{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata validation - wave 1\n",
    "\n",
    "# ASAP CRN Metadata validation - wave 1\n",
    "\n",
    "15 September 2023\n",
    "Andy Henrie\n",
    "\n",
    "\n",
    "\n",
    "## STEPS\n",
    "\n",
    "### imports\n",
    "- pandas\n",
    "- pathlib\n",
    "\n",
    "### Load CDE for validation\n",
    "- check all columns1\n",
    "\n",
    "### Team Lee\n",
    "- load .tsv, csv tables\n",
    "- fix format\n",
    "- load additional metadata\n",
    "\n",
    "- add batch columns\n",
    "- add missing columns\n",
    "\n",
    "\n",
    "### Team Hafler\n",
    "- load excel file with tables\n",
    "- add batch info\n",
    "- add missing columns\n",
    "\n",
    "\n",
    "\n",
    "### Team Hardy\n",
    "- load excel file with tables\n",
    "- add batch info\n",
    "- add missing columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from pathlib import Path\n",
    "import scanpy as sc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load CDE\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Table</th>\n",
       "      <th>Field</th>\n",
       "      <th>Description</th>\n",
       "      <th>DataType</th>\n",
       "      <th>Required</th>\n",
       "      <th>Validation</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>ClinPath field</th>\n",
       "      <th>team_Hafler type</th>\n",
       "      <th>ClinPath description</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_name</td>\n",
       "      <td>Project Name:   A Title of the overall project...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Dataset Name:   A unique name is required for ...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_description</td>\n",
       "      <td>Project Description:   Brief description of th...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP Team Name:   Name of the ASAP CRN Team. i...</td>\n",
       "      <td>Enum</td>\n",
       "      <td>Required</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>Lab Name. :   Lab name that is submitting data...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Table                Field   \n",
       "0  STUDY         project_name  \\\n",
       "1  STUDY      project_dataset   \n",
       "2  STUDY  project_description   \n",
       "3  STUDY       ASAP_team_name   \n",
       "4  STUDY        ASAP_lab_name   \n",
       "\n",
       "                                         Description DataType  Required   \n",
       "0  Project Name:   A Title of the overall project...   String  Required  \\\n",
       "1  Dataset Name:   A unique name is required for ...   String  Required   \n",
       "2  Project Description:   Brief description of th...   String  Required   \n",
       "3  ASAP Team Name:   Name of the ASAP CRN Team. i...     Enum  Required   \n",
       "4  Lab Name. :   Lab name that is submitting data...   String  Required   \n",
       "\n",
       "                                     Validation  Unnamed: 6 ClinPath field   \n",
       "0                                           NaN         NaN            NaN  \\\n",
       "1                                           NaN         NaN            NaN   \n",
       "2                                           NaN         NaN            NaN   \n",
       "3  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]         NaN            NaN   \n",
       "4                                           NaN         NaN            NaN   \n",
       "\n",
       "  team_Hafler type ClinPath description Unnamed: 10  \n",
       "0              NaN                  NaN        bard  \n",
       "1              NaN                  NaN         NaN  \n",
       "2              NaN                  NaN         NaN  \n",
       "3              NaN                  NaN         NaN  \n",
       "4              NaN                  NaN         NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = Path.home() / (\"Projects/ASAP/harmonized-wf-dev\")\n",
    "CDE = pd.read_csv(data_path / \"metadata/ASAP_CDE_final09152023.csv\")\n",
    "\n",
    "CDE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Field</th>\n",
       "      <th>Validation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>project_name</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>project_dataset</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>project_description</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PI_full_name</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PI_email</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>contributor_names</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>submitter_name</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>submittor_email</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ASAP_grant_id</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>other_funding_source</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>publication_DOI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>publication_PMID</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>number_of_brain_samples</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>brain_regions</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>types_of_samples</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PI_ORCHID</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PI_google_scholar_id</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>DUA_version</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>metadata_version_date</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Field                                    Validation\n",
       "0              project_name                                           NaN\n",
       "1           project_dataset                                           NaN\n",
       "2       project_description                                           NaN\n",
       "3            ASAP_team_name  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]\n",
       "4             ASAP_lab_name                                           NaN\n",
       "5              PI_full_name                                           NaN\n",
       "6                  PI_email                                           NaN\n",
       "7         contributor_names                                           NaN\n",
       "8            submitter_name                                           NaN\n",
       "9           submittor_email                                           NaN\n",
       "10            ASAP_grant_id                                           NaN\n",
       "11     other_funding_source                                           NaN\n",
       "12          publication_DOI                                           NaN\n",
       "13         publication_PMID                                           NaN\n",
       "14  number_of_brain_samples                                           NaN\n",
       "15            brain_regions                                           NaN\n",
       "16         types_of_samples                                           NaN\n",
       "17                PI_ORCHID                                           NaN\n",
       "18     PI_google_scholar_id                                           NaN\n",
       "19              DUA_version                                           NaN\n",
       "20    metadata_version_date                                           NaN"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_name = 'STUDY'\n",
    "validate = CDE[CDE['Table'] == table_name][['Field', 'Validation']]\n",
    "validate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def validate_table(table, table_name, CDE):\n",
    "    \"\"\" validate the table against the CDE (i.e. CDE[table_name])\n",
    "    \"\"\"\n",
    "    validate = CDE[CDE['Table'] == table_name][['Field', 'DataType','Validation']]\n",
    "\n",
    "    for _, row in table.iterrows():\n",
    "        if row['Field'] not in table.columns:\n",
    "            print(f\"Field {row['Field']} not found in {table_name}.csv\")\n",
    "            print(f\"adding empty {row['Field']}\")\n",
    "            table[row['Field']] = \"\"\n",
    "        # apply validation field to whats in the table in that row.\n",
    "        col_vals = table[row['Field']]\n",
    "        # if row['DataType'] == 'Enum':\n",
    "\n",
    "\n",
    "    return table\n",
    "\n",
    "\n",
    "\n",
    "    # # Iterate over each unique table\n",
    "    # for table_name, table_data in CDE.groupby(table):\n",
    "        \n",
    "    #     # For each field in the table\n",
    "    #     for _, row in table_data.iterrows():\n",
    "            \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Team Lee"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert to seurat Object\n",
    "data_path = Path.home() / (\"Projects/ASAP/team-lee\")\n",
    "metadata_path = data_path / \"metadata/ogmetadata\"\n",
    "\n",
    "SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.tsv\", delimiter=\"\\t\")\n",
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.tsv\",delimiter=\"\\t\")\n",
    "\n",
    "CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\",delimiter=\",\")\n",
    "STUDY = pd.read_csv(f\"{metadata_path}/STUDY.tsv\",delimiter=\"\\t\")\n",
    "PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.tsv\",delimiter=\"\\t\")\n",
    "\n",
    "# # fix STUDY & PROTOCOL\n",
    "# PROTOCOL.to_csv(f\"{data_path}/PROTOCOL_.csv\", index=False)\n",
    "# CLINPATH.to_csv(f\"{data_path}/CLINPATH_.csv\", index=False)\n",
    "# SAMPLE.to_csv(f\"{data_path}/SAMPLE_.csv\", index=False)\n",
    "# SUBJECT.to_csv(f\"{data_path}/SUBJECT_.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_collection_summary</th>\n",
       "      <th>cell_extraction_summary</th>\n",
       "      <th>lib_prep_summary</th>\n",
       "      <th>data_processing_summary</th>\n",
       "      <th>github_url</th>\n",
       "      <th>protocols_io_DOI</th>\n",
       "      <th>other_reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>The samples were collected by the Banner Sun H...</td>\n",
       "      <td>See protocol.</td>\n",
       "      <td>Library preps were generated via the 5' 10x Ge...</td>\n",
       "      <td>We use cellranger count (6.0.1) to align and g...</td>\n",
       "      <td>private repo at https://github.com/TheGBLab/ASAP</td>\n",
       "      <td>https://dx.doi.org/10.17504/protocols.io.bzmgp43w</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           sample_collection_summary cell_extraction_summary   \n",
       "0  The samples were collected by the Banner Sun H...           See protocol.  \\\n",
       "\n",
       "                                    lib_prep_summary   \n",
       "0  Library preps were generated via the 5' 10x Ge...  \\\n",
       "\n",
       "                             data_processing_summary   \n",
       "0  We use cellranger count (6.0.1) to align and g...  \\\n",
       "\n",
       "                                         github_url   \n",
       "0  private repo at https://github.com/TheGBLab/ASAP  \\\n",
       "\n",
       "                                    protocols_io_DOI  other_reference  \n",
       "0  https://dx.doi.org/10.17504/protocols.io.bzmgp43w              NaN  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PROTOCOL.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "STUDY = pd.read_csv(metadata_path / \"STUDY.tsv\",delimiter=\"\\t\")\n",
    "STUDY.to_csv(data_path / \"STUDY_.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Team-Lee-Bras-Lab-Info</th>\n",
       "      <th>Field</th>\n",
       "      <th>Description</th>\n",
       "      <th>Data type</th>\n",
       "      <th>Validation</th>\n",
       "      <th>Note</th>\n",
       "      <th>Required/Optional</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>project_name</td>\n",
       "      <td>Project Name/Title</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unique and clear title.</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Dataset name</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A Dataset name is required for each submission...</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>project_description</td>\n",
       "      <td>Brief description of the goals and objectives ...</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP Team e.g. \"Scherzer\"</td>\n",
       "      <td>Enum</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bras</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>ASAP Lab under the above team e.g. \"Dong\"</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Unnamed: 0           Unnamed: 1   \n",
       "0  Is senescence a component of human PD and does...         project_name  \\\n",
       "1    Human snRNA-seq PD Senesence Jose Bras Team Lee      project_dataset   \n",
       "2  Characterize the neuropathological progression...  project_description   \n",
       "3                                           TEAM-LEE       ASAP_team_name   \n",
       "4                                               Bras        ASAP_lab_name   \n",
       "\n",
       "                                          Unnamed: 2 Team-Lee-Bras-Lab-Info   \n",
       "0                                 Project Name/Title                 String  \\\n",
       "1                                       Dataset name                 String   \n",
       "2  Brief description of the goals and objectives ...                 String   \n",
       "3                          ASAP Team e.g. \"Scherzer\"                   Enum   \n",
       "4          ASAP Lab under the above team e.g. \"Dong\"                 String   \n",
       "\n",
       "                                          Field   \n",
       "0                                           NaN  \\\n",
       "1                                           NaN   \n",
       "2                                           NaN   \n",
       "3  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]   \n",
       "4                                           NaN   \n",
       "\n",
       "                                         Description Data type  Validation   \n",
       "0                            Unique and clear title.  Required         NaN  \\\n",
       "1  A Dataset name is required for each submission...  Required         NaN   \n",
       "2                                                NaN  Required         NaN   \n",
       "3                                                NaN  Required         NaN   \n",
       "4                                                NaN  Required         NaN   \n",
       "\n",
       "   Note  Required/Optional  \n",
       "0   NaN                NaN  \n",
       "1   NaN                NaN  \n",
       "2   NaN                NaN  \n",
       "3   NaN                NaN  \n",
       "4   NaN                NaN  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "STUDY = pd.read_csv(data_path / \"STUDY_.csv\")\n",
    "STUDY.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submittor_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>project_name</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>project_description</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>PI_full_name</td>\n",
       "      <td>PI_email</td>\n",
       "      <td>submitter_id</td>\n",
       "      <td>submitter_name</td>\n",
       "      <td>submittor_email</td>\n",
       "      <td>...</td>\n",
       "      <td>other_funding_source</td>\n",
       "      <td>publication_DOI</td>\n",
       "      <td>publication_PMID</td>\n",
       "      <td>number_of_brain_samples</td>\n",
       "      <td>brain_regions</td>\n",
       "      <td>types_of_samples</td>\n",
       "      <td>PI_ORCHID</td>\n",
       "      <td>PI_google_scholar_id</td>\n",
       "      <td>DUA_version</td>\n",
       "      <td>metadata_version_date</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>Bras</td>\n",
       "      <td>Jose, Bras</td>\n",
       "      <td>jose.bras@vai.org</td>\n",
       "      <td>Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...</td>\n",
       "      <td>Kaitlyn E Westra</td>\n",
       "      <td>kaitlyn.westra@vai.org</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75</td>\n",
       "      <td>hippocampus; middle frontal gyrus; substantia ...</td>\n",
       "      <td>human PD and control postmortem brains</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unsure</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name   \n",
       "0                                       project_name  \\\n",
       "1  Is senescence a component of human PD and does...   \n",
       "\n",
       "0                                  project_dataset   \n",
       "0                                  project_dataset  \\\n",
       "1  Human snRNA-seq PD Senesence Jose Bras Team Lee   \n",
       "\n",
       "0                                project_description  ASAP_team_name   \n",
       "0                                project_description  ASAP_team_name  \\\n",
       "1  Characterize the neuropathological progression...        TEAM-LEE   \n",
       "\n",
       "0  ASAP_lab_name  PI_full_name           PI_email   \n",
       "0  ASAP_lab_name  PI_full_name           PI_email  \\\n",
       "1           Bras    Jose, Bras  jose.bras@vai.org   \n",
       "\n",
       "0                                       submitter_id    submitter_name   \n",
       "0                                       submitter_id    submitter_name  \\\n",
       "1  Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...  Kaitlyn E Westra   \n",
       "\n",
       "0         submittor_email  ...  other_funding_source  publication_DOI   \n",
       "0         submittor_email  ...  other_funding_source  publication_DOI  \\\n",
       "1  kaitlyn.westra@vai.org  ...                   NaN              NaN   \n",
       "\n",
       "0  publication_PMID  number_of_brain_samples   \n",
       "0  publication_PMID  number_of_brain_samples  \\\n",
       "1               NaN                       75   \n",
       "\n",
       "0                                      brain_regions   \n",
       "0                                      brain_regions  \\\n",
       "1  hippocampus; middle frontal gyrus; substantia ...   \n",
       "\n",
       "0                        types_of_samples  PI_ORCHID  PI_google_scholar_id   \n",
       "0                        types_of_samples  PI_ORCHID  PI_google_scholar_id  \\\n",
       "1  human PD and control postmortem brains        NaN                   NaN   \n",
       "\n",
       "0  DUA_version  metadata_version_date  \n",
       "0  DUA_version  metadata_version_date  \n",
       "1       unsure                    NaN  \n",
       "\n",
       "[2 rows x 21 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "tmp = pd.DataFrame()\n",
    "tmp = STUDY[[\"Unnamed: 1\",\"Unnamed: 0\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "tmp.columns = tmp.iloc[0]\n",
    "STUDY = tmp.drop([0])\n",
    "# STUDY[[\"Unnamed: 1\"]].transpose().reset_index().drop(columns=[\"index\"]), tmp\n",
    "\n",
    "tmp.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submittor_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>Bras</td>\n",
       "      <td>Jose, Bras</td>\n",
       "      <td>jose.bras@vai.org</td>\n",
       "      <td>Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...</td>\n",
       "      <td>Kaitlyn E Westra</td>\n",
       "      <td>kaitlyn.westra@vai.org</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75</td>\n",
       "      <td>hippocampus; middle frontal gyrus; substantia ...</td>\n",
       "      <td>human PD and control postmortem brains</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unsure</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name   \n",
       "1  Is senescence a component of human PD and does...  \\\n",
       "\n",
       "0                                  project_dataset   \n",
       "1  Human snRNA-seq PD Senesence Jose Bras Team Lee  \\\n",
       "\n",
       "0                                project_description ASAP_team_name   \n",
       "1  Characterize the neuropathological progression...       TEAM-LEE  \\\n",
       "\n",
       "0 ASAP_lab_name PI_full_name           PI_email   \n",
       "1          Bras   Jose, Bras  jose.bras@vai.org  \\\n",
       "\n",
       "0                                       submitter_id    submitter_name   \n",
       "1  Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...  Kaitlyn E Westra  \\\n",
       "\n",
       "0         submittor_email  ... other_funding_source publication_DOI   \n",
       "1  kaitlyn.westra@vai.org  ...                  NaN             NaN  \\\n",
       "\n",
       "0 publication_PMID number_of_brain_samples   \n",
       "1              NaN                      75  \\\n",
       "\n",
       "0                                      brain_regions   \n",
       "1  hippocampus; middle frontal gyrus; substantia ...  \\\n",
       "\n",
       "0                        types_of_samples PI_ORCHID PI_google_scholar_id   \n",
       "1  human PD and control postmortem brains       NaN                  NaN  \\\n",
       "\n",
       "0 DUA_version metadata_version_date  \n",
       "1      unsure                   NaN  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Field'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniforge3/envs/scverse10/lib/python3.10/site-packages/pandas/core/indexes/base.py:3652\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3651\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 3652\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_engine\u001b[39m.\u001b[39;49mget_loc(casted_key)\n\u001b[1;32m   3653\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/miniforge3/envs/scverse10/lib/python3.10/site-packages/pandas/_libs/index.pyx:147\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/miniforge3/envs/scverse10/lib/python3.10/site-packages/pandas/_libs/index.pyx:176\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7080\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7088\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Field'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[64], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m newSTUDY \u001b[39m=\u001b[39m validate_table(STUDY, \u001b[39m\"\u001b[39;49m\u001b[39mSTUDY\u001b[39;49m\u001b[39m\"\u001b[39;49m, CDE)\n",
      "Cell \u001b[0;32mIn[55], line 7\u001b[0m, in \u001b[0;36mvalidate_table\u001b[0;34m(table, table_name, CDE)\u001b[0m\n\u001b[1;32m      4\u001b[0m validate \u001b[39m=\u001b[39m CDE[CDE[\u001b[39m'\u001b[39m\u001b[39mTable\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m==\u001b[39m table_name][[\u001b[39m'\u001b[39m\u001b[39mField\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mDataType\u001b[39m\u001b[39m'\u001b[39m,\u001b[39m'\u001b[39m\u001b[39mValidation\u001b[39m\u001b[39m'\u001b[39m]]\n\u001b[1;32m      6\u001b[0m \u001b[39mfor\u001b[39;00m _, row \u001b[39min\u001b[39;00m table\u001b[39m.\u001b[39miterrows():\n\u001b[0;32m----> 7\u001b[0m     \u001b[39mif\u001b[39;00m row[\u001b[39m'\u001b[39;49m\u001b[39mField\u001b[39;49m\u001b[39m'\u001b[39;49m] \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m table\u001b[39m.\u001b[39mcolumns:\n\u001b[1;32m      8\u001b[0m         \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mField \u001b[39m\u001b[39m{\u001b[39;00mrow[\u001b[39m'\u001b[39m\u001b[39mField\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m}\u001b[39;00m\u001b[39m not found in \u001b[39m\u001b[39m{\u001b[39;00mtable_name\u001b[39m}\u001b[39;00m\u001b[39m.csv\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m      9\u001b[0m         \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39madding empty \u001b[39m\u001b[39m{\u001b[39;00mrow[\u001b[39m'\u001b[39m\u001b[39mField\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[0;32m~/miniforge3/envs/scverse10/lib/python3.10/site-packages/pandas/core/series.py:1007\u001b[0m, in \u001b[0;36mSeries.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1004\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_values[key]\n\u001b[1;32m   1006\u001b[0m \u001b[39melif\u001b[39;00m key_is_scalar:\n\u001b[0;32m-> 1007\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_get_value(key)\n\u001b[1;32m   1009\u001b[0m \u001b[39mif\u001b[39;00m is_hashable(key):\n\u001b[1;32m   1010\u001b[0m     \u001b[39m# Otherwise index.get_value will raise InvalidIndexError\u001b[39;00m\n\u001b[1;32m   1011\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m   1012\u001b[0m         \u001b[39m# For labels that don't resolve as scalars like tuples and frozensets\u001b[39;00m\n",
      "File \u001b[0;32m~/miniforge3/envs/scverse10/lib/python3.10/site-packages/pandas/core/series.py:1116\u001b[0m, in \u001b[0;36mSeries._get_value\u001b[0;34m(self, label, takeable)\u001b[0m\n\u001b[1;32m   1113\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_values[label]\n\u001b[1;32m   1115\u001b[0m \u001b[39m# Similar to Index.get_value, but we do not fall back to positional\u001b[39;00m\n\u001b[0;32m-> 1116\u001b[0m loc \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mindex\u001b[39m.\u001b[39;49mget_loc(label)\n\u001b[1;32m   1118\u001b[0m \u001b[39mif\u001b[39;00m is_integer(loc):\n\u001b[1;32m   1119\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_values[loc]\n",
      "File \u001b[0;32m~/miniforge3/envs/scverse10/lib/python3.10/site-packages/pandas/core/indexes/base.py:3654\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3652\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_engine\u001b[39m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3653\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n\u001b[0;32m-> 3654\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(key) \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m   3655\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mTypeError\u001b[39;00m:\n\u001b[1;32m   3656\u001b[0m     \u001b[39m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3657\u001b[0m     \u001b[39m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3658\u001b[0m     \u001b[39m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3659\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Field'"
     ]
    }
   ],
   "source": [
    "newSTUDY = validate_table(STUDY, \"STUDY\", CDE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>06-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>00-09</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>19-21</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFG_PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>20-58</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id source_sample_id subject_id replicate  replicate_count   \n",
       "0  MFG_HC_1225            12-25      12-25      rep1                1  \\\n",
       "1  MFG_HC_0602            06-02      06-02      rep1                1   \n",
       "2  MFG_PD_0009            00-09      00-09      rep1                1   \n",
       "3  MFG_PD_1921            19-21      19-21      rep1                1   \n",
       "4  MFG_PD_2058            20-58      20-58      rep1                1   \n",
       "\n",
       "   repeated_sample tissue          brain_region  source_RIN  RIN  ...   \n",
       "0                0  brain  Middle_Frontal_Gyrus         NaN  NaN  ...  \\\n",
       "1                0  brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "2                0  brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "3                0  brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "4                0  brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "\n",
       "  sex_ontology_term_id  self_reported_ethnicity_ontology_term_id   \n",
       "0  PATO:0000384 (male)                                   unknown  \\\n",
       "1  PATO:0000384 (male)                                   unknown   \n",
       "2  PATO:0000384 (male)                                   unknown   \n",
       "3  PATO:0000384 (male)                                   unknown   \n",
       "4  PATO:0000384 (male)                                   unknown   \n",
       "\n",
       "  disease_ontology_term_id tissue_ontology_term_id   \n",
       "0             PATO:0000461          UBERON:0002702  \\\n",
       "1             PATO:0000461          UBERON:0002702   \n",
       "2            MONDO:0005180          UBERON:0002702   \n",
       "3            MONDO:0005180          UBERON:0002702   \n",
       "4            MONDO:0005180          UBERON:0002702   \n",
       "\n",
       "   cell_type_ontology_term_id assay_ontology_term_id suspension_type DV2000   \n",
       "0                NA(multiple)            EFO:0030004         nucleus    NaN  \\\n",
       "1                NA(multiple)            EFO:0030004         nucleus    NaN   \n",
       "2                NA(multiple)            EFO:0030004         nucleus    NaN   \n",
       "3                NA(multiple)            EFO:0030004         nucleus    NaN   \n",
       "4                NA(multiple)            EFO:0030004         nucleus    NaN   \n",
       "\n",
       "  pm_PH donor_id  \n",
       "0   NaN      NaN  \n",
       "1   NaN      NaN  \n",
       "2   NaN      NaN  \n",
       "3   NaN      NaN  \n",
       "4   NaN      NaN  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path / \"metadata/PROTOCOL.csv\"\n",
    "STUDY = pd.read_csv(data_path / \"metadata/STUDY.csv\")\n",
    "\n",
    "SAMPLE.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>region_level_1</th>\n",
       "      <th>region_level_2</th>\n",
       "      <th>region_level_3</th>\n",
       "      <th>AMPPD_id</th>\n",
       "      <th>family_history</th>\n",
       "      <th>...</th>\n",
       "      <th>sn_neuronal_loss</th>\n",
       "      <th>path_infarcs</th>\n",
       "      <th>path_nia_ri</th>\n",
       "      <th>path_nia_aa_a</th>\n",
       "      <th>path_nia_aa_b</th>\n",
       "      <th>path_nia_aa_c</th>\n",
       "      <th>TDP43</th>\n",
       "      <th>arteriolosclerosis_severity_scale</th>\n",
       "      <th>amyloid_angiopathy_severity_scale</th>\n",
       "      <th>path_ad_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not AD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFG_PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id source_sample_id  time_from_baseline  GP2_id hemisphere   \n",
       "0  MFG_HC_1225            12-25                   0     NaN    Unknown  \\\n",
       "1  MFG_HC_0602            06-02                   0     NaN    Unknown   \n",
       "2  MFG_PD_0009            00-09                   0     NaN    Unknown   \n",
       "3  MFG_PD_1921            19-21                   0     NaN    Unknown   \n",
       "4  MFG_PD_2058            20-58                   0     NaN    Unknown   \n",
       "\n",
       "  region_level_1        region_level_2 region_level_3  AMPPD_id   \n",
       "0   Frontal lobe  Middle frontal gyrus        unknown       NaN  \\\n",
       "1   Frontal lobe  Middle frontal gyrus        unknown       NaN   \n",
       "2   Frontal lobe  Middle frontal gyrus        unknown       NaN   \n",
       "3   Frontal lobe  Middle frontal gyrus        unknown       NaN   \n",
       "4   Frontal lobe  Middle frontal gyrus        unknown       NaN   \n",
       "\n",
       "  family_history  ...  sn_neuronal_loss  path_infarcs       path_nia_ri   \n",
       "0   Not Reported  ...               NaN           NaN  Criteria not met  \\\n",
       "1   Not Reported  ...               NaN           NaN  Criteria not met   \n",
       "2   Not Reported  ...               NaN           NaN            Not AD   \n",
       "3   Not Reported  ...               NaN           NaN  Criteria not met   \n",
       "4   Not Reported  ...               NaN           NaN               Low   \n",
       "\n",
       "   path_nia_aa_a path_nia_aa_b  path_nia_aa_c                         TDP43   \n",
       "0            NaN           NaN            NaN                           NaN  \\\n",
       "1            NaN           NaN            NaN                           NaN   \n",
       "2            NaN           NaN            NaN                           NaN   \n",
       "3            NaN           NaN            NaN  None in medial temporal lobe   \n",
       "4            NaN           NaN            NaN  None in medial temporal lobe   \n",
       "\n",
       "  arteriolosclerosis_severity_scale  amyloid_angiopathy_severity_scale   \n",
       "0                               NaN                                NaN  \\\n",
       "1                               NaN                                NaN   \n",
       "2                               NaN                                NaN   \n",
       "3                               NaN                                NaN   \n",
       "4                               NaN                                NaN   \n",
       "\n",
       "                                       path_ad_level  \n",
       "0  No evidence of Alzheimer's disease neuropathol...  \n",
       "1  No evidence of Alzheimer's disease neuropathol...  \n",
       "2  No evidence of Alzheimer's disease neuropathol...  \n",
       "3  No evidence of Alzheimer's disease neuropathol...  \n",
       "4  Microscopic changes of Alzheimer's disease, in...  \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write clean metadata tables accourding to CDE "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "STUDY.to_csv(data_path / \"metadata/STUDY.csv\")\n",
    "PROTOCOL.to_csv(data_path / \"metadata/PROTOCOL.csv\")\n",
    "CLINPATH.to_csv(data_path / \"metadata/CLINPATH.csv\")\n",
    "SAMPLE.to_csv(data_path / \"metadata/SAMPLE.csv\")\n",
    "\n",
    "# fix SUBJECT before \n",
    "SUBJECT.to_csv(data_path / \"metadata/SUBJECT.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>06-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>00-09</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>MFG_PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>19-21</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>MFG_PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>20-58</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0.1  Unnamed: 0    sample_id source_sample_id subject_id   \n",
       "0             0           0  MFG_HC_1225            12-25      12-25  \\\n",
       "1             1           1  MFG_HC_0602            06-02      06-02   \n",
       "2             2           2  MFG_PD_0009            00-09      00-09   \n",
       "3             3           3  MFG_PD_1921            19-21      19-21   \n",
       "4             4           4  MFG_PD_2058            20-58      20-58   \n",
       "\n",
       "  replicate  replicate_count  repeated_sample tissue          brain_region   \n",
       "0      rep1                1                0  brain  Middle_Frontal_Gyrus  \\\n",
       "1      rep1                1                0  brain  Middle_Frontal_Gyrus   \n",
       "2      rep1                1                0  brain  Middle_Frontal_Gyrus   \n",
       "3      rep1                1                0  brain  Middle_Frontal_Gyrus   \n",
       "4      rep1                1                0  brain  Middle_Frontal_Gyrus   \n",
       "\n",
       "   ...  sex_ontology_term_id  self_reported_ethnicity_ontology_term_id   \n",
       "0  ...   PATO:0000384 (male)                                   unknown  \\\n",
       "1  ...   PATO:0000384 (male)                                   unknown   \n",
       "2  ...   PATO:0000384 (male)                                   unknown   \n",
       "3  ...   PATO:0000384 (male)                                   unknown   \n",
       "4  ...   PATO:0000384 (male)                                   unknown   \n",
       "\n",
       "  disease_ontology_term_id  tissue_ontology_term_id   \n",
       "0             PATO:0000461           UBERON:0002702  \\\n",
       "1             PATO:0000461           UBERON:0002702   \n",
       "2            MONDO:0005180           UBERON:0002702   \n",
       "3            MONDO:0005180           UBERON:0002702   \n",
       "4            MONDO:0005180           UBERON:0002702   \n",
       "\n",
       "  cell_type_ontology_term_id assay_ontology_term_id  suspension_type DV2000   \n",
       "0               NA(multiple)            EFO:0030004          nucleus    NaN  \\\n",
       "1               NA(multiple)            EFO:0030004          nucleus    NaN   \n",
       "2               NA(multiple)            EFO:0030004          nucleus    NaN   \n",
       "3               NA(multiple)            EFO:0030004          nucleus    NaN   \n",
       "4               NA(multiple)            EFO:0030004          nucleus    NaN   \n",
       "\n",
       "  pm_PH donor_id  \n",
       "0   NaN      NaN  \n",
       "1   NaN      NaN  \n",
       "2   NaN      NaN  \n",
       "3   NaN      NaN  \n",
       "4   NaN      NaN  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = Path.home() / (\"Projects/ASAP/team-lee\")\n",
    "metadata_path = data_path / \"metadata\"\n",
    "\n",
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.csv\",delimiter=\",\")\n",
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "metadata_path = Path.home() / (\"Projects/ASAP/team-lee/metadata\")\n",
    "HIP_covar = pd.read_csv(f\"{metadata_path}/HIP/covar.csv\")\n",
    "HIP_cases = pd.read_csv(f\"{metadata_path}/HIP/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "HIP_control = pd.read_csv(f\"{metadata_path}/HIP/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n",
    "\n",
    "MFG_covar = pd.read_csv(f\"{metadata_path}/MFG/covar.csv\") # includes 'PMI' ?\n",
    "MFG_cases = pd.read_csv(f\"{metadata_path}/MFG/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "MFG_control = pd.read_csv(f\"{metadata_path}/MFG/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n",
    "\n",
    "\n",
    "SN_covar = pd.read_csv(f\"{metadata_path}/SN/covar.csv\")\n",
    "SN_cases = pd.read_csv(f\"{metadata_path}/SN/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "SN_control = pd.read_csv(f\"{metadata_path}/SN/PD_ASAP_Sample_batch_information_banner_controls.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hippocampus samples\n",
    "\n",
    "\n",
    "WARNING:  there is an error here in merging cases and controls... the *_meta tables are NOT correct.  Just using covar for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "HIP_cases[\"GROUPcv\"]=\"PD\"\n",
    "HIP_control[\"GROUPcv\"]=\"HC\"\n",
    "HIP_meta = pd.concat([HIP_cases, HIP_control], axis=0, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "HIP_meta['MERGE_ID'] = \"HIP_\" + HIP_meta['GROUPcv'] +\"_\" + HIP_meta['CaseID'].str.replace('-','')\n",
    "HIP_covar['MERGE_ID'] = HIP_covar['COUNT_ID']\n",
    "# the fastqs follow COUNT_ID insteald of SEQ_ID naming convention\n",
    "HIP_covar['SEQ_ID'] = HIP_covar['COUNT_ID']\n",
    "\n",
    "\n",
    "HIP_TABLE = pd.merge(HIP_covar, HIP_meta, on='MERGE_ID', how='inner')\n",
    "HIP_TABLE.columns\n",
    "\n",
    "HIP_TABLE['subdir']=\"HIP\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### medial frontal gyrus samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "MFG_cases[\"GROUPcv\"]=\"PD\"\n",
    "MFG_control[\"GROUPcv\"]=\"HC\"\n",
    "MFG_meta = pd.concat([MFG_cases, MFG_control], axis=0, ignore_index=True)\n",
    "\n",
    "# make a MERGE_ID column because the formatting is inconsistent\n",
    "MFG_meta['MERGE_ID'] = \"MFG_\" + MFG_meta['GROUPcv'] +\"_\" + MFG_meta['CaseID'].str.replace('-','')\n",
    "MFG_covar['MERGE_ID'] = MFG_covar['SAMPLE']\n",
    "# the fastqs are in SEQ_ID \n",
    "\n",
    "MFG_TABLE = pd.merge(MFG_covar, MFG_meta, on='MERGE_ID', how='inner')\n",
    "MFG_TABLE['subdir']=\"MFG\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Substantia Nigra samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BRAIN_REGION</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>ID</th>\n",
       "      <th>SEQ_ID</th>\n",
       "      <th>BATCH</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>DIAGNOSIS</th>\n",
       "      <th>SEX</th>\n",
       "      <th>AGE</th>\n",
       "      <th>DIAGNOSIS_AGE</th>\n",
       "      <th>DIAGNOSIS_YEARS</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>ALZHEIMERS_BRAAK</th>\n",
       "      <th>DEMENTIA</th>\n",
       "      <th>DEMENTIA_AGE</th>\n",
       "      <th>DEMENTIA_YEARS</th>\n",
       "      <th>WEIGHT_USED</th>\n",
       "      <th>TOTAL_WEIGHT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Substantia_nigra</td>\n",
       "      <td>SN_HC_1939</td>\n",
       "      <td>HC_19-39</td>\n",
       "      <td>SN_1939_HC</td>\n",
       "      <td>BATCH_5</td>\n",
       "      <td>HC</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>93</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LBS</td>\n",
       "      <td>II</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Substantia_nigra</td>\n",
       "      <td>SN_HC_0602</td>\n",
       "      <td>MCI_06-02</td>\n",
       "      <td>SN_0602_HC</td>\n",
       "      <td>BATCH_5</td>\n",
       "      <td>HC</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>84</td>\n",
       "      <td>81.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>LBS</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Substantia_nigra</td>\n",
       "      <td>SN_PD_0413</td>\n",
       "      <td>PD_04-13</td>\n",
       "      <td>SN_0413_PD</td>\n",
       "      <td>BATCH_5</td>\n",
       "      <td>PD</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>72</td>\n",
       "      <td>62.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>PD</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Substantia_nigra</td>\n",
       "      <td>SN_PD_1317</td>\n",
       "      <td>PD_13-17</td>\n",
       "      <td>SN_1317_PD</td>\n",
       "      <td>BATCH_5</td>\n",
       "      <td>PD</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>83</td>\n",
       "      <td>70.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>PD</td>\n",
       "      <td>IV</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Substantia_nigra</td>\n",
       "      <td>SN_PD_1973</td>\n",
       "      <td>PD_19-73</td>\n",
       "      <td>SN_1973_PD</td>\n",
       "      <td>BATCH_5</td>\n",
       "      <td>PD</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>84</td>\n",
       "      <td>70.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>PD</td>\n",
       "      <td>IV</td>\n",
       "      <td>1</td>\n",
       "      <td>79.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>40</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       BRAIN_REGION      SAMPLE         ID      SEQ_ID    BATCH GROUP   \n",
       "0  Substantia_nigra  SN_HC_1939   HC_19-39  SN_1939_HC  BATCH_5    HC  \\\n",
       "1  Substantia_nigra  SN_HC_0602  MCI_06-02  SN_0602_HC  BATCH_5    HC   \n",
       "2  Substantia_nigra  SN_PD_0413   PD_04-13  SN_0413_PD  BATCH_5    PD   \n",
       "3  Substantia_nigra  SN_PD_1317   PD_13-17  SN_1317_PD  BATCH_5    PD   \n",
       "4  Substantia_nigra  SN_PD_1973   PD_19-73  SN_1973_PD  BATCH_5    PD   \n",
       "\n",
       "   DIAGNOSIS  SEX  AGE  DIAGNOSIS_AGE  DIAGNOSIS_YEARS DISEASE   \n",
       "0          0    2   93            NaN              NaN     LBS  \\\n",
       "1          0    1   84           81.0              3.0     LBS   \n",
       "2          1    1   72           62.0             10.0      PD   \n",
       "3          1    1   83           70.0             13.0      PD   \n",
       "4          1    1   84           70.0             14.0      PD   \n",
       "\n",
       "  ALZHEIMERS_BRAAK  DEMENTIA  DEMENTIA_AGE  DEMENTIA_YEARS  WEIGHT_USED   \n",
       "0               II         0           NaN             NaN           27  \\\n",
       "1                I         0           NaN             NaN           25   \n",
       "2                I         0           NaN             NaN           33   \n",
       "3               IV         0           NaN             NaN           30   \n",
       "4               IV         1          79.0             5.0           40   \n",
       "\n",
       "   TOTAL_WEIGHT  \n",
       "0            27  \n",
       "1            46  \n",
       "2            33  \n",
       "3            30  \n",
       "4            65  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SN_covar.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "SN_cases[\"GROUPcv\"]=\"PD\"\n",
    "SN_control[\"GROUPcv\"]=\"HC\"\n",
    "SN_meta = pd.concat([SN_cases, SN_control], axis=0, ignore_index=True)\n",
    "\n",
    "SN_meta['MERGE_ID'] = \"SN_\" + MFG_meta['GROUPcv'] +\"_\" + MFG_meta['CaseID'].str.replace('-','')\n",
    "SN_covar['MERGE_ID'] = SN_covar['SAMPLE']\n",
    "\n",
    "SN_TABLE = pd.merge(SN_covar, SN_meta, on='MERGE_ID', how='inner')\n",
    "SN_TABLE['subdir']=\"SN\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### concatenate SN, MSG, and HIP tables into one 'all_samples' table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_samples = pd.concat([HIP_TABLE, MFG_TABLE, SN_TABLE], axis=0, ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BRAIN_REGION</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>COUNT_ID</th>\n",
       "      <th>SAMPLE_FULL</th>\n",
       "      <th>ID</th>\n",
       "      <th>SEQ_ID</th>\n",
       "      <th>BATCH</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>DIAGNOSIS</th>\n",
       "      <th>SEX</th>\n",
       "      <th>...</th>\n",
       "      <th>neurological_dx</th>\n",
       "      <th>neurological_dx_age</th>\n",
       "      <th>dementia_yn</th>\n",
       "      <th>dementia_years</th>\n",
       "      <th>demential_age</th>\n",
       "      <th>GROUPcv</th>\n",
       "      <th>neurological_dx_years</th>\n",
       "      <th>subdir</th>\n",
       "      <th>PMI_x</th>\n",
       "      <th>PMI_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>HC_2067</td>\n",
       "      <td>HIP_HC_2067</td>\n",
       "      <td>HIP_HC_20-67</td>\n",
       "      <td>HC_20-67</td>\n",
       "      <td>HIP_HC_2067</td>\n",
       "      <td>BATCH_8</td>\n",
       "      <td>HC</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>PD_0413</td>\n",
       "      <td>HIP_PD_0413</td>\n",
       "      <td>HIP_PD_04-13</td>\n",
       "      <td>PD_04-13</td>\n",
       "      <td>HIP_PD_0413</td>\n",
       "      <td>BATCH_8</td>\n",
       "      <td>PD</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Parkinson's disease</td>\n",
       "      <td>62.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>PD_1504</td>\n",
       "      <td>HIP_PD_1504</td>\n",
       "      <td>HIP_PD_15-04</td>\n",
       "      <td>PD_15-04</td>\n",
       "      <td>HIP_PD_1504</td>\n",
       "      <td>BATCH_8</td>\n",
       "      <td>PD</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>Parkinson's disease</td>\n",
       "      <td>54.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>HC_2062</td>\n",
       "      <td>HIP_HC_2062</td>\n",
       "      <td>HIP_HC_20-62</td>\n",
       "      <td>HC_20-62</td>\n",
       "      <td>HIP_HC_2062</td>\n",
       "      <td>BATCH_9</td>\n",
       "      <td>HC</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>PD_2038</td>\n",
       "      <td>HIP_PD_2038</td>\n",
       "      <td>HIP_PD_20-38</td>\n",
       "      <td>PD_20-38</td>\n",
       "      <td>HIP_PD_2038</td>\n",
       "      <td>BATCH_9</td>\n",
       "      <td>PD</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Parkinson's disease</td>\n",
       "      <td>66.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>PD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HIP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 95 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  BRAIN_REGION   SAMPLE     COUNT_ID   SAMPLE_FULL        ID       SEQ_ID   \n",
       "0  Hippocampus  HC_2067  HIP_HC_2067  HIP_HC_20-67  HC_20-67  HIP_HC_2067  \\\n",
       "1  Hippocampus  PD_0413  HIP_PD_0413  HIP_PD_04-13  PD_04-13  HIP_PD_0413   \n",
       "2  Hippocampus  PD_1504  HIP_PD_1504  HIP_PD_15-04  PD_15-04  HIP_PD_1504   \n",
       "3  Hippocampus  HC_2062  HIP_HC_2062  HIP_HC_20-62  HC_20-62  HIP_HC_2062   \n",
       "4  Hippocampus  PD_2038  HIP_PD_2038  HIP_PD_20-38  PD_20-38  HIP_PD_2038   \n",
       "\n",
       "     BATCH GROUP  DIAGNOSIS  SEX  ...      neurological_dx   \n",
       "0  BATCH_8    HC          0    1  ...              Control  \\\n",
       "1  BATCH_8    PD          1    1  ...  Parkinson's disease   \n",
       "2  BATCH_8    PD          1    2  ...  Parkinson's disease   \n",
       "3  BATCH_9    HC          0    1  ...              Control   \n",
       "4  BATCH_9    PD          1    1  ...  Parkinson's disease   \n",
       "\n",
       "   neurological_dx_age  dementia_yn dementia_years demential_age  GROUPcv   \n",
       "0                  NaN          0.0            NaN           NaN       HC  \\\n",
       "1                 62.0          0.0            NaN           NaN       PD   \n",
       "2                 54.0          1.0            4.0          68.0       PD   \n",
       "3                  NaN          0.0            NaN           NaN       HC   \n",
       "4                 66.0          1.0            1.0          79.0       PD   \n",
       "\n",
       "   neurological_dx_years  subdir  PMI_x  PMI_y  \n",
       "0                    NaN     HIP    NaN    NaN  \n",
       "1                    NaN     HIP    NaN    NaN  \n",
       "2                    NaN     HIP    NaN    NaN  \n",
       "3                    NaN     HIP    NaN    NaN  \n",
       "4                    NaN     HIP    NaN    NaN  \n",
       "\n",
       "[5 rows x 95 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "all_samples.to_csv(\"team-Lee-all_samples.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_csv = pd.DataFrame()\n",
    "samples_csv['sample'] = all_samples['SEQ_ID']\n",
    "samples_csv['batch'] = all_samples['BATCH']\n",
    "samples_csv['subdir'] = all_samples['subdir']\n",
    "\n",
    "samples_csv.to_csv(\"team-Lee-samples.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Team Hafler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_path = Path.home() / (\"Projects/ASAP/team-hafler\")\n",
    "\n",
    "# metadata_path = data_path / \"metadata\"\n",
    "\n",
    "# SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.csv\")\n",
    "# CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\")\n",
    "# STUDY = pd.read_csv(f\"{metadata_path}/STUDY.csv\")\n",
    "# PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.csv\")\n",
    "\n",
    "# SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.csv\")\n",
    "# SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.csv\")\n",
    "\n",
    "\n",
    "\n",
    "# First batch: HSDG07HC HSDG10HC HSDG148PD HSDG199PD\n",
    "\n",
    "# Second batch: hsDG101HC hsDG13HC hsDG151PD hsDG197PD hsDG30HC hsDG99HC\n",
    "\n",
    "# Third batch: hsDG142PD hsDG208PD\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/team-hafler/metadata')"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## convert to seurat Object\n",
    "data_path = Path.home() / (\"Projects/ASAP\")\n",
    "metadata_path = data_path / \"team-hafler/metadata\"\n",
    "\n",
    "sheets = [\"SAMPLE\",\"SUBJECT\",\"CLINPATH\",\"STUDY\",\"PROTOCOL\"]\n",
    "excel_path = data_path / \"ASAP_CDE_ALL_Team_Hafler_v1.xlsx\"\n",
    "STUDY = pd.read_excel(excel_path,sheet_name=\"STUDY\",header=1).drop(columns=\"Field\")\n",
    "CLINPATH = pd.read_excel(excel_path,sheet_name=\"CLINPATH\",header=1).drop(columns=\"Field\")\n",
    "SUBJECT = pd.read_excel(excel_path,sheet_name=\"SUBJECT\",header=1).drop(columns=\"Field\")\n",
    "SAMPLE = pd.read_excel(excel_path,sheet_name=\"SAMPLE\",header=1).drop(columns=\"Field\")\n",
    "PROTOCOL = pd.read_excel(excel_path,sheet_name=\"PROTOCOL\",header=1).drop(columns=\"Field\")\n",
    "metadata_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Rep3</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Rep3</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id source_sample_id subject_id  replicate replicate_count   \n",
       "0    hSDG07           hSDG07       HC01        NaN            Rep1  \\\n",
       "1    hSDG07           hSDG07       HC01        NaN            Rep1   \n",
       "2    hSDG07           hSDG07       HC01        NaN            Rep1   \n",
       "3   hSDG101          hSDG101       HC03        NaN            Rep3   \n",
       "4   hSDG101          hSDG101       HC03        NaN            Rep3   \n",
       "\n",
       "   repeated_sample  batch tissue       brain_region  source_RIN  ...   \n",
       "0                0    NaN  Brain  Prefrontal Cortex         NaN  ...  \\\n",
       "1                0    NaN  Brain  Prefrontal Cortex         NaN  ...   \n",
       "2                0    NaN  Brain  Prefrontal Cortex         NaN  ...   \n",
       "3                0    NaN  Brain  Prefrontal Cortex         NaN  ...   \n",
       "4                0    NaN  Brain  Prefrontal Cortex         NaN  ...   \n",
       "\n",
       "   sex_ontology_term_id self_reported_ethnicity_ontology_term_id   \n",
       "0                   NaN                                      NaN  \\\n",
       "1                   NaN                                      NaN   \n",
       "2                   NaN                                      NaN   \n",
       "3                   NaN                                      NaN   \n",
       "4                   NaN                                      NaN   \n",
       "\n",
       "  disease_ontology_term_id tissue_ontology_term_id cell_type_ontology_term_id   \n",
       "0                      NaN                     NaN                        NaN  \\\n",
       "1                      NaN                     NaN                        NaN   \n",
       "2                      NaN                     NaN                        NaN   \n",
       "3                      NaN                     NaN                        NaN   \n",
       "4                      NaN                     NaN                        NaN   \n",
       "\n",
       "  assay_ontology_term_id suspension_type DV2000 pm_PH  donor_id  \n",
       "0                    NaN             NaN    NaN   NaN       NaN  \n",
       "1                    NaN             NaN    NaN   NaN       NaN  \n",
       "2                    NaN             NaN    NaN   NaN       NaN  \n",
       "3                    NaN             NaN    NaN   NaN       NaN  \n",
       "4                    NaN             NaN    NaN   NaN       NaN  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>region_level_1</th>\n",
       "      <th>region_level_2</th>\n",
       "      <th>region_level_3</th>\n",
       "      <th>AMPPD_id</th>\n",
       "      <th>family_history</th>\n",
       "      <th>...</th>\n",
       "      <th>sn_neuronal_loss</th>\n",
       "      <th>path_infarcs</th>\n",
       "      <th>path_nia_ri</th>\n",
       "      <th>path_nia_aa_a</th>\n",
       "      <th>path_nia_aa_b</th>\n",
       "      <th>path_nia_aa_c</th>\n",
       "      <th>TDP43</th>\n",
       "      <th>arteriolosclerosis_severity_scale</th>\n",
       "      <th>amyloid_angiopathy_severity_scale</th>\n",
       "      <th>path_ad_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HC01</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Prefrontal cortex</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HC02</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Prefrontal cortex</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HC03</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Prefrontal cortex</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC04</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Prefrontal cortex</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC05</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Prefrontal cortex</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id source_sample_id  time_from_baseline  GP2_id hemisphere   \n",
       "0      HC01           hSDG07                 NaN     NaN    Unknown  \\\n",
       "1      HC02           hSDG13                 NaN     NaN    Unknown   \n",
       "2      HC03          hSDG101                 NaN     NaN    Unknown   \n",
       "3      HC04           hSDG10                 NaN     NaN    Unknown   \n",
       "4      HC05           hSDG30                 NaN     NaN    Unknown   \n",
       "\n",
       "  region_level_1     region_level_2 region_level_3  AMPPD_id family_history   \n",
       "0   Frontal lobe  Prefrontal cortex    Grey matter       NaN        Unknown  \\\n",
       "1   Frontal lobe  Prefrontal cortex    Grey matter       NaN        Unknown   \n",
       "2   Frontal lobe  Prefrontal cortex    Grey matter       NaN        Unknown   \n",
       "3   Frontal lobe  Prefrontal cortex    Grey matter       NaN        Unknown   \n",
       "4   Frontal lobe  Prefrontal cortex    Grey matter       NaN        Unknown   \n",
       "\n",
       "   ...  sn_neuronal_loss  path_infarcs  path_nia_ri  path_nia_aa_a   \n",
       "0  ...               NaN           NaN          NaN            NaN  \\\n",
       "1  ...               NaN           NaN          NaN            NaN   \n",
       "2  ...               NaN           NaN          NaN            NaN   \n",
       "3  ...               NaN           NaN          NaN            NaN   \n",
       "4  ...               NaN           NaN          NaN            NaN   \n",
       "\n",
       "   path_nia_aa_b  path_nia_aa_c  TDP43  arteriolosclerosis_severity_scale   \n",
       "0            NaN            NaN    NaN                                NaN  \\\n",
       "1            NaN            NaN    NaN                                NaN   \n",
       "2            NaN            NaN    NaN                                NaN   \n",
       "3            NaN            NaN    NaN                                NaN   \n",
       "4            NaN            NaN    NaN                                NaN   \n",
       "\n",
       "   amyloid_angiopathy_severity_scale path_ad_level  \n",
       "0                                NaN           NaN  \n",
       "1                                NaN           NaN  \n",
       "2                                NaN           NaN  \n",
       "3                                NaN           NaN  \n",
       "4                                NaN           NaN  \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def add_hafler_batch(sample_df):\n",
    "\n",
    "    # First batch: HSDG07HC HSDG10HC HSDG148PD HSDG199PD\n",
    "    # batch[batch.sample_id in ['hSDG07HC', 'hSDG10HC', 'hSDG148PD', 'hSDG199PD']]=1\n",
    "    Batch_1 = ['hSDG07', 'hSDG10', 'hSDG148', 'hSDG199'] \n",
    "    # Second batch: hsDG101HC hsDG13HC hsDG151PD hsDG197PD hsDG30HC hsDG99HC\n",
    "    Batch_2 = ['hSDG101', 'hSDG13', 'hSDG151', 'hSDG197', 'hSDG30', 'hSDG99']\n",
    "    # Third batch: hsDG142PD hsDG208PD\n",
    "    Batch_3 = ['hSDG142', 'hSDG208'] \n",
    "\n",
    "\n",
    "    batch_col = []\n",
    "    for row in sample_df.sample_id:\n",
    "        if row in Batch_1:\n",
    "            batch_col.append(\"Batch_1\")\n",
    "        elif row in Batch_2:\n",
    "            batch_col.append(\"Batch_2\")\n",
    "        elif row in Batch_3:\n",
    "            batch_col.append(\"Batch_3\")\n",
    "        else:\n",
    "            print(\"ERROR >>>>>>>> not no batch info\")\n",
    "            batch_col.append(\"\")\n",
    "\n",
    "\n",
    "    sample_df['batch'] = batch_col\n",
    "    return sample_df\n",
    "\n",
    "SAMPLE = add_hafler_batch(SAMPLE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HC04</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HC04</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HC04</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>HC02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>HC02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>HC02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td>PD04</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_3</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td>PD04</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_3</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td>PD04</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_3</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td>PD05</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td>PD05</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td>PD05</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td>PD01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td>PD01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td>PD01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td>PD02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td>PD02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td>PD02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td>PD06</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td>PD06</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td>PD06</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td>PD03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_3</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td>PD03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_3</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td>PD03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_3</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>HC05</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>HC05</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>HC05</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td>HC06</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td>HC06</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td>HC06</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id source_sample_id subject_id replicate  replicate_count   \n",
       "0     hSDG07           hSDG07       HC01      Rep1                1  \\\n",
       "1     hSDG07           hSDG07       HC01      Rep1                1   \n",
       "2     hSDG07           hSDG07       HC01      Rep1                1   \n",
       "3    hSDG101          hSDG101       HC03      Rep1                1   \n",
       "4    hSDG101          hSDG101       HC03      Rep1                1   \n",
       "5    hSDG101          hSDG101       HC03      Rep1                1   \n",
       "6     hSDG10           hSDG10       HC04      Rep1                1   \n",
       "7     hSDG10           hSDG10       HC04      Rep1                1   \n",
       "8     hSDG10           hSDG10       HC04      Rep1                1   \n",
       "9     hSDG13           hSDG13       HC02      Rep1                1   \n",
       "10    hSDG13           hSDG13       HC02      Rep1                1   \n",
       "11    hSDG13           hSDG13       HC02      Rep1                1   \n",
       "12   hSDG142          hSDG142       PD04      Rep1                1   \n",
       "13   hSDG142          hSDG142       PD04      Rep1                1   \n",
       "14   hSDG142          hSDG142       PD04      Rep1                1   \n",
       "15   hSDG148          hSDG148       PD05      Rep1                1   \n",
       "16   hSDG148          hSDG148       PD05      Rep1                1   \n",
       "17   hSDG148          hSDG148       PD05      Rep1                1   \n",
       "18   hSDG151          hSDG151       PD01      Rep1                1   \n",
       "19   hSDG151          hSDG151       PD01      Rep1                1   \n",
       "20   hSDG151          hSDG151       PD01      Rep1                1   \n",
       "21   hSDG197          hSDG197       PD02      Rep1                1   \n",
       "22   hSDG197          hSDG197       PD02      Rep1                1   \n",
       "23   hSDG197          hSDG197       PD02      Rep1                1   \n",
       "24   hSDG199          hSDG199       PD06      Rep1                1   \n",
       "25   hSDG199          hSDG199       PD06      Rep1                1   \n",
       "26   hSDG199          hSDG199       PD06      Rep1                1   \n",
       "27   hSDG208          hSDG208       PD03      Rep1                1   \n",
       "28   hSDG208          hSDG208       PD03      Rep1                1   \n",
       "29   hSDG208          hSDG208       PD03      Rep1                1   \n",
       "30    hSDG30           hSDG30       HC05      Rep1                1   \n",
       "31    hSDG30           hSDG30       HC05      Rep1                1   \n",
       "32    hSDG30           hSDG30       HC05      Rep1                1   \n",
       "33    hSDG99           hSDG99       HC06      Rep1                1   \n",
       "34    hSDG99           hSDG99       HC06      Rep1                1   \n",
       "35    hSDG99           hSDG99       HC06      Rep1                1   \n",
       "\n",
       "    repeated_sample    batch tissue       brain_region  source_RIN  ...   \n",
       "0                 0  Batch_1  Brain  Prefrontal Cortex         NaN  ...  \\\n",
       "1                 0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "2                 0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "3                 0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "4                 0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "5                 0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "6                 0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "7                 0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "8                 0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "9                 0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "10                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "11                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "12                0  Batch_3  Brain  Prefrontal Cortex         NaN  ...   \n",
       "13                0  Batch_3  Brain  Prefrontal Cortex         NaN  ...   \n",
       "14                0  Batch_3  Brain  Prefrontal Cortex         NaN  ...   \n",
       "15                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "16                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "17                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "18                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "19                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "20                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "21                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "22                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "23                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "24                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "25                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "26                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "27                0  Batch_3  Brain  Prefrontal Cortex         NaN  ...   \n",
       "28                0  Batch_3  Brain  Prefrontal Cortex         NaN  ...   \n",
       "29                0  Batch_3  Brain  Prefrontal Cortex         NaN  ...   \n",
       "30                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "31                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "32                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "33                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "34                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "35                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "\n",
       "    sex_ontology_term_id self_reported_ethnicity_ontology_term_id   \n",
       "0                    NaN                                      NaN  \\\n",
       "1                    NaN                                      NaN   \n",
       "2                    NaN                                      NaN   \n",
       "3                    NaN                                      NaN   \n",
       "4                    NaN                                      NaN   \n",
       "5                    NaN                                      NaN   \n",
       "6                    NaN                                      NaN   \n",
       "7                    NaN                                      NaN   \n",
       "8                    NaN                                      NaN   \n",
       "9                    NaN                                      NaN   \n",
       "10                   NaN                                      NaN   \n",
       "11                   NaN                                      NaN   \n",
       "12                   NaN                                      NaN   \n",
       "13                   NaN                                      NaN   \n",
       "14                   NaN                                      NaN   \n",
       "15                   NaN                                      NaN   \n",
       "16                   NaN                                      NaN   \n",
       "17                   NaN                                      NaN   \n",
       "18                   NaN                                      NaN   \n",
       "19                   NaN                                      NaN   \n",
       "20                   NaN                                      NaN   \n",
       "21                   NaN                                      NaN   \n",
       "22                   NaN                                      NaN   \n",
       "23                   NaN                                      NaN   \n",
       "24                   NaN                                      NaN   \n",
       "25                   NaN                                      NaN   \n",
       "26                   NaN                                      NaN   \n",
       "27                   NaN                                      NaN   \n",
       "28                   NaN                                      NaN   \n",
       "29                   NaN                                      NaN   \n",
       "30                   NaN                                      NaN   \n",
       "31                   NaN                                      NaN   \n",
       "32                   NaN                                      NaN   \n",
       "33                   NaN                                      NaN   \n",
       "34                   NaN                                      NaN   \n",
       "35                   NaN                                      NaN   \n",
       "\n",
       "   disease_ontology_term_id tissue_ontology_term_id   \n",
       "0                       NaN                     NaN  \\\n",
       "1                       NaN                     NaN   \n",
       "2                       NaN                     NaN   \n",
       "3                       NaN                     NaN   \n",
       "4                       NaN                     NaN   \n",
       "5                       NaN                     NaN   \n",
       "6                       NaN                     NaN   \n",
       "7                       NaN                     NaN   \n",
       "8                       NaN                     NaN   \n",
       "9                       NaN                     NaN   \n",
       "10                      NaN                     NaN   \n",
       "11                      NaN                     NaN   \n",
       "12                      NaN                     NaN   \n",
       "13                      NaN                     NaN   \n",
       "14                      NaN                     NaN   \n",
       "15                      NaN                     NaN   \n",
       "16                      NaN                     NaN   \n",
       "17                      NaN                     NaN   \n",
       "18                      NaN                     NaN   \n",
       "19                      NaN                     NaN   \n",
       "20                      NaN                     NaN   \n",
       "21                      NaN                     NaN   \n",
       "22                      NaN                     NaN   \n",
       "23                      NaN                     NaN   \n",
       "24                      NaN                     NaN   \n",
       "25                      NaN                     NaN   \n",
       "26                      NaN                     NaN   \n",
       "27                      NaN                     NaN   \n",
       "28                      NaN                     NaN   \n",
       "29                      NaN                     NaN   \n",
       "30                      NaN                     NaN   \n",
       "31                      NaN                     NaN   \n",
       "32                      NaN                     NaN   \n",
       "33                      NaN                     NaN   \n",
       "34                      NaN                     NaN   \n",
       "35                      NaN                     NaN   \n",
       "\n",
       "   cell_type_ontology_term_id assay_ontology_term_id suspension_type DV2000   \n",
       "0                         NaN                    NaN             NaN    NaN  \\\n",
       "1                         NaN                    NaN             NaN    NaN   \n",
       "2                         NaN                    NaN             NaN    NaN   \n",
       "3                         NaN                    NaN             NaN    NaN   \n",
       "4                         NaN                    NaN             NaN    NaN   \n",
       "5                         NaN                    NaN             NaN    NaN   \n",
       "6                         NaN                    NaN             NaN    NaN   \n",
       "7                         NaN                    NaN             NaN    NaN   \n",
       "8                         NaN                    NaN             NaN    NaN   \n",
       "9                         NaN                    NaN             NaN    NaN   \n",
       "10                        NaN                    NaN             NaN    NaN   \n",
       "11                        NaN                    NaN             NaN    NaN   \n",
       "12                        NaN                    NaN             NaN    NaN   \n",
       "13                        NaN                    NaN             NaN    NaN   \n",
       "14                        NaN                    NaN             NaN    NaN   \n",
       "15                        NaN                    NaN             NaN    NaN   \n",
       "16                        NaN                    NaN             NaN    NaN   \n",
       "17                        NaN                    NaN             NaN    NaN   \n",
       "18                        NaN                    NaN             NaN    NaN   \n",
       "19                        NaN                    NaN             NaN    NaN   \n",
       "20                        NaN                    NaN             NaN    NaN   \n",
       "21                        NaN                    NaN             NaN    NaN   \n",
       "22                        NaN                    NaN             NaN    NaN   \n",
       "23                        NaN                    NaN             NaN    NaN   \n",
       "24                        NaN                    NaN             NaN    NaN   \n",
       "25                        NaN                    NaN             NaN    NaN   \n",
       "26                        NaN                    NaN             NaN    NaN   \n",
       "27                        NaN                    NaN             NaN    NaN   \n",
       "28                        NaN                    NaN             NaN    NaN   \n",
       "29                        NaN                    NaN             NaN    NaN   \n",
       "30                        NaN                    NaN             NaN    NaN   \n",
       "31                        NaN                    NaN             NaN    NaN   \n",
       "32                        NaN                    NaN             NaN    NaN   \n",
       "33                        NaN                    NaN             NaN    NaN   \n",
       "34                        NaN                    NaN             NaN    NaN   \n",
       "35                        NaN                    NaN             NaN    NaN   \n",
       "\n",
       "   pm_PH  donor_id  \n",
       "0    NaN       NaN  \n",
       "1    NaN       NaN  \n",
       "2    NaN       NaN  \n",
       "3    NaN       NaN  \n",
       "4    NaN       NaN  \n",
       "5    NaN       NaN  \n",
       "6    NaN       NaN  \n",
       "7    NaN       NaN  \n",
       "8    NaN       NaN  \n",
       "9    NaN       NaN  \n",
       "10   NaN       NaN  \n",
       "11   NaN       NaN  \n",
       "12   NaN       NaN  \n",
       "13   NaN       NaN  \n",
       "14   NaN       NaN  \n",
       "15   NaN       NaN  \n",
       "16   NaN       NaN  \n",
       "17   NaN       NaN  \n",
       "18   NaN       NaN  \n",
       "19   NaN       NaN  \n",
       "20   NaN       NaN  \n",
       "21   NaN       NaN  \n",
       "22   NaN       NaN  \n",
       "23   NaN       NaN  \n",
       "24   NaN       NaN  \n",
       "25   NaN       NaN  \n",
       "26   NaN       NaN  \n",
       "27   NaN       NaN  \n",
       "28   NaN       NaN  \n",
       "29   NaN       NaN  \n",
       "30   NaN       NaN  \n",
       "31   NaN       NaN  \n",
       "32   NaN       NaN  \n",
       "33   NaN       NaN  \n",
       "34   NaN       NaN  \n",
       "35   NaN       NaN  \n",
       "\n",
       "[36 rows x 42 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# fix replicate & replicate_count\n",
    "\n",
    "SAMPLE['replicate'] = \"Rep1\"\n",
    "SAMPLE['replicate_count'] = 1\n",
    "\n",
    "SAMPLE.head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_id', 'source_sample_id', 'subject_id', 'replicate',\n",
       "       'replicate_count', 'repeated_sample', 'batch', 'tissue', 'brain_region',\n",
       "       'source_RIN', 'RIN', 'molecular_source', 'input_cell_count', 'assay',\n",
       "       'sequencing_end', 'sequencing_length', 'sequencing_instrument',\n",
       "       'file_type', 'file_name', 'file_description', 'file_MD5', 'technology',\n",
       "       'omic', 'adjustment', 'content', 'time', 'header', 'annotation',\n",
       "       'preprocessing_references', 'configuration_file',\n",
       "       'organism_ontology_term_id', 'development_stage_ontology_term_id',\n",
       "       'sex_ontology_term_id', 'self_reported_ethnicity_ontology_term_id',\n",
       "       'disease_ontology_term_id', 'tissue_ontology_term_id',\n",
       "       'cell_type_ontology_term_id', 'assay_ontology_term_id',\n",
       "       'suspension_type', 'DV2000', 'pm_PH', 'donor_id'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "STUDY.to_csv( metadata_path / \"STUDY.csv\")\n",
    "CLINPATH.to_csv( metadata_path / \"CLINPATH.csv\")\n",
    "SUBJECT.to_csv( metadata_path / \"SUBJECT.csv\")\n",
    "SAMPLE.to_csv( metadata_path / \"SAMPLE.csv\")\n",
    "PROTOCOL.to_csv( metadata_path / \"PROTOCOL.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/team-hafler/fastq')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fastq_path = data_path / \"team-hafler/fastq\"\n",
    "fastq_path.glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/team-lee/ASAP CDE ALL_Team Hafler_v1.xlsx')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data_path / \"ASAP CDE ALL_Team Hafler_v1.xlsx\"\n",
    "\n",
    "#SUBJECT = pd.read_excel(data_path / \"ASAP CDE ALL_Team Hafler_v1.xlsx\",sheet_name=\"SUBJECT\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "CLINPATH = pd.read_excel(data_path / \"ASAP CDE ALL_Team Hafler_v1.xlsx\",sheet_name=\"CLINPATH\")\n",
    "SAMPLE = pd.read_excel(data_path / \"ASAP CDE ALL_Team Hafler_v1.xlsx\",sheet_name=\"SAMPLE\")\n",
    "PROTOCOL = pd.read_excel(data_path / \"ASAP CDE ALL_Team Hafler_v1.xlsx\",sheet_name=\"PROTOCOL\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.tsv\",delimiter=\"\\t\")\n",
    "\n",
    "SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.tsv\",delimiter=\"\\t\")\n",
    "CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\",delimiter=\",\")\n",
    "STUDY = pd.read_csv(f\"{metadata_path}/STUDY.tsv\",delimiter=\"\\t\")\n",
    "PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.tsv\",delimiter=\"\\t\")\n",
    "\n",
    "# read count matriz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ALIGNMENT\n",
    "\n",
    "> SOURCE: [gs://asap-raw-data-team-lee/scripts/PD_MFG_snRNAseq_with_redos.Rmd](gs://asap-raw-data-team-lee/scripts/PD_MFG_snRNAseq_with_redos.Rmd)\n",
    "\n",
    "\n",
    "## Cellranger\n",
    "\n",
    "Transfer sequencing data from genomics to rawdata  \n",
    "cellranger-6.0.1 to align and count  \n",
    "performed mRNA alignment for standard single cell counts  \n",
    "also performed mRNA and premRNA (introns) counts for trajectory analysis\n",
    "\n",
    "Need to run cellranger count for each sample.. i.e. \n",
    "\n",
    "HIP/HIP_HC_2067,HIP/HIP_PD_0413,HIP/HIP_PD_1504,HIP/HIP_HC_2062,HIP/HIP_PD_2038...\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/ergonyc/Projects/ASAP/harmonized-wf-dev/examples\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#---------- cellranger\n",
    "#----- mRNA alignment and counting\n",
    "# used for single cell\n",
    "# don't use --include-introns\n",
    "\n",
    "#---------- cellranger_mRNA_premRNA\n",
    "#----- mRNA and premRNA alignment and counting\n",
    "# used for single nuclei\n",
    "module load cellranger/cellranger-6.0.1\n",
    "\n",
    "\n",
    "\"cellranger count  \\\n",
    "  --id=TEAM_LEE \\\n",
    "  --transcriptome=refdata-gex-GRCh38-2020-A \\\n",
    "  --fastqs=../data/team-lee/fastq/ \n",
    "  --sample=SAMPLE \n",
    "  --localcores=4\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PRE-PROCESSING\n",
    "\n",
    "- convert to seurat Object\n",
    "- qc features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert to seurat Object\n",
    "data_path = Path.home() / (\"Projects/ASAP/team-lee\")\n",
    "metadata_path = data_path / \"metadata\"\n",
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.tsv\",delimiter=\"\\t\")\n",
    "\n",
    "SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.tsv\",delimiter=\"\\t\")\n",
    "CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\",delimiter=\",\")\n",
    "STUDY = pd.read_csv(f\"{metadata_path}/STUDY.tsv\",delimiter=\"\\t\")\n",
    "PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.tsv\",delimiter=\"\\t\")\n",
    "\n",
    "# read count matriz\n",
    "adata_path = data_path / \"cellranger_counts/SN\"\n",
    "\n",
    "# SN_2061_HC_count_filtered_feature_bc_matrix.h5\n",
    "# SN_2061_HC_count_molecule_info.h5\n",
    "# SN_2061_HC_count_raw_feature_bc_matrix.h5\n",
    "\n",
    "\n",
    "raw_counts = adata_path / \"SN_2061_HC_count_raw_feature_bc_matrix.h5\"\n",
    "filtered_counts = adata_path / \"SN_2061_HC_count_filtered_feature_bc_matrix.h5\"\n",
    "molec_info_counts = adata_path / \"SN_2061_HC_count_molecule_info.h5\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ergonyc/miniforge3/envs/scverse10/lib/python3.10/site-packages/anndata/_core/anndata.py:1832: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/Users/ergonyc/miniforge3/envs/scverse10/lib/python3.10/site-packages/anndata/_core/anndata.py:1832: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    }
   ],
   "source": [
    "raw_adata = sc.read_10x_h5(raw_counts, genome='GRCh38')\n",
    "filt_adata = sc.read_10x_h5(filtered_counts, genome='GRCh38')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 514945 × 36601\n",
       "    var: 'gene_ids', 'feature_types', 'genome'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 514945 × 36601\n",
       "    var: 'gene_ids', 'feature_types', 'genome'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_adata.var_names_make_unique()\n",
    "raw_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACCAT-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACCTA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAACGCC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACAACT-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACAATC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACTCGG-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAACTGTA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGAAGC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGCCCA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGACA-1</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrameView\n",
       "Columns: []\n",
       "Index: [AAACCTGAGAAACCAT-1, AAACCTGAGAAACCTA-1, AAACCTGAGAAACGCC-1, AAACCTGAGAACAACT-1, AAACCTGAGAACAATC-1, AAACCTGAGAACTCGG-1, AAACCTGAGAACTGTA-1, AAACCTGAGAAGAAGC-1, AAACCTGAGAAGCCCA-1, AAACCTGAGAAGGACA-1]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_adata[:10,:10].obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGCCT-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGACCACGA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGACCCACC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAGGGATA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCCAGAAC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTGTGGCTC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCATGCTCC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCCTTTCGG-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTGTCAAG-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTGTTTGT-1</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6004 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [AAACCTGAGAAGGCCT-1, AAACCTGAGACCACGA-1, AAACCTGAGACCCACC-1, AAACCTGAGAGGGATA-1, AAACCTGAGCCAGAAC-1, AAACCTGAGCCGGTAA-1, AAACCTGAGCTGAAAT-1, AAACCTGAGGAGTTGC-1, AAACCTGAGTGAAGTT-1, AAACCTGAGTGTACCT-1, AAACCTGCATATGCTG-1, AAACCTGGTAGGCTGA-1, AAACCTGGTCACACGC-1, AAACCTGGTCGAAAGC-1, AAACCTGTCTGCTGCT-1, AAACCTGTCTTACCGC-1, AAACGGGAGAATGTGT-1, AAACGGGAGACAGACC-1, AAACGGGAGACTGTAA-1, AAACGGGAGATCTGAA-1, AAACGGGAGCTGTTCA-1, AAACGGGAGGTGCTAG-1, AAACGGGAGTGAACGC-1, AAACGGGCAATAGCGG-1, AAACGGGCAATCCAAC-1, AAACGGGCATCTCGCT-1, AAACGGGGTAGCCTCG-1, AAACGGGTCAACGGGA-1, AAACGGGTCACTCTTA-1, AAACGGGTCAGCTCGG-1, AAACGGGTCAGTGCAT-1, AAACGGGTCTCCAGGG-1, AAAGATGAGGTAGCCA-1, AAAGATGAGGTGATTA-1, AAAGATGAGTTAACGA-1, AAAGATGCACGAAGCA-1, AAAGATGGTATTACCG-1, AAAGATGGTCGAGTTT-1, AAAGATGGTCTAGTGT-1, AAAGATGGTCTGCGGT-1, AAAGATGTCAGGATCT-1, AAAGATGTCATCGCTC-1, AAAGATGTCCTGCTTG-1, AAAGATGTCGTGGGAA-1, AAAGATGTCTCTAAGG-1, AAAGATGTCTCTGAGA-1, AAAGATGTCTTAGCCC-1, AAAGCAAAGAGACTAT-1, AAAGCAAAGATCTGCT-1, AAAGCAAAGCACAGGT-1, AAAGCAAAGCGATGAC-1, AAAGCAAAGGTCGGAT-1, AAAGCAAAGGTGACCA-1, AAAGCAACAATGAATG-1, AAAGCAACATTCCTCG-1, AAAGCAAGTATGAATG-1, AAAGCAAGTCTGCGGT-1, AAAGCAAGTGAGTATA-1, AAAGCAATCAGATAAG-1, AAAGCAATCCAGTAGT-1, AAAGCAATCCGTCATC-1, AAAGCAATCTCCAACC-1, AAAGTAGAGGCACATG-1, AAAGTAGCAAACTGTC-1, AAAGTAGGTAAGTGGC-1, AAAGTAGGTCTCACCT-1, AAAGTAGGTGTCGCTG-1, AAAGTAGGTGTGGTTT-1, AAAGTAGGTGTGTGCC-1, AAAGTAGGTTGAGTTC-1, AAAGTAGTCACAATGC-1, AAAGTAGTCATCGCTC-1, AAAGTAGTCCGTCATC-1, AAAGTAGTCTGCGTAA-1, AAATGCCAGAAGAAGC-1, AAATGCCAGGTTCCTA-1, AAATGCCAGTGAAGAG-1, AAATGCCAGTGGAGAA-1, AAATGCCCACATGACT-1, AAATGCCCACCACGTG-1, AAATGCCCACGAAAGC-1, AAATGCCCATCGGAAG-1, AAATGCCCATCGGACC-1, AAATGCCGTACACCGC-1, AAATGCCGTCTGATCA-1, AAATGCCGTTGCGCAC-1, AAATGCCTCACCGGGT-1, AAATGCCTCCATTCTA-1, AAATGCCTCTAGAGTC-1, AAATGCCTCTCCTATA-1, AACACGTAGAGAACAG-1, AACACGTAGCTGTTCA-1, AACACGTCAAAGGCGT-1, AACACGTCACATAACC-1, AACACGTCACTCGACG-1, AACACGTCATCACGTA-1, AACACGTGTAAGGGAA-1, AACACGTGTGTCGCTG-1, AACACGTTCCGCAAGC-1, AACACGTTCTACCTGC-1, ...]\n",
       "\n",
       "[6004 rows x 0 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filt_adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'AnnData' object has no attribute 'gene_ids'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[25], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m filt_adata\u001b[39m.\u001b[39;49mgene_ids\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'AnnData' object has no attribute 'gene_ids'"
     ]
    }
   ],
   "source": [
    "filt_adata.gene_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'scanpy' has no attribute 'read_seurat'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[26], line 8\u001b[0m\n\u001b[1;32m      5\u001b[0m filtered \u001b[39m=\u001b[39m seurat_path \u001b[39m/\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mseurat_filtered.rds\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m      6\u001b[0m combined \u001b[39m=\u001b[39m seurat_path \u001b[39m/\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mseurat_combined.rds\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m----> 8\u001b[0m filt \u001b[39m=\u001b[39m sc\u001b[39m.\u001b[39;49mread_seurat(filtered_original)\n",
      "\u001b[0;31mAttributeError\u001b[0m: module 'scanpy' has no attribute 'read_seurat'"
     ]
    }
   ],
   "source": [
    "seurat_path = data_path / \"seurat_objects/SN\"\n",
    "\n",
    "\n",
    "filtered_original = seurat_path / \"seurat_filtered_original.rds\"\n",
    "filtered = seurat_path / \"seurat_filtered.rds\"\n",
    "combined = seurat_path / \"seurat_combined.rds\"\n",
    "\n",
    "filt = sc.read_seurat(filtered_original)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scverse10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
