{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d6b5a7bd",
   "metadata": {},
   "source": [
    "ASAP CRN Prototype workflow \n",
    "\n",
    "# ASAP CRN Prototype workflow \n",
    "\n",
    "Skeleton workflow to begin development of a unified workflow for _all_ ASAP CRN single cell (or nucleus) RNA sequence data.\n",
    "based on:\n",
    "* PD_MFG_snRNAseq_with_redos.Rmd from Team Lee ***sequence alignment AND processing***\n",
    "* Harmony-RNA-Workflow [snakefile repo](https://github.com/shahrozeabbas/Harmony-RNA-Workflow) ***processing only***\n",
    "  \n",
    "28 July 2023\n",
    "Andy Henrie\n",
    "\n",
    "\n",
    "Note that the \n",
    "\n",
    "\n",
    "# TOOLS / PACKAGES\n",
    "\n",
    "\n",
    "# STEPS\n",
    "\n",
    "## housekeeping\n",
    "- define names\n",
    "- list of fastq\n",
    "- collect meta-data\n",
    "\n",
    "\n",
    "## alignment \n",
    "- cellranger\n",
    "\n",
    "## preprocessing\n",
    "- convert to data object (e.g. Scanpy, Surat)\n",
    "- add metadata\n",
    "    - from metadata\n",
    "    - calculated\n",
    "        - doublets\n",
    "        - mt\n",
    "        - rb\n",
    "\n",
    "## QC\n",
    "- filtering\n",
    "    - mt\n",
    "    - doublets\n",
    "    - counts\n",
    "    - features\n",
    "\n",
    "## processing\n",
    "- normalize\n",
    "- identify highly variable genes\n",
    "- scale\n",
    "- assign\n",
    "    - cell cycle scores\n",
    "\n",
    "## harmonize (LATER)\n",
    "- concatenate\n",
    "- RunHarmony\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9b2c037",
   "metadata": {},
   "source": [
    "## HOUSEKEEPING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b457ea4c",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Project"
   },
   "outputs": [],
   "source": [
    "project = \"asap_crn_XXX_snRNAseq\"\n",
    "\n",
    "input_table = \"input_table.csv\"\n",
    "\n",
    "\n",
    "## import meta-data \n",
    "\n",
    "\n",
    "## merge tables\n",
    "\n",
    "\n",
    "\n",
    "#---------- QC Filtering\n",
    "#----- Filtering\n",
    "\n",
    "            # nCount_RNA %between% c(500, 2e5) &\n",
    "            # nFeature_RNA %between% c(500, 12500) &\n",
    "            # percent.mt < 0.5 & percent.rb < 0.5,\n",
    "\n",
    "nUMI_low = 1000    # harmony = 500\n",
    "nUMI_high = 20000  # harmony = 20000\n",
    "nGene_low = 700    # harmony = 500\n",
    "nGene_high = 6000  # harmony = 12500\n",
    "percent_mito_high = 5  # \n",
    "percent_ribo_high = 30\n",
    "log10GenesPerUMI_low = 0.85\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "902eb63b",
   "metadata": {},
   "source": [
    "## ALIGNMENT\n",
    "\n",
    "> SOURCE: [gs://asap-raw-data-team-lee/scripts/PD_MFG_snRNAseq_with_redos.Rmd](gs://asap-raw-data-team-lee/scripts/PD_MFG_snRNAseq_with_redos.Rmd)\n",
    "\n",
    "\n",
    "### Cellranger\n",
    "\n",
    "Transfer sequencing data from genomics to rawdata  \n",
    "cellranger-6.0.1 to align and count  \n",
    "performed mRNA alignment for standard single cell counts  \n",
    "also performed mRNA and premRNA (introns) counts for trajectory analysis  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "353d4061",
   "metadata": {
    "eval": false,
    "name": "cellranger"
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "#---------- cellranger\n",
    "#----- mRNA alignment and counting\n",
    "# used for single cell\n",
    "# don't use --include-introns\n",
    "\n",
    "#---------- cellranger_mRNA_premRNA\n",
    "#----- mRNA and premRNA alignment and counting\n",
    "# used for single nuclei\n",
    "module load cellranger/cellranger-6.0.1\n",
    "\n",
    "\n",
    "\"cellranger count  \n",
    "  --include-introns \n",
    "  --id=SAMPLE_count \n",
    "  --transcriptome=/secondary/projects/bras/primary/software/cellranger_v6/refdata-gex-GRCh38-2020-A \n",
    "  --fastqs=../fastq \n",
    "  --sample=SAMPLE \n",
    "  --localcores=4\"\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7bf12944",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "22595000",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "data_path = \"../data/team-lee/metadata\"\n",
    "HIP_covar = pd.read_csv(f\"{data_path}/HIP/covar.csv\")\n",
    "HIP_cases = pd.read_csv(f\"{data_path}/HIP/PD_ASAP_Sample_batch_information_banner_cases.csv\")\n",
    "HIP_control = pd.read_csv(f\"{data_path}/HIP/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n",
    "\n",
    "MFG_covar = pd.read_csv(f\"{data_path}/MFG/covar.csv\")\n",
    "MFG_cases = pd.read_csv(f\"{data_path}/MFG/PD_ASAP_Sample_batch_information_banner_cases.csv\")\n",
    "MFG_control = pd.read_csv(f\"{data_path}/MFG/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n",
    "\n",
    "\n",
    "SN_covar = pd.read_csv(f\"{data_path}/SN/covar.csv\")\n",
    "SN_cases = pd.read_csv(f\"{data_path}/SN/PD_ASAP_Sample_batch_information_banner_cases.csv\")\n",
    "SN_control = pd.read_csv(f\"{data_path}/SN/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e1e13a93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BRAIN_REGION</th>\n",
       "      <th>SAMPLE</th>\n",
       "      <th>COUNT_ID</th>\n",
       "      <th>SAMPLE_FULL</th>\n",
       "      <th>ID</th>\n",
       "      <th>SEQ_ID</th>\n",
       "      <th>BATCH</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>DIAGNOSIS</th>\n",
       "      <th>SEX</th>\n",
       "      <th>AGE</th>\n",
       "      <th>DIAGNOSIS_AGE</th>\n",
       "      <th>DIAGNOSIS_YEARS</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>ALZHEIMERS_BRAAK</th>\n",
       "      <th>DEMENTIA</th>\n",
       "      <th>DEMENTIA_AGE</th>\n",
       "      <th>DEMENTIA_YEARS</th>\n",
       "      <th>WEIGHT_USED</th>\n",
       "      <th>TOTAL_WEIGHT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>HC_2067</td>\n",
       "      <td>HIP_HC_2067</td>\n",
       "      <td>HIP_HC_20-67</td>\n",
       "      <td>HC_20-67</td>\n",
       "      <td>HIP2067HC</td>\n",
       "      <td>BATCH_8</td>\n",
       "      <td>HC</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>72</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HC</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>HC_1939</td>\n",
       "      <td>HIP_HC_1939</td>\n",
       "      <td>HIP_HC_19-39</td>\n",
       "      <td>HC_19-39</td>\n",
       "      <td>HIP1939HC</td>\n",
       "      <td>BATCH_8</td>\n",
       "      <td>HC</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>93</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HC</td>\n",
       "      <td>II</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>PD_0413</td>\n",
       "      <td>HIP_PD_0413</td>\n",
       "      <td>HIP_PD_04-13</td>\n",
       "      <td>PD_04-13</td>\n",
       "      <td>HIP0413PD</td>\n",
       "      <td>BATCH_8</td>\n",
       "      <td>PD</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>72</td>\n",
       "      <td>62.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>PD</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>PD_1504</td>\n",
       "      <td>HIP_PD_1504</td>\n",
       "      <td>HIP_PD_15-04</td>\n",
       "      <td>PD_15-04</td>\n",
       "      <td>HIP1504PD</td>\n",
       "      <td>BATCH_8</td>\n",
       "      <td>PD</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>72</td>\n",
       "      <td>54.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>PD</td>\n",
       "      <td>II</td>\n",
       "      <td>1</td>\n",
       "      <td>68.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>35</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Hippocampus</td>\n",
       "      <td>HC_1225</td>\n",
       "      <td>HIP_HC_1225</td>\n",
       "      <td>HIP_HC_12-25</td>\n",
       "      <td>HC_12-25</td>\n",
       "      <td>HIP1225HC</td>\n",
       "      <td>BATCH_9</td>\n",
       "      <td>HC</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HC</td>\n",
       "      <td>IV</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  BRAIN_REGION   SAMPLE     COUNT_ID   SAMPLE_FULL        ID     SEQ_ID   \n",
       "0  Hippocampus  HC_2067  HIP_HC_2067  HIP_HC_20-67  HC_20-67  HIP2067HC  \\\n",
       "1  Hippocampus  HC_1939  HIP_HC_1939  HIP_HC_19-39  HC_19-39  HIP1939HC   \n",
       "2  Hippocampus  PD_0413  HIP_PD_0413  HIP_PD_04-13  PD_04-13  HIP0413PD   \n",
       "3  Hippocampus  PD_1504  HIP_PD_1504  HIP_PD_15-04  PD_15-04  HIP1504PD   \n",
       "4  Hippocampus  HC_1225  HIP_HC_1225  HIP_HC_12-25  HC_12-25  HIP1225HC   \n",
       "\n",
       "     BATCH GROUP  DIAGNOSIS  SEX  AGE  DIAGNOSIS_AGE  DIAGNOSIS_YEARS DISEASE   \n",
       "0  BATCH_8    HC          0    1   72            NaN              NaN      HC  \\\n",
       "1  BATCH_8    HC          0    2   93            NaN              NaN      HC   \n",
       "2  BATCH_8    PD          1    1   72           62.0             10.0      PD   \n",
       "3  BATCH_8    PD          1    2   72           54.0             18.0      PD   \n",
       "4  BATCH_9    HC          0    1   80            NaN              NaN      HC   \n",
       "\n",
       "  ALZHEIMERS_BRAAK  DEMENTIA  DEMENTIA_AGE  DEMENTIA_YEARS  WEIGHT_USED   \n",
       "0                I         0           NaN             NaN           35  \\\n",
       "1               II         0           NaN             NaN           29   \n",
       "2                I         0           NaN             NaN           33   \n",
       "3               II         1          68.0             4.0           35   \n",
       "4               IV         0           NaN             NaN           19   \n",
       "\n",
       "   TOTAL_WEIGHT  \n",
       "0            53  \n",
       "1            49  \n",
       "2            49  \n",
       "3            62  \n",
       "4            19  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HIP_covar.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "602cf4f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CaseID</th>\n",
       "      <th>race</th>\n",
       "      <th>gender</th>\n",
       "      <th>expired_age</th>\n",
       "      <th>PMI</th>\n",
       "      <th>last_mmse_test_score</th>\n",
       "      <th>last_mmse_interval_before_death</th>\n",
       "      <th>motor_updrs_score_on</th>\n",
       "      <th>motor_updrs_score_off</th>\n",
       "      <th>motor_updrs_score_months_prior_to_death</th>\n",
       "      <th>...</th>\n",
       "      <th>nctx_frontal</th>\n",
       "      <th>sum_lb_density</th>\n",
       "      <th>nctx_parietal</th>\n",
       "      <th>PathDXSummary</th>\n",
       "      <th>neurological_dx</th>\n",
       "      <th>neurological_dx_years</th>\n",
       "      <th>neurological_dx_age</th>\n",
       "      <th>dementia_yn</th>\n",
       "      <th>dementia_years</th>\n",
       "      <th>demential_age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18-64</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>72</td>\n",
       "      <td>4.60</td>\n",
       "      <td>29.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Control (history); Neurofibrillary tangles, me...</td>\n",
       "      <td>Control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20-57</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>81</td>\n",
       "      <td>4.58</td>\n",
       "      <td>28.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Control; Neurofibrillary tangles and non-speci...</td>\n",
       "      <td>Control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20-61</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>96</td>\n",
       "      <td>2.83</td>\n",
       "      <td>27.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Cognitive impairment without dementia (history...</td>\n",
       "      <td>Mild Cognitive Impairment</td>\n",
       "      <td>2.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>20-62</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>81</td>\n",
       "      <td>4.42</td>\n",
       "      <td>28.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Cognitively normal control (history); Cerebral...</td>\n",
       "      <td>Control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20-67</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>72</td>\n",
       "      <td>5.25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>Control (history); Microscopic changes of Alzh...</td>\n",
       "      <td>Control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CaseID  race  gender  expired_age   PMI  last_mmse_test_score   \n",
       "0  18-64     1       1           72  4.60                  29.0  \\\n",
       "1  20-57     1       1           81  4.58                  28.0   \n",
       "2  20-61     1       2           96  2.83                  27.0   \n",
       "3  20-62     1       1           81  4.42                  28.0   \n",
       "4  20-67     1       1           72  5.25                   NaN   \n",
       "\n",
       "   last_mmse_interval_before_death  motor_updrs_score_on   \n",
       "0                             18.0                   NaN  \\\n",
       "1                              8.0                   NaN   \n",
       "2                             26.0                   NaN   \n",
       "3                             12.0                   NaN   \n",
       "4                              NaN                   NaN   \n",
       "\n",
       "   motor_updrs_score_off  motor_updrs_score_months_prior_to_death  ...   \n",
       "0                    NaN                                      NaN  ...  \\\n",
       "1                    4.0                                      8.0  ...   \n",
       "2                   12.0                                     26.0  ...   \n",
       "3                   14.0                                     12.0  ...   \n",
       "4                    NaN                                      NaN  ...   \n",
       "\n",
       "  nctx_frontal sum_lb_density nctx_parietal   \n",
       "0            0              0             0  \\\n",
       "1            0              0             0   \n",
       "2            0              0             0   \n",
       "3            0              0             0   \n",
       "4            0              0             0   \n",
       "\n",
       "                                       PathDXSummary   \n",
       "0  Control (history); Neurofibrillary tangles, me...  \\\n",
       "1  Control; Neurofibrillary tangles and non-speci...   \n",
       "2  Cognitive impairment without dementia (history...   \n",
       "3  Cognitively normal control (history); Cerebral...   \n",
       "4  Control (history); Microscopic changes of Alzh...   \n",
       "\n",
       "             neurological_dx neurological_dx_years neurological_dx_age   \n",
       "0                    Control                   NaN                 NaN  \\\n",
       "1                    Control                   NaN                 NaN   \n",
       "2  Mild Cognitive Impairment                   2.0                94.0   \n",
       "3                    Control                   NaN                 NaN   \n",
       "4                    Control                   NaN                 NaN   \n",
       "\n",
       "  dementia_yn dementia_years demential_age  \n",
       "0           0            NaN           NaN  \n",
       "1           0            NaN           NaN  \n",
       "2           0            NaN           NaN  \n",
       "3           0            NaN           NaN  \n",
       "4           0            NaN           NaN  \n",
       "\n",
       "[5 rows x 70 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HIP_control.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b718f537",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CaseID</th>\n",
       "      <th>race</th>\n",
       "      <th>gender</th>\n",
       "      <th>expired_age</th>\n",
       "      <th>PMI</th>\n",
       "      <th>last_mmse_test_score</th>\n",
       "      <th>last_mmse_interval_before_death</th>\n",
       "      <th>motor_updrs_score_on</th>\n",
       "      <th>motor_updrs_score_off</th>\n",
       "      <th>motor_updrs_score_months_prior_to_death</th>\n",
       "      <th>...</th>\n",
       "      <th>nctx_temporal</th>\n",
       "      <th>nctx_frontal</th>\n",
       "      <th>sum_lb_density</th>\n",
       "      <th>nctx_parietal</th>\n",
       "      <th>PathDXSummary</th>\n",
       "      <th>neurological_dx</th>\n",
       "      <th>neurological_dx_age</th>\n",
       "      <th>dementia_yn</th>\n",
       "      <th>dementia_years</th>\n",
       "      <th>demential_age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>00-09</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>4.00</td>\n",
       "      <td>22.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Parkinson's disease; Dementia (history); Charc...</td>\n",
       "      <td>Parkinson's disease</td>\n",
       "      <td>49.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>58.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>03-48</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>3.00</td>\n",
       "      <td>27.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Parkinson's disease; microscopic lesions of Al...</td>\n",
       "      <td>Parkinson's disease</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>04-13</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>2.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Parkinson’s disease; Astrocytoma, grade IV (Da...</td>\n",
       "      <td>Parkinson's disease</td>\n",
       "      <td>62.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>06-02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>2.66</td>\n",
       "      <td>26.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Control; Mild cognitive impairment (history); ...</td>\n",
       "      <td>Mild Cognitive Impairment</td>\n",
       "      <td>81.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12-25</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>3.50</td>\n",
       "      <td>29.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Control; Neurofibrillary tangles, argyrophilic...</td>\n",
       "      <td>Control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 69 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CaseID  race  gender  expired_age   PMI  last_mmse_test_score   \n",
       "0  00-09   1.0     1.0         64.0  4.00                  22.0  \\\n",
       "1  03-48   1.0     1.0         90.0  3.00                  27.0   \n",
       "2  04-13   1.0     1.0         72.0  2.00                   NaN   \n",
       "3  06-02   1.0     1.0         84.0  2.66                  26.0   \n",
       "4  12-25   1.0     1.0         80.0  3.50                  29.0   \n",
       "\n",
       "   last_mmse_interval_before_death  motor_updrs_score_on   \n",
       "0                             34.0                   NaN  \\\n",
       "1                             26.0                   NaN   \n",
       "2                              NaN                   NaN   \n",
       "3                             25.0                   NaN   \n",
       "4                              5.0                   NaN   \n",
       "\n",
       "   motor_updrs_score_off  motor_updrs_score_months_prior_to_death  ...   \n",
       "0                    NaN                                      NaN  ...  \\\n",
       "1                    NaN                                      NaN  ...   \n",
       "2                   17.5                                      1.0  ...   \n",
       "3                    7.0                                      2.0  ...   \n",
       "4                    0.0                                      5.0  ...   \n",
       "\n",
       "  nctx_temporal nctx_frontal sum_lb_density nctx_parietal   \n",
       "0           1.0          1.0           19.0           1.0  \\\n",
       "1           1.0          1.0           16.0           1.0   \n",
       "2           1.0          1.0           11.0           0.0   \n",
       "3           0.0          0.0           11.0           0.0   \n",
       "4           0.0          0.0            3.0           0.0   \n",
       "\n",
       "                                       PathDXSummary   \n",
       "0  Parkinson's disease; Dementia (history); Charc...  \\\n",
       "1  Parkinson's disease; microscopic lesions of Al...   \n",
       "2  Parkinson’s disease; Astrocytoma, grade IV (Da...   \n",
       "3  Control; Mild cognitive impairment (history); ...   \n",
       "4  Control; Neurofibrillary tangles, argyrophilic...   \n",
       "\n",
       "             neurological_dx neurological_dx_age dementia_yn dementia_years   \n",
       "0        Parkinson's disease                49.0         1.0            6.0  \\\n",
       "1        Parkinson's disease                75.0         0.0            NaN   \n",
       "2        Parkinson's disease                62.0         0.0            NaN   \n",
       "3  Mild Cognitive Impairment                81.0         0.0            NaN   \n",
       "4                    Control                 NaN         0.0            NaN   \n",
       "\n",
       "  demential_age  \n",
       "0          58.0  \n",
       "1           NaN  \n",
       "2           NaN  \n",
       "3           NaN  \n",
       "4           NaN  \n",
       "\n",
       "[5 rows x 69 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HIP_cases.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fbe00ad",
   "metadata": {},
   "source": [
    "# PREPROCESSING\n",
    "\n",
    "> SOURCE: [Harmony-RNA-Workflow/scripts/main/preprocess.R](https://github.com/shahrozeabbas/Harmony-RNA-Workflow/blob/main/scripts/main/preprocess.R)\n",
    "\n",
    "\n",
    "This step creates a suitable data object including the relavent metadata.\n",
    "\n",
    "1. convert raw cellranger counts to data objects.  e.g. Seurat object as per examples. Could be scanpy anndata object (preferred?)\n",
    "2. add metadata\n",
    "    - collect metadata - sample IDs, conditions, etc.\n",
    "    - calculate features/metrics - doublets, mt, rb\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5c9529d",
   "metadata": {
    "eval": false,
    "name": "Seurat_QC"
   },
   "outputs": [],
   "source": [
    "# TEAM LEE\n",
    "#---------- Seurat_Object\n",
    "#----- Count Matrix\n",
    "seurat_counts <- lapply(NamesCount, function(Name){\n",
    "  count <- Read10X(data.dir = paste0(DirCounts,\n",
    "                                     Name,\n",
    "                                     \"_count/outs/filtered_feature_bc_matrix/\"),\n",
    "                   gene.column=2, # col1 Ensemble, col2 Symbol\n",
    "                   unique.features=TRUE,\n",
    "                   strip.suffix=TRUE) \n",
    "  seurat_obj <- CreateSeuratObject(counts = count, \n",
    "                                   project = Name,\n",
    "                                   min.cells = 10,\n",
    "                                   min.genes = 200)\n",
    "})\n",
    "# Make sound to notify that evaluation is complete.\n",
    "beepr::beep()\n",
    "#----- Merge Seurat Objects\n",
    "seurat_combined <- merge(x = seurat_counts[[1]], \n",
    "                         y = seurat_counts[2:length(seurat_counts)],\n",
    "                         add.cell.ids = unlist(NamesCount), \n",
    "                         project = NameProject)\n",
    "\n",
    "#---------- Add Metadata\n",
    "#----- Add Percentage of Mitocondrial and Ribosomal Counts\n",
    "seurat_combined <- Add_Mito_Ribo_Seurat(seurat_object = seurat_combined, \n",
    "                                        species = \"Human\")\n",
    "# View(seurat_combined@meta.data)\n",
    "# orig.ident: contains the sample identity if known\n",
    "# nCount_RNA: number of UMIs per cell\n",
    "# nFeature_RNA: number of genes detected per cell\n",
    "seurat_meta <- function(seurat_obj, name_count) {\n",
    "  #----- Create metadata dataframe\n",
    "  metadata <- seurat_obj@meta.data\n",
    "  #----- Rename \n",
    "  names(metadata)[1:3] <- c(name_count,\"nUMI\",\"nGene\")\n",
    "  #----- Add Cell IDs\n",
    "  metadata$cells <- rownames(metadata)\n",
    "  #----- Add number of genes per UMI\n",
    "  metadata$log10GenesPerUMI <- \n",
    "    log10(metadata$nGene) / log10(metadata$nUMI)\n",
    "  #----- Compute percent mito ratio\n",
    "  # metadata <- cbind(metadata,\n",
    "  #                   PercentageFeatureSet(object = seurat_combined, \n",
    "  #                                        pattern = \"^MT-\") / 100)\n",
    "  # names(metadata)[6] <- \"mitoRatio\"\n",
    "  #----- Merge Covar\n",
    "  metadata <- metadata %>% \n",
    "                left_join(covar, by = name_count)\n",
    "  rownames(metadata) <- metadata$cells\n",
    "  #----- Add metadata back to Seurat object\n",
    "  seurat_obj@meta.data <- metadata\n",
    "  return(seurat_obj)\n",
    "}\n",
    "seurat_combined <- seurat_meta(seurat_obj = seurat_combined,\n",
    "                                name_count = NameCount)\n",
    "\n",
    "\n",
    "#---------- Senescence\n",
    "#----- Senescence Markers, p16 = CDKN2A, p21 = CDKN1A\n",
    "#----- + extra senescence-relevant genes of interest: p19 = CDKN2D, uPAR = PLAUR\n",
    "seurat_combined$CDKN2A_Count <- \n",
    "  seurat_combined@assays$RNA@counts[rownames(seurat_combined) == \"CDKN2A\",]\n",
    "seurat_combined$CDKN1A_Count <- \n",
    "  seurat_combined@assays$RNA@counts[rownames(seurat_combined) == \"CDKN1A\",]\n",
    "seurat_combined$CDKN2D_Count <- \n",
    "  seurat_combined@assays$RNA@counts[rownames(seurat_combined) == \"CDKN2D\",]\n",
    "seurat_combined$GPNMB_Count <- \n",
    "  seurat_combined@assays$RNA@counts[rownames(seurat_combined) == \"GPNMB\",]\n",
    "seurat_combined$PLAUR_Count <- \n",
    "  seurat_combined@assays$RNA@counts[rownames(seurat_combined) == \"PLAUR\",]\n",
    "\n",
    "#---------- Save Data\n",
    "#----- Remove Data\n",
    "rm(seurat_counts)\n",
    "#----- Save Data\n",
    "saveRDS(seurat_combined, file=\"seurat_results/seurat_combined.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12cdaf86",
   "metadata": {},
   "source": [
    "## QC \n",
    "\n",
    "> SOURCE: [Harmony-RNA-Workflow/scripts/main/gmm_doublet_calling.R](https://github.com/shahrozeabbas/Harmony-RNA-Workflow/blob/main/scripts/main/gmm_doublet_calling.R) and [Harmony-RNA-Workflow/scripts/main/filter.R](https://github.com/shahrozeabbas/Harmony-RNA-Workflow/blob/main/scripts/main/filter.R)\n",
    "\n",
    "\n",
    "Quality control thresholds calculated/set.   \n",
    "Probably we should simply Add QC features (columns) instead of removing entries at this point.\n",
    "\n",
    "We also can generate figure artifacts.\n",
    "\n",
    "e.g. Senescence Cell not selectively removed  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35631a81",
   "metadata": {
    "name": "Seurat_QC_Figures"
   },
   "outputs": [],
   "source": [
    "# TEAM LEE\n",
    "#---------- Seurat_QC_Figures\n",
    "#----- Load data\n",
    "seurat_combined <- readRDS(file = \"seurat_results/seurat_combined.rds\")\n",
    "\n",
    "#---------- Cell Counts\n",
    "#----- Number of Cells per Sample\n",
    "plot <- ggSample + \n",
    "  geom_bar(data = seurat_combined@meta.data, \n",
    "           aes(x=SAMPLE, fill=SAMPLE)) + \n",
    "  ggtitle(\"Number of Cells\") + \n",
    "  theme(axis.title.x = element_blank()) + \n",
    "  theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Number_of_Cells_sample.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Number of Cells per Group\n",
    "plot <- ggGroup + \n",
    "  geom_boxplot(data = seurat_combined@meta.data %>% \n",
    "                        dplyr::count(GROUP, SAMPLE) %>% \n",
    "                        dplyr::rename(Count = n), \n",
    "           aes(x=GROUP, y=Count, fill=GROUP)) + \n",
    "  ggtitle(\"Number of Cells\") + \n",
    "  theme(axis.title.x = element_blank()) + \n",
    "  theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Number_of_Cells_group.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "\n",
    "#---------- UMI Transcript Counts \n",
    "#----- Number of Transcripts per cell per Sample\n",
    "plot <- ggSample + \n",
    "  geom_density(data = seurat_combined@meta.data, \n",
    "               alpha = 0.2,\n",
    "               aes(x=nUMI, fill=SAMPLE, color=SAMPLE)) + \n",
    "  ggtitle(\"Number of Transcripts per Cell\") + \n",
    "  ylab(\"Cell density\") + \n",
    "  geom_vline(xintercept = nUMI_low, linetype=\"dashed\", colour = \"red3\") + \n",
    "  geom_vline(xintercept = nUMI_high, linetype=\"dashed\", colour = \"red3\") + \n",
    "  scale_x_log10() + \n",
    "  facet_wrap(~SAMPLE) +\n",
    "  theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Number_of_Transcripts_per_Cell_sample.png\", \n",
    "    width=1000, \n",
    "    height=1000)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Number of Transcripts per cell per Group\n",
    "plot <- ggGroup + \n",
    "  geom_density(data = seurat_combined@meta.data, \n",
    "               alpha = 0.2,\n",
    "               aes(x=nUMI, fill=GROUP, color=GROUP)) + \n",
    "  ggtitle(\"Number of Transcripts per Cell\") + \n",
    "  ylab(\"Cell density\") + \n",
    "  geom_vline(xintercept = nUMI_low, linetype=\"dashed\", colour = \"red3\") + \n",
    "  geom_vline(xintercept = nUMI_high, linetype=\"dashed\", colour = \"red3\") + \n",
    "  scale_x_log10()\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Number_of_Transcripts_per_Cell_group.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "\n",
    "#---------- Gene Counts\n",
    "#----- Number of Genes per cell per Sample\n",
    "plot <- ggSample + \n",
    "  geom_density(data = seurat_combined@meta.data, \n",
    "               alpha = 0.2,\n",
    "               aes(x=nGene, fill=SAMPLE, color=SAMPLE)) + \n",
    "  ggtitle(\"Number of Genes per Cell\") + \n",
    "  ylab(\"Cell density\") + \n",
    "  geom_vline(xintercept = nGene_low, linetype=\"dashed\", colour = \"red3\") + \n",
    "  geom_vline(xintercept = nGene_high, linetype=\"dashed\", colour = \"red3\") + \n",
    "  scale_x_log10() + \n",
    "  facet_wrap(~SAMPLE) + \n",
    "  theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Number_of_Genes_per_Cell_sample.png\", \n",
    "    width=1000, \n",
    "    height=1000)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Number of Genes per cell per Group\n",
    "plot <- ggGroup + \n",
    "  geom_density(data = seurat_combined@meta.data, \n",
    "               alpha = 0.2,\n",
    "               aes(x=nGene, fill=GROUP, color=GROUP)) + \n",
    "  ggtitle(\"Number of Genes per Cell\") + \n",
    "  ylab(\"Cell density\") + \n",
    "  geom_vline(xintercept = nGene_low, linetype=\"dashed\", colour = \"red3\") + \n",
    "  geom_vline(xintercept = nGene_high, linetype=\"dashed\", colour = \"red3\") + \n",
    "  scale_x_log10()\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Number_of_Genes_per_Cell_group.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "\n",
    "\n",
    "#----- Number of Genes per cell per Sample\n",
    "plot <- ggSample + \n",
    "  geom_boxplot(data = seurat_combined@meta.data,\n",
    "               aes(x=SAMPLE, y=log10(nGene), fill=SAMPLE)) + \n",
    "  ggtitle(\"Number of Genes per Cell\") + \n",
    "  theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Number_of_Genes_per_Cell_per_Sample.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Number of Genes per cell per Group\n",
    "plot <- ggGroup + \n",
    "  geom_boxplot(data = seurat_combined@meta.data,\n",
    "               aes(x=GROUP, y=log10(nGene), fill=GROUP)) + \n",
    "  ggtitle(\"Number of Genes per Cell\") + \n",
    "  theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Number_of_Genes_per_Cell_per_Group.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "\n",
    "#---------- Mitochondria Percentage\n",
    "#----- Mitochondrial Transcripts per cell per Sample\n",
    "lowest_mito <- min(seurat_combined@meta.data$percent_mito[seurat_combined@meta.data$percent_mito > 0])\n",
    "plot <- ggSample + \n",
    "  geom_density(data = seurat_combined@meta.data %>% \n",
    "                 # handle the log 0 -- let these show up on plot \n",
    "                 # (replace 0s with the min non-zero value)\n",
    "                 mutate(percent_mito = case_when(percent_mito == 0 ~ lowest_mito,\n",
    "                                                 TRUE ~ percent_mito)), \n",
    "               alpha = 0.2,\n",
    "               aes(x=percent_mito, fill=SAMPLE, color=SAMPLE)) + \n",
    "  ggtitle(\"Mitochondrial Transcripts per Cell\") + \n",
    "  ylab(\"Percentage\") + \n",
    "  geom_vline(xintercept = percent_mito_high, linetype=\"dashed\", colour = \"red3\") + \n",
    "  scale_x_log10() + \n",
    "  facet_wrap(~SAMPLE) + \n",
    "  theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Mitochondrial_Transcripts_per_Cell_sample.png\", \n",
    "    width=1000, \n",
    "    height=1000)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Mitochondrial Transcripts per cell per Group\n",
    "plot <- ggGroup + \n",
    "  geom_density(data = seurat_combined@meta.data %>% \n",
    "                 # handle the log 0 -- let these show up on plot \n",
    "                 # (replace 0s with the min non-zero value)\n",
    "                 mutate(percent_mito = case_when(percent_mito == 0 ~ lowest_mito,\n",
    "                                                 TRUE ~ percent_mito)), \n",
    "               alpha = 0.2,\n",
    "               aes(x=percent_mito, fill=GROUP, color=GROUP)) + \n",
    "  ggtitle(\"Mitochondrial Transcripts per Cell\") + \n",
    "  ylab(\"Percentage\") + \n",
    "  geom_vline(xintercept = percent_mito_high, linetype=\"dashed\", colour = \"red3\") + \n",
    "  scale_x_log10()\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Mitochondrial_Transcripts_per_Cell_group.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "rm(lowest_mito)\n",
    "\n",
    "#---------- Ribosomal Percentage\n",
    "#----- Ribosomal Transcripts per cell per Sample\n",
    "lowest_ribo <- min(seurat_combined@meta.data$percent_ribo[seurat_combined@meta.data$percent_ribo > 0])\n",
    "plot <- ggSample + \n",
    "  geom_density(data = seurat_combined@meta.data %>% \n",
    "                 # handle the log 0 -- let these show up on plot \n",
    "                 # (replace 0s with the min non-zero value)\n",
    "                 mutate(percent_ribo = case_when(percent_ribo == 0 ~ lowest_ribo,\n",
    "                                                 TRUE ~ percent_ribo)), \n",
    "               alpha = 0.2,\n",
    "               aes(x=percent_ribo, fill=SAMPLE, color=SAMPLE)) + \n",
    "  ggtitle(\"Ribosomal Transcripts per Cell\") + \n",
    "  ylab(\"Percentage\") + \n",
    "  geom_vline(xintercept = percent_ribo_high, linetype=\"dashed\", colour = \"red3\") + \n",
    "  scale_x_log10() + \n",
    "  facet_wrap(~SAMPLE) + \n",
    "  theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Ribosomal_Transcripts_per_Cell_sample.png\", \n",
    "    width=1000, \n",
    "    height=1000)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Ribosomal Transcripts per cell per Group\n",
    "plot <- ggGroup + \n",
    "  geom_density(data = seurat_combined@meta.data %>% \n",
    "                 # handle the log 0 -- let these show up on plot \n",
    "                 # (replace 0s with the min non-zero value)\n",
    "                 mutate(percent_ribo = case_when(percent_ribo == 0 ~ lowest_ribo,\n",
    "                                                 TRUE ~ percent_ribo)), \n",
    "               alpha = 0.2,\n",
    "               aes(x=percent_ribo, fill=GROUP, color=GROUP)) + \n",
    "  ggtitle(\"Ribosomal Transcripts per Cell\") + \n",
    "  ylab(\"Percentage\") + \n",
    "  geom_vline(xintercept = percent_ribo_high, linetype=\"dashed\", colour = \"red3\") + \n",
    "  scale_x_log10()\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Ribosomal_Transcripts_per_Cell_group.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "rm(lowest_ribo)\n",
    "\n",
    "#---------- Complexity\n",
    "#----- Complexity Genes per Transcript per cell per Sample\n",
    "plot <- ggSample + \n",
    "  geom_density(data = seurat_combined@meta.data, \n",
    "               alpha = 0.2,\n",
    "               aes(x=log10GenesPerUMI, fill=SAMPLE, color=SAMPLE)) + \n",
    "  ggtitle(\"Complexity Genes per Transcript per Cell\") + \n",
    "  ylab(\"Cell density\") + \n",
    "  geom_vline(xintercept = log10GenesPerUMI_low, \n",
    "             linetype = \"dashed\", \n",
    "             colour = \"red3\") + \n",
    "  facet_wrap(~SAMPLE) + \n",
    "  theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Complexity_Genes_per_Transcript_per_Cell_sample.png\", \n",
    "    width=1000, \n",
    "    height=1000)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Complexity Genes per Transcript per cell per Group\n",
    "plot <- ggGroup + \n",
    "  geom_density(data = seurat_combined@meta.data, \n",
    "               alpha = 0.2,\n",
    "               aes(x=log10GenesPerUMI, fill=GROUP, color=GROUP)) + \n",
    "  ggtitle(\"Complexity Genes per Transcript per Cell\") + \n",
    "  ylab(\"Cell density\") + \n",
    "  geom_vline(xintercept = log10GenesPerUMI_low, \n",
    "             linetype = \"dashed\", \n",
    "             colour = \"red3\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Complexity_Genes_per_Transcript_per_Cell_group.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "\n",
    "#---------- UMI Transcript Counts vs Gene Counts\n",
    "#----- Number of Transcripts vs Number of Genes per cell per Sample\n",
    "plot <- ggSample + \n",
    "  geom_point(data = seurat_combined@meta.data, \n",
    "               aes(x=nUMI, y=nGene, color=percent_mito),\n",
    "               size=0.1) + \n",
    "  scale_colour_viridis_c() +\n",
    "  ggtitle(\"Number of Transcripts vs Gene Counts per Cell\") + \n",
    "  geom_vline(xintercept = nUMI_low, linetype=\"dashed\", colour = \"red3\") + \n",
    "  geom_vline(xintercept = nUMI_high, linetype=\"dashed\", colour = \"red3\") + \n",
    "  geom_hline(yintercept = nGene_low, linetype=\"dashed\", colour = \"red3\") + \n",
    "  geom_hline(yintercept = nGene_high, linetype=\"dashed\", colour = \"red3\") +  \n",
    "  scale_x_log10() + \n",
    "  scale_y_log10() + \n",
    "  stat_smooth(method=lm) + \n",
    "  facet_wrap(~SAMPLE)\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Number_of_Transcripts_vs_Genes_per_Cell_sample.png\", \n",
    "    width=1000, \n",
    "    height=1000)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Number of Transcripts vs Number of Genes per cell per Group\n",
    "plot <- ggSample + \n",
    "  geom_point(data = seurat_combined@meta.data, \n",
    "               aes(x=nUMI, y=nGene, color=percent_mito),\n",
    "               size=0.1) + \n",
    "  scale_colour_viridis_c() +\n",
    "  ggtitle(\"Number of Transcripts vs Gene Counts per Cell\") + \n",
    "  geom_vline(xintercept = nUMI_low, linetype=\"dashed\", colour = \"red3\") + \n",
    "  geom_vline(xintercept = nUMI_high, linetype=\"dashed\", colour = \"red3\") + \n",
    "  geom_hline(yintercept = nGene_low, linetype=\"dashed\", colour = \"red3\") + \n",
    "  geom_hline(yintercept = nGene_high, linetype=\"dashed\", colour = \"red3\") +  \n",
    "  scale_x_log10() + \n",
    "  scale_y_log10() + \n",
    "  stat_smooth(method=lm) + \n",
    "  facet_wrap(~GROUP)\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_QC_Number_of_Transcripts_vs_Genes_per_Cell_group.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "087400d3",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Seurat_QC_Senescence"
   },
   "outputs": [],
   "source": [
    "#---------- Senescence Markers \n",
    "#----- Number of Transcripts or Genes vs Senescence Marker per Cell sample\n",
    "plot_qc_senescence_markers <- function(senescence_gene, qc_metric) {\n",
    "  #----- check arguments: senescence_gene\n",
    "  possible_genes <- gsub(\"_Count\", \"\",\n",
    "                         names(seurat_combined@meta.data[endsWith(names(seurat_combined@meta.data), \"_Count\")]))\n",
    "  if (!(senescence_gene %in% possible_genes)) {\n",
    "    stop(paste(\"senescence_gene must be one of:\", \n",
    "               paste(possible_genes, collapse = \", \"),\n",
    "               \"\\nYou provided: senescence_gene =\", senescence_gene)\n",
    "         )\n",
    "  }\n",
    "  #----- check arguments: qc_metric\n",
    "  if (qc_metric == \"Transcripts\") {\n",
    "    vline_variable <- \"nUMI\"\n",
    "  } else if (qc_metric == \"Genes\") {\n",
    "    vline_variable <- \"nGene\"\n",
    "  } else {\n",
    "    stop(paste(\"qc_metric must either be 'Transcripts' or 'Genes'.\\n\",\n",
    "               \"You provided: qc_metric =\", qc_metric))\n",
    "  }\n",
    "  #----- make plot\n",
    "  plot <- ggSample + \n",
    "    geom_point(data = seurat_combined@meta.data, \n",
    "                 aes(x=nUMI, y=get(paste0(senescence_gene, \"_Count\")), color=percent_mito),\n",
    "                 size=0.1) + \n",
    "    scale_colour_viridis_c() +\n",
    "    ggtitle(paste(\"Number of\", qc_metric, \"vs\", senescence_gene, \"per Cell\")) + \n",
    "    labs(y = paste0(senescence_gene, \"_Count\")) +\n",
    "    geom_vline(xintercept = get(paste0(vline_variable, \"_low\")), linetype=\"dashed\", colour = \"red3\") + \n",
    "    geom_vline(xintercept = get(paste0(vline_variable, \"_high\")), linetype=\"dashed\", colour = \"red3\") + \n",
    "    scale_x_log10() + \n",
    "    facet_wrap(~SAMPLE)\n",
    "  print(plot)\n",
    "  png(file=paste0(\"seurat_results/figures/Seurat_QC_Number_of_\", qc_metric,\n",
    "                  \"_vs_\", senescence_gene, \"_per_Cell_sample.png\"), \n",
    "      width=1000, \n",
    "      height=1000)\n",
    "  print(plot)\n",
    "  dev.off()\n",
    "} \n",
    "\n",
    "# loop through all senescence genes, plotting transcripts & genes:\n",
    "senescence_genes <- gsub(\"_Count\", \"\", names(seurat_combined@meta.data[endsWith(names(seurat_combined@meta.data), \"_Count\")]))\n",
    "for (gene in senescence_genes) {\n",
    "  print(gene)\n",
    "  plot_qc_senescence_markers(gene, \"Transcripts\")\n",
    "  plot_qc_senescence_markers(gene, \"Genes\")\n",
    "}\n",
    "\n",
    "#----- Simple lm of senescence ~ age.\n",
    "p16_metadata <- seurat_combined@meta.data %>%\n",
    "  mutate(CDKN2A_Positive = case_when(CDKN2A_Count > 0 ~ TRUE,\n",
    "                                     CDKN2A_Count == 0 ~ FALSE)) %>%\n",
    "  group_by(SAMPLE, CaseID) %>%\n",
    "  summarize(total_cells = n(),\n",
    "            total_p16_transcripts = sum(CDKN2A_Count),\n",
    "            total_p16_cells = sum(CDKN2A_Positive),\n",
    "            prop_p16_cells = total_p16_cells/total_cells) %>%\n",
    "  left_join(y = covar, by = c(\"CaseID\", \"SAMPLE\"))\n",
    "\n",
    "lm(prop_p16_cells ~ AGE, data = p16_metadata) %>% summary()\n",
    "\n",
    "#----- Test lm of senescence ~ each numeric variable + age.\n",
    "numeric_vars <- p16_metadata %>% dplyr::select(where(is.numeric)) %>% dplyr::select_if(~ !any(is.na(.))) %>% names()\n",
    "numeric_vars <- numeric_vars[!numeric_vars %in% c(\"SAMPLE\", \"AGE\", \"prop_p16_cells\", \"race\", \"expired_age\")]\n",
    "lmp <- function (modelobject) {\n",
    "    # from https://stackoverflow.com/questions/5587676/pull-out-p-values-and-r-squared-from-a-linear-regression\n",
    "    if (class(modelobject) != \"lm\") stop(\"Not an object of class 'lm' \")\n",
    "    f <- summary(modelobject)$fstatistic\n",
    "    p <- pf(f[1],f[2],f[3],lower.tail=F)\n",
    "    attributes(p) <- NULL\n",
    "    return(p)\n",
    "}\n",
    "for (numeric_var in numeric_vars) {\n",
    "  # print(numeric_var)\n",
    "  lm_model <- lm(as.formula(paste(\"prop_p16_cells ~\", numeric_var, \"+ AGE\")), data = p16_metadata)\n",
    "  lm_summary <- lm_model %>% summary() \n",
    "  # print(lm_summary)\n",
    "  # lm_summary$coefficients[\"dementia_yn\"]\n",
    "  if ((lm_summary$coefficients[numeric_var,\"Pr(>|t|)\"] < 0.05) #| \n",
    "      # (lm_summary$coefficients[\"AGE\",\"Pr(>|t|)\"] < 0.05) | \n",
    "      # (lmp(lm_model) < 0.05)\n",
    "      ) {\n",
    "    message(paste(numeric_var, \"has *significant* coefficients:\"))\n",
    "    print(lm_summary)\n",
    "    print(lmp(lm_model))\n",
    "  } else {\n",
    "    message(paste(numeric_var, \"has no significant coefficients.\"))\n",
    "  }\n",
    "  \n",
    "}\n",
    "\n",
    "#----- Group * Age Interaction in Explaining p16+ Proportion\n",
    "lm(prop_p16_cells ~ GROUP + AGE, data = p16_metadata) %>% summary()\n",
    "lm(prop_p16_cells ~ GROUP*AGE, data = p16_metadata) %>% summary()\n",
    "for (any_var in names(p16_metadata)[!names(p16_metadata) %in% \n",
    "                                    c(\"SAMPLE\", \"CaseID\", \"BRAIN_REGION\", \"ID\", \n",
    "                                      \"SEQ_ID\", \"AD\", \"DLB\", \"VAD\", \"PSP\", \"HS\",\n",
    "                                      \"DLDH\", \"MND\", \"CBD\", \"PICKS\", \"HD\", \"MSA\",\n",
    "                                      \"FTLD-TDP\", \"TDP-43\", \"MS\", \"ACUTE_INFARCTS\",\n",
    "                                      \"TangleP\", \"bf_nbm\")]) { # don't have enough difference values for an interaction\n",
    "  print(any_var)\n",
    "  lm_model <- lm(as.formula(paste0(\"prop_p16_cells ~ `\", any_var, \"`*AGE\")), data = p16_metadata)\n",
    "  lm_summary <- lm_model %>% summary() \n",
    "  sig_age_interaction <- data.frame(lm_summary$coefficients[,\"Pr(>|t|)\"]) %>% \n",
    "    tibble::rownames_to_column(\"variable\") %>% \n",
    "    dplyr::rename(\"p_val\" = \"lm_summary.coefficients....Pr...t....\") %>%\n",
    "    filter(endsWith(variable, \":AGE\"),\n",
    "           p_val < .05)\n",
    "  if (nrow(sig_age_interaction) > 0) {\n",
    "    message(paste(any_var, \"has *significant* interaction coefficients:\"))\n",
    "    print(lm_summary)\n",
    "    print(lmp(lm_model))\n",
    "  } else {\n",
    "    message(paste(any_var, \"has no significant interaction coefficients.\"))\n",
    "  }\n",
    "}\n",
    "\n",
    "# MuMIn::dredge(lm(prop_p16_cells ~ GROUP*AGE + AGE*DEMENTIA, \n",
    "#                  data = p16_metadata, \n",
    "#                  na.action = na.fail),\n",
    "#               rank = 'AIC')\n",
    "\n",
    "# ggformula::gf_point(prop_p16_cells ~ AGE, color = ~MCI, shape = ~MCI, data = p16_metadata) %>%\n",
    "#   ggformula::gf_lm() + \n",
    "#   labs(title = \"MFG: MCI*AGE\", caption = \"note, we don't have that many samples with MCI...\") +\n",
    "#   lims(y = c(0, .08))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f0113b7",
   "metadata": {
    "eval": false,
    "name": "Seurat_Filtering"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Filtering\n",
    "# Filter out low quality reads using selected thresholds \n",
    "# these will change with experiment\n",
    "# seurat_filtered <- seurat_combined@meta.data %>% \n",
    "#                      filter(nUMI >= 500 & nUMI <= 20000)\n",
    "# seurat_filtered <- subset(seurat_combined, cells = seurat_filtered$cells)\n",
    "seurat_filtered <- subset(x = seurat_combined, \n",
    "                          subset = (nUMI >= nUMI_low) & \n",
    "                                   (nUMI <= nUMI_high) & \n",
    "                                   (nGene >= nGene_low) & \n",
    "                                   (nGene <= nGene_high) &\n",
    "                                   (percent_mito < percent_mito_high) & \n",
    "                                   (percent_ribo < percent_ribo_high) &\n",
    "                                   (log10GenesPerUMI > log10GenesPerUMI_low))\n",
    "\n",
    "#---- save rds\n",
    "saveRDS(seurat_filtered, file=\"seurat_results/seurat_filtered.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e73ec6a3",
   "metadata": {
    "name": "Seurat_Filtering_Figures"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a2bc58b",
   "metadata": {
    "name": "Seurat_Filtering_Senescence"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Filtering_Senescence\n",
    "#----- (repeating the same senescence demographic modeling as above, post-filter)\n",
    "\n",
    "#----- Load data\n",
    "seurat_filtered <- readRDS(file = \"seurat_results/seurat_filtered.rds\")\n",
    "\n",
    "#----- Simple lm of senescence ~ age.\n",
    "p16_metadata <- seurat_filtered@meta.data %>%\n",
    "  mutate(CDKN2A_Positive = case_when(CDKN2A_Count > 0 ~ TRUE,\n",
    "                                     CDKN2A_Count == 0 ~ FALSE)) %>%\n",
    "  group_by(SAMPLE, CaseID) %>%\n",
    "  summarize(total_cells = n(),\n",
    "            total_p16_transcripts = sum(CDKN2A_Count),\n",
    "            total_p16_cells = sum(CDKN2A_Positive),\n",
    "            prop_p16_cells = total_p16_cells/total_cells) %>%\n",
    "  left_join(y = covar, by = c(\"CaseID\", \"SAMPLE\"))\n",
    "\n",
    "lm(prop_p16_cells ~ AGE, data = p16_metadata) %>% summary()\n",
    "\n",
    "#----- Test lm of senescence ~ each numeric variable + age.\n",
    "numeric_vars <- p16_metadata %>% dplyr::select(where(is.numeric)) %>% dplyr::select_if(~ !any(is.na(.))) %>% names()\n",
    "numeric_vars <- numeric_vars[!numeric_vars %in% c(\"SAMPLE\", \"AGE\", \"prop_p16_cells\", \"race\", \"expired_age\")]\n",
    "lmp <- function (modelobject) {\n",
    "    # from https://stackoverflow.com/questions/5587676/pull-out-p-values-and-r-squared-from-a-linear-regression\n",
    "    if (class(modelobject) != \"lm\") stop(\"Not an object of class 'lm' \")\n",
    "    f <- summary(modelobject)$fstatistic\n",
    "    p <- pf(f[1],f[2],f[3],lower.tail=F)\n",
    "    attributes(p) <- NULL\n",
    "    return(p)\n",
    "}\n",
    "for (numeric_var in numeric_vars) {\n",
    "  # print(numeric_var)\n",
    "  lm_model <- lm(as.formula(paste(\"prop_p16_cells ~\", numeric_var, \"+ AGE\")), data = p16_metadata)\n",
    "  lm_summary <- lm_model %>% summary() \n",
    "  # print(lm_summary)\n",
    "  # lm_summary$coefficients[\"dementia_yn\"]\n",
    "  if ((lm_summary$coefficients[numeric_var,\"Pr(>|t|)\"] < 0.05) #| \n",
    "      # (lm_summary$coefficients[\"AGE\",\"Pr(>|t|)\"] < 0.05) |\n",
    "      # (lmp(lm_model) < 0.05)\n",
    "      ) {\n",
    "    message(paste(numeric_var, \"has *significant* coefficients:\"))\n",
    "    print(lm_summary)\n",
    "    print(lmp(lm_model))\n",
    "  } else {\n",
    "    message(paste(numeric_var, \"has no significant coefficients.\"))\n",
    "  }\n",
    "  \n",
    "}\n",
    "\n",
    "#----- Group * Age Interaction in Explaining p16+ Proportion\n",
    "lm(prop_p16_cells ~ GROUP + AGE, data = p16_metadata) %>% summary()\n",
    "lm(prop_p16_cells ~ GROUP*AGE, data = p16_metadata) %>% summary()\n",
    "for (any_var in names(p16_metadata)[!names(p16_metadata) %in% \n",
    "                                    c(\"SAMPLE\", \"CaseID\", \"BRAIN_REGION\", \"ID\", \n",
    "                                      \"SEQ_ID\", \"AD\", \"DLB\", \"VAD\", \"PSP\", \"HS\",\n",
    "                                      \"DLDH\", \"MND\", \"CBD\", \"PICKS\", \"HD\", \"MSA\",\n",
    "                                      \"FTLD-TDP\", \"TDP-43\", \"MS\", \"ACUTE_INFARCTS\",\n",
    "                                      \"TangleP\", \"bf_nbm\")]) { # don't have enough difference values for an interaction\n",
    "  print(any_var)\n",
    "  lm_model <- lm(as.formula(paste0(\"prop_p16_cells ~ `\", any_var, \"`*AGE\")), data = p16_metadata)\n",
    "  lm_summary <- lm_model %>% summary() \n",
    "  sig_age_interaction <- data.frame(lm_summary$coefficients[,\"Pr(>|t|)\"]) %>% \n",
    "    tibble::rownames_to_column(\"variable\") %>% \n",
    "    dplyr::rename(\"p_val\" = \"lm_summary.coefficients....Pr...t....\") %>%\n",
    "    filter(endsWith(variable, \":AGE\"),\n",
    "           p_val < .05)\n",
    "  if (nrow(sig_age_interaction) > 0) {\n",
    "    message(paste(any_var, \"has *significant* interaction coefficients:\"))\n",
    "    print(lm_summary)\n",
    "    print(lmp(lm_model))\n",
    "  } else {\n",
    "    message(paste(any_var, \"has no significant interaction coefficients.\"))\n",
    "  }\n",
    "}\n",
    "\n",
    "# MuMIn::dredge(lm(prop_p16_cells ~ GROUP*AGE + AGE*DEMENTIA, \n",
    "#                  data = p16_metadata, \n",
    "#                  na.action = na.fail),\n",
    "#               rank = 'AIC')\n",
    "\n",
    "ggplot(p16_metadata, aes(y = prop_p16_cells, x = AGE, color = dementia_nos, shape = dementia_nos)) +\n",
    "  geom_point() +\n",
    "  geom_smooth(method = \"lm\", se = FALSE) +\n",
    "  lims(y = c(0, NA)) +\n",
    "  labs(title = \"MFG (post-filter): Dementia*Age\")\n",
    "\n",
    "ggplot(p16_metadata, aes(y = prop_p16_cells, x = AGE, color = MCI, shape = MCI)) +\n",
    "  geom_point() +\n",
    "  geom_smooth(method = \"lm\", se = FALSE) +\n",
    "  lims(y = c(0, NA)) +\n",
    "  labs(title = \"MFG (post-filter): MCI*AGE\")\n",
    "\n",
    "p16_metadata %>%\n",
    "  mutate(bf_trans = as.character(bf_trans)) %>%\n",
    "  ggplot(aes(y = prop_p16_cells, x = AGE, color = bf_trans)) +\n",
    "  geom_point() +\n",
    "  geom_smooth(method = \"lm\", se = FALSE) +\n",
    "  lims(y = c(0, NA)) +\n",
    "  labs(title = \"MFG (post-filter): bf_trans*AGE\")\n",
    "\n",
    "p16_metadata %>%\n",
    "  # mutate(bf_trans = as.character(bf_trans)) %>%\n",
    "  ggplot(aes(y = prop_p16_cells, x = AGE, color = bf_trans)) +\n",
    "  geom_point() +\n",
    "  geom_smooth(method = \"lm\", se = FALSE) +\n",
    "  lims(y = c(0, NA)) +\n",
    "  labs(title = \"MFG (post-filter): bf_trans*AGE\") +\n",
    "  facet_wrap(~bf_trans)\n",
    "\n",
    "p16_metadata %>%\n",
    "  # mutate(bf_trans = as.character(bf_trans)) %>%\n",
    "  ggplot(aes(y = prop_p16_cells, x = AGE, color = bf_cing)) +\n",
    "  geom_point() +\n",
    "  geom_smooth(method = \"lm\", se = FALSE) +\n",
    "  lims(y = c(0, NA)) +\n",
    "  labs(title = \"MFG (post-filter): bf_cing*AGE\") +\n",
    "  facet_wrap(~bf_cing)\n",
    "\n",
    "\n",
    "p16_metadata %>%\n",
    "  # mutate(bf_trans = as.character(bf_trans)) %>%\n",
    "  ggplot(aes(y = prop_p16_cells, x = AGE, color = `Braak score`)) +\n",
    "  geom_point() +\n",
    "  geom_smooth(method = \"lm\", se = FALSE) +\n",
    "  lims(y = c(0, NA)) +\n",
    "  labs(title = \"MFG (post-filter): Braak score*AGE\") +\n",
    "  facet_wrap(~`Braak score`)\n",
    "\n",
    "p16_metadata %>%\n",
    "  # mutate(bf_trans = as.character(bf_trans)) %>%\n",
    "  ggplot(aes(y = prop_p16_cells, x = AGE)) +\n",
    "  geom_point() +\n",
    "  geom_smooth(method = \"lm\", se = TRUE) +\n",
    "  lims(y = c(0, NA)) +\n",
    "  labs(title = \"MFG (post-filter): AGE\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "283f4f7b",
   "metadata": {
    "eval": false,
    "name": "Seurat_Cell_Cycle_Scoring"
   },
   "outputs": [],
   "source": [
    "#---------- Cell_Cycle_Scoring\n",
    "#----- Load cell cycle markers\n",
    "load(file=\"seurat_results/seurat_cycle.rda\")\n",
    "\n",
    "#----- Identify Cell cycle phase for each cell\n",
    "seurat_cellcycle <- function(seurat_obj, number_features) {\n",
    "  #----- Normalize the counts\n",
    "  phase <- NormalizeData(seurat_obj)\n",
    "  #----- Identify the most variable genes\n",
    "  phase <- FindVariableFeatures(object = phase, \n",
    "                                selection.method = \"vst\",\n",
    "                                nfeatures = number_features, \n",
    "                                verbose = FALSE)\n",
    "  #----- Scale the counts\n",
    "  phase <- ScaleData(object = phase)\n",
    "  #----- Perform PCA\n",
    "  phase <- RunPCA(object = phase,\n",
    "                  verbose = FALSE)\n",
    "  #----- Score cells for cell cycle\n",
    "  phase <- CellCycleScoring(object = phase, \n",
    "                            g2m.features = g2m_genes, \n",
    "                            s.features = s_genes)\n",
    "  return(phase)\n",
    "} \n",
    "seurat_phase <- seurat_cellcycle(seurat_obj = seurat_filtered, \n",
    "                                 number_features = 2000)\n",
    "  \n",
    "#---------- Data\n",
    "#----- Save Data\n",
    "saveRDS(seurat_phase, file=\"seurat_results/seurat_phase.rds\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5418599",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "dc7d1f09",
   "metadata": {},
   "source": [
    "# PROCESSING\n",
    "\n",
    "> SOURCE: [Harmony-RNA-Workflow/scripts/main/process.R](https://github.com/shahrozeabbas/Harmony-RNA-Workflow/blob/main/scripts/main/process.R) \n",
    "\n",
    "The main processing requires subsetting to the QC features, followed by normalization.  From there identifying highly variable genes to scale, score (e.g. cell cycle).  Batch correction ought to be considered as part of scaling.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f92a7ae1",
   "metadata": {
    "eval": false,
    "name": "Seurat_Batch_Correction_RPCA"
   },
   "outputs": [],
   "source": [
    "# TEAM LEE\n",
    "\n",
    "library(Seurat)\n",
    "library(SeuratObject)\n",
    "\n",
    "#----- Load data\n",
    "seurat_filtered <- readRDS(file = \"seurat_results/seurat_filtered.rds\")\n",
    "\n",
    "#---------- Seurat_Batch_Correction_RPCA\n",
    "# Ensure the dataset uses the RNA assay\n",
    "DefaultAssay(seurat_filtered) <- \"RNA\"\n",
    "\n",
    "#----- Split Data \n",
    "# Split data into individual ‘assays’.\n",
    "seurat_split <- SplitObject(object = seurat_filtered, \n",
    "                            split.by = \"SAMPLE\")\n",
    "\n",
    "#----- Normalize and Identify Variable Features\n",
    "# LogNormalize: Feature counts for each cell are divided by the  \n",
    "# total counts for that cell and multiplied by the scale.factor  \n",
    "# (10,000 by default), and log-transforms the result. \n",
    "# Normalized values are stored in seurat_obj[[\"RNA\"]]@data.\n",
    "seurat_split <- lapply(X = seurat_split, FUN = function(x){\n",
    "    x <- NormalizeData(object = x)\n",
    "    x <- FindVariableFeatures(object = x, \n",
    "                              selection.method = \"vst\", \n",
    "                              nfeatures = 2000)\n",
    "})\n",
    "\n",
    "#----- Select features repeated across datasets for Integration\n",
    "features <- SelectIntegrationFeatures(object.list = seurat_split)\n",
    "\n",
    "#---------- Cell_Cycle_Scoring\n",
    "#----- Load cell cycle markers\n",
    "load(file=\"seurat_results/seurat_cycle.rda\")\n",
    "seurat_split <- lapply(X = seurat_split, FUN = function(x) {\n",
    "    x <- ScaleData(object = x, \n",
    "                   features = features, \n",
    "                   vars.to.regress = \"percent_mito\",\n",
    "                   verbose = FALSE)\n",
    "    x <- RunPCA(object = x, \n",
    "                features = features, \n",
    "                verbose = FALSE)\n",
    "    x <- CellCycleScoring(object = x, \n",
    "                          g2m.features = g2m_genes, \n",
    "                          s.features = s_genes)\n",
    "})\n",
    "\n",
    "#----- Perform integration\n",
    "seurat_anchors <- FindIntegrationAnchors(object.list = seurat_split, \n",
    "                                         anchor.features = features, \n",
    "                                         reduction = \"rpca\")\n",
    "rm(seurat_split)\n",
    "\n",
    "#---- Integrate across samples\n",
    "# this command creates an 'integrated' data assay\n",
    "seurat_rpca <- IntegrateData(anchorset = seurat_anchors)\n",
    "rm(seurat_anchors)\n",
    "\n",
    "# specify that we will perform downstream analysis on the corrected data\n",
    "# original unmodified data still resides in the 'RNA' assay\n",
    "DefaultAssay(seurat_rpca) <- \"integrated\"\n",
    "\n",
    "# Run the standard workflow for visualization and clustering\n",
    "seurat_rpca <- ScaleData(seurat_rpca, \n",
    "                         vars.to.regress = \"percent_mito\",\n",
    "                         verbose = FALSE)\n",
    "\n",
    "#---- Run PCA and UMAP\n",
    "seurat_rpca <- RunPCA(object = seurat_rpca, \n",
    "                      verbose = FALSE)\n",
    "# ElbowPlot(object = seurat_rpca, \n",
    "#           ndims = 50)\n",
    "seurat_rpca <- RunUMAP(seurat_rpca, \n",
    "                       reduction = \"pca\", \n",
    "                       dims = 1:30)\n",
    "\n",
    "#---- save rds\n",
    "saveRDS(seurat_rpca, file=\"seurat_results/seurat_rpca.rds\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe859290",
   "metadata": {
    "lines_to_next_cell": 0,
    "name": "Seurat_Batch_Correction_RPCA_Figures"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Batch_Correction_Harmony\n",
    "#----- Load data\n",
    "seurat_rpca <- readRDS(file = \"seurat_results/seurat_rpca.rds\")\n",
    "\n",
    "#---------- PCA\n",
    "#----- Elbow Plot\n",
    "plot <- ElbowPlot(object = seurat_rpca, \n",
    "                  ndims = 50) + \n",
    "          theme(plot.title = element_text(size = 20,\n",
    "                                          hjust = 0.5, \n",
    "                                          face = \"bold\")) +\n",
    "          ggtitle(\"PCA Elbow Plot\")\n",
    "\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_RPCA_PCA_Elbowplot.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "#----- PCA Group\n",
    "plot <- PCAPlot(object = seurat_rpca,\n",
    "                group.by = \"SAMPLE\",\n",
    "                split.by = \"GROUP\")  +\n",
    "          ggtitle(\"PCA\") + \n",
    "          scale_color_manual(values = color_sample)\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_RPCA_PCA_group.png\", \n",
    "    width=800, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "\n",
    "#----- Explore heatmap of PCs\n",
    "# Printing out the most variable genes driving PCs\n",
    "# print(x = seurat_rpca[[\"pca\"]], dims = 1:9, nfeatures = 5)\n",
    "# -- doesn't return a ggplot object, so we'll run twice...\n",
    "# -- one for RStudio viewing, one to save as png\n",
    "DimHeatmap(object = seurat_rpca, \n",
    "           dims = 1:9, \n",
    "           cells = 500, \n",
    "           balanced = TRUE)\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_RPCA_PCA_heatmap.png\", \n",
    "    width=1500, \n",
    "    height=1500,\n",
    "    pointsize = 16)\n",
    "DimHeatmap(object = seurat_rpca, \n",
    "           dims = 1:9, \n",
    "           cells = 500, \n",
    "           balanced = TRUE)\n",
    "dev.off()\n",
    "\n",
    "#---------- UMAP                             \n",
    "plot <- DimPlot(object = seurat_rpca, \n",
    "                reduction = \"umap\",\n",
    "                shuffle = TRUE,\n",
    "                group.by = \"GROUP\") + \n",
    "          scale_color_manual(values = color_group)\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_RPCA_UMAP_group.png\", \n",
    "    width=600, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "plot <- DimPlot(object = seurat_rpca, \n",
    "                reduction = \"umap\",\n",
    "                shuffle = TRUE,\n",
    "                split.by = \"GROUP\") + \n",
    "          scale_color_manual(values = color_sample)\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_RPCA_UMAP_sample.png\", \n",
    "    width=1000, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "plot <- DimPlot(object = seurat_rpca, \n",
    "                reduction = \"umap\",\n",
    "                shuffle = TRUE,\n",
    "                split.by = \"GROUP\") + \n",
    "          scale_color_manual(values = color_sample) +\n",
    "          theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_RPCA_UMAP_sample_no_legend.png\", \n",
    "    width=1000, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "plot <- DimPlot(object = seurat_rpca, \n",
    "                reduction = \"umap\",\n",
    "                shuffle = TRUE,\n",
    "                split.by = \"SAMPLE\",\n",
    "                ncol = 5) + \n",
    "          scale_color_manual(values = color_sample) +\n",
    "          theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_RPCA_UMAP_sample_all.png\", \n",
    "    width=2000, \n",
    "    height=1000)\n",
    "plot\n",
    "dev.off()\n",
    "\n",
    "\n",
    "\n",
    "rm(seurat_rpca)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b854a44a",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a1a0c6c",
   "metadata": {},
   "source": [
    "# HARMONIZE\n",
    "\n",
    "> SOURCE: [Harmony-RNA-Workflow/scripts/main/harmony.R](https://github.com/shahrozeabbas/Harmony-RNA-Workflow/blob/main/scripts/main/harmony.R) \n",
    "\n",
    "This is the final step towards our endpoint.  Dimension reduction, and cell-type identification will be performed on the \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b759cbd7",
   "metadata": {},
   "source": [
    "Harmony for batch integration when using large samples that cause memory issues  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1068350e",
   "metadata": {
    "eval": false,
    "name": "Seurat_Batch_Correction"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Batch_Correction\n",
    "# The data is preprocessed by splitting into individual ‘assays’. \n",
    "# In this experiment, each individual sample is considered an ‘assay’.\n",
    "# Ensure the dataset used here is the non-normalized dataset\n",
    "DefaultAssay(seurat_filtered) <- \"RNA\"\n",
    "#----- Load cell cycle markers\n",
    "load(file=\"seurat_results/seurat_cycle.rda\")\n",
    "\n",
    "#---------- Harmony\n",
    "#----- Batch Correction\n",
    "seurat_batch <- function(seurat_obj, number_features, sample, dimensions) {\n",
    "  #----- Normalize the counts\n",
    "  harmony <- NormalizeData(seurat_obj)\n",
    "  #----- Identify the most variable genes\n",
    "  harmony <- FindVariableFeatures(object = harmony, \n",
    "                                  selection.method = \"vst\",\n",
    "                                  nfeatures = number_features, \n",
    "                                  verbose = FALSE)\n",
    "  #----- Scale the counts\n",
    "  harmony <- ScaleData(object = harmony)\n",
    "  #----- Perform PCA\n",
    "  harmony <- RunPCA(object = harmony,\n",
    "                    verbose = FALSE)\n",
    "  #----- Score cells for cell cycle\n",
    "  harmony <- CellCycleScoring(object = harmony, \n",
    "                              g2m.features = g2m_genes, \n",
    "                              s.features = s_genes)\n",
    "  #---- Harmony Integration\n",
    "  harmony <- RunHarmony(object = harmony, \n",
    "                        group.by.vars = sample)\n",
    "  #---- Perform UMAP\n",
    "  harmony <- RunUMAP(object = harmony, \n",
    "                     reduction = \"harmony\", \n",
    "                     dims = dimensions)\n",
    "  return(harmony)\n",
    "} \n",
    "seurat_harmony <- seurat_batch(seurat_obj = seurat_filtered, \n",
    "                               number_features = 2000,\n",
    "                               sample = \"SAMPLE\",\n",
    "                               dimensions = 1:30)  \n",
    "# ElbowPlot(object = seurat_harmony, ndims = 50)\n",
    "# DimPlot(seurat_harmony, split.by = \"sample\", ncol = 6) + NoLegend()\n",
    "\n",
    "#---------- Data\n",
    "#----- Save Data\n",
    "saveRDS(seurat_harmony, file=\"seurat_results/seurat_harmony.rds\")\n",
    "#----- Remove Data\n",
    "rm(seurat_filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b40648c",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Seurat_Batch_Correction_Harmony"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Batch_Correction_Harmony\n",
    "#----- Load data\n",
    "seurat_harmony <- readRDS(file = \"seurat_results/seurat_harmony.rds\")\n",
    "\n",
    "#---------- PCA\n",
    "#----- Elbow Plot\n",
    "plot <- ElbowPlot(object = seurat_harmony, \n",
    "                  ndims = 50) + \n",
    "          theme(plot.title = element_text(size = 20,\n",
    "                                          hjust = 0.5, \n",
    "                                          face = \"bold\")) +\n",
    "          ggtitle(\"PCA Elbow Plot\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_Harmony_PCA_Elbowplot.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "#----- PCA Group\n",
    "plot <- PCAPlot(object = seurat_harmony,\n",
    "                group.by = \"SAMPLE\",\n",
    "                split.by = \"GROUP\") +\n",
    "          ggtitle(\"PCA\") + \n",
    "          scale_color_manual(values = color_sample)\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_Harmony_PCA_group.png\", \n",
    "    width=800, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "plot <- PCAPlot(object = seurat_harmony,\n",
    "                group.by = \"SAMPLE\",\n",
    "                split.by = \"GROUP\") +\n",
    "          ggtitle(\"PCA\") + \n",
    "          scale_color_manual(values = color_sample) +\n",
    "          theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_Harmony_PCA_group_no_legend.png\", \n",
    "    width=800, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "\n",
    "#----- Explore heatmap of PCs\n",
    "# Printing out the most variable genes driving PCs\n",
    "# print(x = seurat_harmony[[\"pca\"]], dims = 1:9, nfeatures = 5)\n",
    "# -- doesn't return a ggplot object, so we'll run twice...\n",
    "# -- one for RStudio viewing, one to save as png\n",
    "DimHeatmap(object = seurat_harmony, \n",
    "           dims = 1:9, \n",
    "           cells = 500, \n",
    "           balanced = TRUE)\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_Harmony_PCA_heatmap.png\", \n",
    "    width=1500, \n",
    "    height=1500,\n",
    "    pointsize = 16)\n",
    "DimHeatmap(object = seurat_harmony, \n",
    "           dims = 1:9, \n",
    "           cells = 500, \n",
    "           balanced = TRUE)\n",
    "dev.off()\n",
    "\n",
    "#---------- UMAP \n",
    "plot <- DimPlot(object = seurat_harmony, \n",
    "                reduction = \"umap\",\n",
    "                shuffle = TRUE,\n",
    "                group.by = \"GROUP\") + \n",
    "          scale_color_manual(values = color_group)\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_Harmony_UMAP_group.png\", \n",
    "    width=600, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "plot <- DimPlot(object = seurat_harmony, \n",
    "                reduction = \"umap\",\n",
    "                shuffle = TRUE,\n",
    "                split.by = \"GROUP\") + \n",
    "          scale_color_manual(values = color_sample) \n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_Harmony_UMAP_sample.png\", \n",
    "    width=1000, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "plot <- DimPlot(object = seurat_harmony, \n",
    "                reduction = \"umap\",\n",
    "                shuffle = TRUE,\n",
    "                split.by = \"GROUP\") + \n",
    "          scale_color_manual(values = color_sample) +\n",
    "          theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_Harmony_UMAP_sample_no_legend.png\", \n",
    "    width=1000, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "plot <- DimPlot(object = seurat_harmony, \n",
    "                reduction = \"umap\",\n",
    "                shuffle = TRUE,\n",
    "                split.by = \"SAMPLE\",\n",
    "                ncol = 5) + \n",
    "          scale_color_manual(values = color_sample) +\n",
    "          theme(legend.position = \"none\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Batch_Correction_Harmony_UMAP_sample_all.png\", \n",
    "    width=2000, \n",
    "    height=1000)\n",
    "plot\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b2f98f7",
   "metadata": {},
   "source": [
    "## Seurat Clustering\n",
    "\n",
    "Seurat uses a graph-based clustering approach, which embeds cells in a graph  \n",
    "structure, using a K-nearest neighbor (KNN) graph (by default), with edges  \n",
    "drawn between cells with similar gene expression patterns.  \n",
    "\n",
    "The resolution is an important argument that sets the “granularity” of the  \n",
    "downstream clustering and will need to be optimized for every individual  \n",
    "experiment. For datasets of 3,000 - 5,000 cells, the resolution set between   \n",
    "0.4-1.4 generally yields good clustering. Increased resolution values lead \n",
    "to a greater number of clusters, which is often required for larger datasets  \n",
    "  \n",
    "* Resolution 0.6  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c520524",
   "metadata": {
    "eval": false,
    "name": "Seurat_Clustering"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Clustering\n",
    "seurat_knn <- function(seurat_obj, reduction_assay, dimensions, resolutions) {\n",
    "  # Determine the K-nearest neighbor graph\n",
    "  cluster <- FindNeighbors(object = seurat_obj, \n",
    "                           reduction = reduction_assay, \n",
    "                           dims = dimensions)\n",
    "  # Clusters for various resolutions\n",
    "  cluster <- FindClusters(object = cluster,\n",
    "                          resolution = resolutions)\n",
    "  return(cluster)\n",
    "} \n",
    "seurat_cluster <- seurat_knn(seurat_obj = seurat_harmony, \n",
    "                             reduction_assay = \"harmony\",\n",
    "                             dimensions = 1:30,\n",
    "                             resolutions = seq(0.2,1.2,0.2))\n",
    "\n",
    "#---------- Seurat_Clustering Resolution\n",
    "#---------- Seurat_Clustering_UMAP\n",
    "#----- Assign identity of clusters\n",
    "Idents(object = seurat_cluster) <- \"RNA_snn_res.0.6\"\n",
    "#----- Save Resolution \n",
    "seurat_cluster@meta.data$seurat_clusters <- Idents(seurat_cluster)\n",
    "#----- Plot UMAP                             \n",
    "DimPlot(object = seurat_cluster, \n",
    "        reduction = \"umap\", \n",
    "        label = TRUE) + \n",
    "          NoLegend()\n",
    "\n",
    "#---------- Data\n",
    "#----- Save Data\n",
    "saveRDS(seurat_cluster, file=\"seurat_results/seurat_cluster.rds\")\n",
    "#----- Remove Data\n",
    "rm(seurat_harmony)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6906ccb",
   "metadata": {
    "name": "Seurat_Clustering_Figures"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Clustering_Figures\n",
    "#----- Load data\n",
    "seurat_cluster <- readRDS(file = \"seurat_results/seurat_cluster.rds\")\n",
    "#----- Plot UMAP                             \n",
    "plot <- DimPlot(object = seurat_cluster, \n",
    "                reduction = \"umap\", \n",
    "                label = TRUE) + \n",
    "          NoLegend()\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Clustering_UMAP_resolution.png\", \n",
    "    width=600, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Violin Plot\n",
    "plot <- VlnPlot(object = seurat_cluster, \n",
    "                features = c(\"nGene\", \"nUMI\", \"log10GenesPerUMI\", \"percent_mito\"), #\"mitoRatio\"), \n",
    "                ncol = 1, \n",
    "                pt.size=0, \n",
    "                group.by=\"seurat_clusters\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Clustering_violin.png\", \n",
    "    width=600, \n",
    "    height=800)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Feature Plot\n",
    "plot <- cowplot::plot_grid(nrow=2, ncol=2,\n",
    "  FeaturePlot(seurat_cluster, \n",
    "              reduction = \"umap\", \n",
    "              features = \"nGene\", \n",
    "              pt.size = 0.1, \n",
    "              order = TRUE,\n",
    "              label = TRUE) + \n",
    "    scale_colour_gradientn(colours = viridis(n = 11)),\n",
    "  FeaturePlot(seurat_cluster, \n",
    "              reduction = \"umap\", \n",
    "              features = \"nUMI\",\n",
    "              pt.size = 0.1, \n",
    "              order = TRUE,\n",
    "              label = TRUE) + \n",
    "    scale_colour_gradientn(colours = viridis(n = 11)),\n",
    "  FeaturePlot(seurat_cluster, \n",
    "              reduction = \"umap\", \n",
    "              features = \"log10GenesPerUMI\", \n",
    "              pt.size = 0.1, \n",
    "              order = TRUE,\n",
    "              label = TRUE) + \n",
    "    scale_colour_gradientn(colours = viridis(n = 11)),\n",
    "  FeaturePlot(seurat_cluster, \n",
    "              reduction = \"umap\", \n",
    "              features = \"percent_mito\", #\"mitoRatio\",\n",
    "              pt.size = 0.1, \n",
    "              order = TRUE,\n",
    "              label = TRUE) + \n",
    "    scale_colour_gradientn(colours = viridis(n = 11)))\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Clustering_feature.png\", \n",
    "    width=800, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "\n",
    "#---------- Seurat Clustering Proportions\n",
    "#----- Clustering Proportions\n",
    "cluster_proportions <- seurat_cluster@meta.data %>% \n",
    "  dplyr::group_by(GROUP, SAMPLE) %>% \n",
    "  dplyr::count(seurat_clusters) %>% \n",
    "  dplyr::rename(Count = n) %>%\n",
    "  dplyr::mutate(Proportion = Count / sum(Count)) %>% \n",
    "  dplyr::select(-Count) %>%\n",
    "  tidyr::spread(seurat_clusters, Proportion)\n",
    "#----- Save Results\n",
    "write_delim(cluster_proportions,\n",
    "            paste0(\"seurat_results/\", NameProject, \"_Clustering_Proportions.txt\"),\n",
    "            delim = \"\\t\")\n",
    "#----- Cell Cycle Proportions Figure Group\n",
    "plot <- ggGroup + \n",
    "  geom_boxplot(data = seurat_cluster@meta.data %>% \n",
    "                        dplyr::group_by(GROUP, SAMPLE) %>% \n",
    "                        dplyr::count(seurat_clusters) %>% \n",
    "                        dplyr::rename(Count = n) %>%\n",
    "                        dplyr::mutate(Proportion = Count / sum(Count)),\n",
    "               aes(x = GROUP, y = Proportion, fill = GROUP)) + \n",
    "  ggtitle(\"Number of Cells per Cluster\") + \n",
    "  theme(legend.position = \"none\") + \n",
    "  facet_wrap(~seurat_clusters, scale=\"free\", ncol = 5)\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Clustering_Proportion_group.png\", \n",
    "    width=800, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "\n",
    "#----- Cell Cycle Proportions Data\n",
    "seurat_cluster@meta.data %>% \n",
    "  dplyr::group_by(GROUP, SAMPLE) %>% \n",
    "  dplyr::count(seurat_clusters) %>% \n",
    "  dplyr::rename(Count = n) %>%\n",
    "  dplyr::mutate(Proportion = Count / sum(Count)) %>%\n",
    "  kable(caption = \"Number of Cells per Cluster\", \n",
    "        format.args = list(big.mark = \",\")) %>% \n",
    "  kable_styling(c(\"striped\", \"bordered\"))\n",
    "\n",
    "seurat_cluster@meta.data %>% \n",
    "  dplyr::group_by(BRAIN_REGION, SAMPLE) %>% \n",
    "  dplyr::count(seurat_clusters) %>% \n",
    "  dplyr::rename(Count = n) %>%\n",
    "  dplyr::mutate(Proportion = Count / sum(Count)) %>%\n",
    "  kable(caption = \"Number of Cells per Cluster\", \n",
    "        format.args = list(big.mark = \",\")) %>% \n",
    "  kable_styling(c(\"striped\", \"bordered\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "099ece8b",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Seurat_Clustering_Phase"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Clustering_Phase\n",
    "#----- Violin Plot\n",
    "plot <- VlnPlot(object = seurat_cluster, \n",
    "                features = c(\"S.Score\", \"G2M.Score\"), \n",
    "                ncol = 1, \n",
    "                pt.size=0, \n",
    "                group.by=\"seurat_clusters\")\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Clustering_Phase_violin.png\", \n",
    "    width=600, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Feature Plot\n",
    "plot <- cowplot::plot_grid(nrow=1, ncol=2,\n",
    "  FeaturePlot(seurat_cluster, \n",
    "              reduction = \"umap\", \n",
    "              features = \"S.Score\", \n",
    "              pt.size = 0.1, \n",
    "              order = TRUE,\n",
    "              label = TRUE) + \n",
    "    scale_colour_gradientn(colours = viridis(n = 11)),\n",
    "  FeaturePlot(seurat_cluster, \n",
    "              reduction = \"umap\", \n",
    "              features = \"G2M.Score\",\n",
    "              pt.size = 0.1, \n",
    "              order = TRUE,\n",
    "              label = TRUE) + \n",
    "    scale_colour_gradientn(colours = viridis(n = 11)))\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Clustering_Phase_feature.png\", \n",
    "    width=800, \n",
    "    height=300)\n",
    "plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04d6cc5f",
   "metadata": {},
   "source": [
    "## Seurat Annotation\n",
    "\n",
    "Databases for cell-type markers  \n",
    "[PanglaoDB](https://panglaodb.se/markers.html?cell_type=%27Astrocytes%27)  \n",
    "[CellMarker](http://bio-bigdata.hrbmu.edu.cn/CellMarker/search.jsp?species=Human&tissue=Brain&cellname=Astrocyte)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cf95e4a",
   "metadata": {
    "name": "Seurat_Annotation"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Annotation\n",
    "#----- Load data\n",
    "seurat_cluster <- readRDS(file = \"seurat_results/seurat_cluster.rds\")\n",
    "#----- Stacked Violin Plots\n",
    "vlnplot_gg <- function(seurat_obj, \n",
    "                       feature, \n",
    "                       pt_size = 0, \n",
    "                       plot_margin = unit(c(-0.75, 0, -0.75, 0), \"cm\"),\n",
    "                       ...){\n",
    "  p <- VlnPlot(seurat_obj, \n",
    "               features = feature, \n",
    "               pt.size = pt_size, \n",
    "               ...) + \n",
    "         ggtitle(\"\") + \n",
    "         xlab(\"\") + \n",
    "         ylab(feature) + \n",
    "         theme(legend.position = \"none\", \n",
    "               axis.text.x = element_blank(), \n",
    "               axis.ticks.x = element_blank(), \n",
    "               axis.title.y = element_text(size = rel(1), angle = 0), \n",
    "               axis.text.y = element_text(size = rel(1)), \n",
    "               plot.margin = plot_margin) \n",
    "  return(p)\n",
    "}\n",
    "vlnplot_max<- function(p){\n",
    "  ymax <- max(ggplot_build(p)$layout$panel_scales_y[[1]]$range$range)\n",
    "  return(ceiling(ymax))\n",
    "}\n",
    "vlnplot_stacked <- function(seurat_obj, \n",
    "                            features,\n",
    "                            pt_size = 0, \n",
    "                            plot_margin = unit(c(-0.75, 0, -0.75, 0), \"cm\"),\n",
    "                            ...){\n",
    "  plot_list <- purrr::map(features, function(feature) \n",
    "    vlnplot_gg(seurat_obj = seurat_obj,\n",
    "               feature = feature, \n",
    "               ...))\n",
    "  # Add back x-axis title to bottom plot. \n",
    "  plot_list[[length(plot_list)]] <- plot_list[[length(plot_list)]] + \n",
    "                                      theme(axis.text.x = element_text(), \n",
    "                                            axis.ticks.x = element_line())\n",
    "  # Change the y-axis tick to only max value \n",
    "  ymaxs <- purrr::map_dbl(plot_list, vlnplot_max)\n",
    "  plot_list <- purrr::map2(plot_list, ymaxs, function(x,y) x + \n",
    "                             scale_y_continuous(breaks = c(y)) + \n",
    "                             expand_limits(y = y))\n",
    "  p <- patchwork::wrap_plots(plotlist = plot_list, ncol = 1)\n",
    "  return(p)\n",
    "}\n",
    "\n",
    "#---------- Cell-type Marker Expression\n",
    "# rownames(seurat_markers@assays$RNA)[grepl(pattern = \"CD3\", rownames(seurat_markers@assays$RNA))]\n",
    "#----- UMAP\n",
    "DimPlot(seurat_cluster, \n",
    "        reduction=\"umap\", \n",
    "        label = TRUE) + NoLegend()\n",
    "\n",
    "#---------- Astrocyte_Cells\n",
    "# GFAP, AQP4 (2)\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster, \n",
    "                features = c(\"AQP4\", \"GFAP\"))\n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"AQP4\", \"GFAP\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Astrocyte_Restricted_Precursors_Cells\n",
    "# CD44 () -- interesting sub cluster by 2/15\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster, \n",
    "                features = c(\"CD44\"))\n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"CD44\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- B_Cells\n",
    "# CD27, CD79A not detected \n",
    "# CD38, CD24 ()\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster, \n",
    "                features = c(\"CD27\", \"CD38\", \"CD24\"))\n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"CD27\", \"CD38\", \"CD24\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- CADPS2 High\n",
    "# CADPS2, TIAM1 (?? kind of everywhere, but not both everywhere...)\n",
    "# NR4A2 for DA neurons\n",
    "# \"MAP2\", \"SCN2A\" for neurons\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster, \n",
    "                features = c(\"CADPS2\", \"TIAM1\", \"NR4A2\", \"MAP2\", \"SCN2A\"))\n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"CADPS2\", \"TIAM1\", \"NR4A2\", \"MAP2\", \"SCN2A\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Choroid_Plexus_Cells\n",
    "# KL, CHMP1A, TTR, SLC26A11 ()\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster, \n",
    "                features = c(\"KL\", \"CHMP1A\", \"SLC26A11\"))\n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"KL\", \"CHMP1A\", \"SLC26A11\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Dopaminergic_Neurons\n",
    "# # TH ()\n",
    "# vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "#                 features = c(\"TH\"))\n",
    "# FeaturePlot(object = seurat_cluster,\n",
    "#             reduction = \"umap\",\n",
    "#             features = c(\"TH\"),\n",
    "#             min.cutoff = \"q10\",\n",
    "#             max.cutoff = \"q90\",\n",
    "#             order = TRUE,\n",
    "#             label = TRUE)\n",
    "\n",
    "# DA scType\n",
    "# \"TH\", \"SLC6A3\", \"FOXA2\", \"KCNJ6\", \"NR4A2\", \"LMX1B\", \"DBH\", \"SLC6A2\", \"PPP1R1B\"\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "                features = c(\"KCNJ6\", \"NR4A2\", \"DBH\", \"PPP1R1B\", \"CADPS2\"))#, \"TH\"))\n",
    "FeaturePlot(object = seurat_cluster,\n",
    "            reduction = \"umap\",\n",
    "            features = c(\"KCNJ6\", \"NR4A2\", \"DBH\", \"PPP1R1B\", \"CADPS2\"),# \"TH\"),\n",
    "            min.cutoff = \"q10\",\n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Dopaminergic_Neurons_PD_Resistant_Macosko\n",
    "# https://www.nature.com/articles/s41593-022-01061-1.pdf\n",
    "# TH, CALB1, TMEM200A () -- TH not a slot in the data\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "                features = c(\"CALB1\", \"TMEM200A\"))\n",
    "FeaturePlot(object = seurat_cluster,\n",
    "            reduction = \"umap\",\n",
    "            features = c(\"CALB1\", \"TMEM200A\"),\n",
    "            min.cutoff = \"q10\",\n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "\n",
    "#---------- Dopaminergic_Neurons_PD_Vulnerable_Macosko\n",
    "# https://www.nature.com/articles/s41593-022-01061-1.pdf\n",
    "# TH, AGTR1, SOX6() -- TH and AGTR1 not a slot in the data\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "                features = c(\"SOX6\"))\n",
    "FeaturePlot(object = seurat_cluster,\n",
    "            reduction = \"umap\",\n",
    "            features = c(\"SOX6\"),\n",
    "            min.cutoff = \"q10\",\n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Endothelial_Cells\n",
    "# FLT1, DUSP1, RGS5, CLDN5 (9)\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster, \n",
    "                features = c(\"FLT1\", \"DUSP1\", \"RGS5\", \"CLDN5\"))\n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"FLT1\", \"DUSP1\", \"RGS5\", \"CLDN5\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Ependymal_Cells\n",
    "# FOXJ1 ()\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "                features = c(\"FOXJ1\"))\n",
    "FeaturePlot(object = seurat_cluster,\n",
    "            reduction = \"umap\",\n",
    "            features = c(\"FOXJ1\"),\n",
    "            min.cutoff = \"q10\",\n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Excitatory_Neurons (Glutaminergic neurons)\n",
    "# I'm adding CBLN2? (ask about this - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5038589/)\n",
    "# SATB2, SLC17A6, SLC17A7 (1,3,4,5?,11,16,18,19,21,22,23) #8,13?, (if 12, 21... 24. (I think 28,29 don't really fit) (12?)\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster, \n",
    "                features = c(\"SATB2\", \"SLC17A6\", \"SLC17A7\", \"CBLN2\"))\n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"SATB2\", \"SLC17A6\", \"SLC17A7\", \"CBLN2\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Inhibitory_Neurons_LHX6_Positive (GABAergic neurons)\n",
    "# https://www.nature.com/articles/s41422-018-0053-3\n",
    "# GAD1, GAD2, GRIK1: + LHX6, NXPH1, PAM (7, 15, 26)\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "                features = c(\"GAD2\", \"GAD1\", \"GRIK1\", \"LHX6\", \"NXPH1\", \"PAM\"))\n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"GAD2\", \"GAD1\", \"GRIK1\", \"LHX6\", \"NXPH1\", \"PAM\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "# #---------- Inhibitory_Neurons_CALB2_Positive (GABAergic neurons)\n",
    "# https://www.nature.com/articles/s41422-018-0053-3\n",
    "# GAD1, GAD2, GRIK1: + CALB2 (+\"NR2F2\", \"CAMLG\"??) (6)\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "                features = c(\"GAD2\", \"GAD1\", \"GRIK1\", \"CALB2\"))\n",
    "FeaturePlot(object = seurat_cluster,\n",
    "            reduction = \"umap\",\n",
    "            features = c(\"GAD2\", \"GAD1\", \"GRIK1\", \"CALB2\"),\n",
    "            min.cutoff = \"q10\",\n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Inhibitory_Neurons_Other (remaining) (GABAergic neurons)\n",
    "# GAD1, GAD2, GRIK1 (14, 7, 17, 25, 6, [ont really 26&27][part of 5?])\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "                features = c(\"GAD2\", \"GAD1\", \"GRIK1\"))\n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"GAD2\", \"GAD1\", \"GRIK1\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Macrophage_Cells\n",
    "# CD68 TREM2 () -- part of 12, but microglial fits better here\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster, \n",
    "                features = c(\"CD68\", \"TREM2\"))\n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"CD68\", \"TREM2\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Microglial_Cells\n",
    "# PLXDC2, CSF3R, \n",
    "# APBB1IP, CSF1R, CD74, (12,20?)\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster, \n",
    "                features = c(\"APBB1IP\", \"CSF1R\", \"CD74\")) \n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"APBB1IP\", \"CSF1R\", \"CD74\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Monocyte_Cells\n",
    "# CD40 CXCR4 ()\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "                features = c(\"CD40\", \"CXCR4\"))\n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"CD40\", \"CXCR4\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Neural_Stem_Cells\n",
    "# SOX2 SOX9 (2? -- no, astrocytes make more sense here.)\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster, \n",
    "                features = c(\"SOX2\", \"SOX9\"))\n",
    "FeaturePlot(object = seurat_cluster, \n",
    "            reduction = \"umap\",\n",
    "            features = c(\"SOX2\", \"SOX9\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Oligodendrocyte_Cells\n",
    "# MOG, MOBP (0) (doublet? = 15, half of 10&12, 27?,28?)\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "                features = c(\"MOG\", \"MOBP\")) \n",
    "FeaturePlot(object = seurat_cluster,\n",
    "            reduction = \"umap\",\n",
    "            features = c(\"MOG\", \"MOBP\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Oligodendrocyte_Precursor_Cells\n",
    "# OLIG1, VCAN (10)\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "                features = c(\"VCAN\", \"OLIG1\"))\n",
    "FeaturePlot(object = seurat_cluster,\n",
    "            reduction = \"umap\",\n",
    "            features = c(\"VCAN\", \"OLIG1\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- Pericyte_Cells\n",
    "# PDGFRB (maybe half of 9, and some in 2)\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "                features = c(\"PDGFRB\")) #, \"CSPG4\", \"ACTA2\", \"NES\"))\n",
    "FeaturePlot(object = seurat_cluster,\n",
    "            reduction = \"umap\",\n",
    "            features = c(\"PDGFRB\"), #, \"CSPG4\", \"ACTA2\", \"NES\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)\n",
    "\n",
    "#---------- T_Cells\n",
    "# CCR6 CD4 ()\n",
    "vlnplot_stacked(seurat_obj = seurat_cluster,\n",
    "                features = c(\"CCR6\", \"CD4\"))\n",
    "FeaturePlot(object = seurat_cluster,\n",
    "            reduction = \"umap\",\n",
    "            features = c(\"CCR6\", \"CD4\"), \n",
    "            min.cutoff = \"q10\", \n",
    "            max.cutoff = \"q90\",\n",
    "            order = TRUE,\n",
    "            label = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0925c63c",
   "metadata": {
    "eval": false,
    "name": "Seurat_Annotation_Assign"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Annotation_Assign\n",
    "#----- Annotate each cluster\n",
    "seurat_anno <- RenameIdents(object = seurat_cluster, \n",
    "                            `0` =  \"Oligodendrocyte_Cells\",\n",
    "                            `1` =  \"Excitatory_Neurons\",\n",
    "                            `2` =  \"Astrocyte_Cells\",\n",
    "                            `3` =  \"Excitatory_Neurons\",\n",
    "                            `4` =  \"Excitatory_Neurons\",\n",
    "                            `5` =  \"Excitatory_Neurons\", #??\n",
    "                            `6` =  \"Inhibitory_Neurons\",\n",
    "                            `7` =  \"Inhibitory_Neurons\",\n",
    "                            `8` =  \"Excitatory_Neurons\",\n",
    "                            `9` =  \"Endothelial_Cells\",\n",
    "                            `10` = \"Oligodendrocyte_Precursor_Cells\", #?Oligodendrocyte_Cells (these are transitioning!!!?\n",
    "                            `11` = \"Excitatory_Neurons\",\n",
    "                            `12` = \"Microglial_Cells\",\n",
    "                            `13` = \"Excitatory_Neurons\",\n",
    "                            `14` = \"Inhibitory_Neurons\",\n",
    "                            `15` = \"doublet_astro_oligo_1\", #?Oligodendrocyte_Cells\n",
    "                            `16` = \"Excitatory_Neurons\",\n",
    "                            `17` = \"Inhibitory_Neurons\", \n",
    "                            `18` = \"Excitatory_Neurons\",\n",
    "                            `19` = \"Excitatory_Neurons\",\n",
    "                            `20` = \"Microglial_Cells\", #??? Microglial_Cells\n",
    "                            `21` = \"Excitatory_Neurons\",\n",
    "                            `22` = \"Excitatory_Neurons\",\n",
    "                            `23` = \"Excitatory_Neurons\",\n",
    "                            `24` = \"doublet_astro_oligo_2\",\n",
    "                            `25` = \"Inhibitory_Neurons\",\n",
    "                            `26` = \"Inhibitory_Neurons\", #??\n",
    "                            `27` = \"doublet_InN_oligo\", #??\n",
    "                            `28` = \"doublet_ExN_oligo\", #Excitatory_Neurons???\n",
    "                            `29` = \"Excitatory_Neurons\") #Excitatory_Neurons\n",
    "                            \n",
    "#----- save these annotations in meta.data\n",
    "seurat_anno@meta.data$cluster_annotation <- as.factor(as.character(Idents(seurat_anno)))\n",
    "#----- UMAP\n",
    "DimPlot(seurat_anno, \n",
    "        reduction=\"umap\", repel = TRUE,\n",
    "        label = TRUE) + NoLegend()\n",
    "\n",
    "#---------- Seurat_Annotation_Manual\n",
    "#----- Metadata\n",
    "metadata <- seurat_anno@meta.data\n",
    "# metadata$cluster_annotation <- \"Na\"\n",
    "# #----- Manual Edit\n",
    "# anno_umap <- DimPlot(seurat_anno, reduction=\"umap\", label = TRUE) \n",
    "# anno_cells <- CellSelector(plot = anno_umap)\n",
    "# metadata$cluster_annotation[metadata$cells %in% anno_cells] <- \"Oligodendrocyte_Precursor_Cells\"\n",
    "#----- Metadata\n",
    "seurat_anno@meta.data <- metadata\n",
    "#----- Assign identity of clusters\n",
    "Idents(object = seurat_anno) <- \"cluster_annotation\"\n",
    "#----- UMAP\n",
    "DimPlot(seurat_anno, \n",
    "        reduction=\"umap\", repel = TRUE,\n",
    "        label = TRUE) + NoLegend()\n",
    "\n",
    "#---------- Data\n",
    "#----- Save Data\n",
    "saveRDS(seurat_anno, file=\"seurat_results/seurat_anno.rds\")\n",
    "#----- Remove Data\n",
    "rm(seurat_cluster, metadata, anno_umap, anno_cells)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e1de426",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Seurat_Annotation_Figures"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Annotation_Figures\n",
    "#----- Load data\n",
    "seurat_anno <- readRDS(file = \"seurat_results/seurat_anno.rds\")\n",
    "#----- Plot UMAP                             \n",
    "plot <- (DimPlot(object = seurat_anno, \n",
    "                 reduction = \"umap\" #,\n",
    "                 # custom colors: not used. \n",
    "                 # cols = c(\"#2887C4\", # oligodendrocytes\n",
    "                 #          \"#ACA4E2\", # exN c1\n",
    "                 #          \"#3FA294\", # astrocytes\n",
    "                 #          \"#92ABE1\", # exN c3\n",
    "                 #          \"#FBCC6B\", # inhibitory neurons\n",
    "                 #          \"#59B7D2\", # exN c4\n",
    "                 #          \"#1B3C74\", # OPCs\n",
    "                 #          \"#76B2DB\", # exN c5\n",
    "                 #          \"#71299A\", # microglial\n",
    "                 #          \"#40BBC5\", # exN c2\n",
    "                 #          \"#A9C47F\", # endothelial\n",
    "                 #          \"#38BDB5\", # exN c6\n",
    "                 #          \"#45BEA4\", # exN c8\n",
    "                 #          \"#5CBD92\"  # exN c7\n",
    "                 #          )\n",
    "                 ) + \n",
    "          NoLegend()) %>%\n",
    "          LabelClusters(plot = ., id = \"ident\", repel = TRUE, box.padding = 1)\n",
    "plot\n",
    "ggsave(filename = \"seurat_results/figures/Seurat_Annotation_UMAP_resolution_high_VAI.png\", \n",
    "       plot = plot,\n",
    "       width = 600*6,\n",
    "       height = 600*6,\n",
    "       dpi = 500,\n",
    "       units = \"px\")\n",
    "png(file=\"seurat_results/figures/Seurat_Annotation_UMAP_resolution_VAI.png\", \n",
    "    width=600, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Plot UMAP group\n",
    "plot <- (DimPlot(object = seurat_anno, \n",
    "                reduction = \"umap\", \n",
    "                split.by = \"GROUP\") + \n",
    "          NoLegend()) %>%\n",
    "          LabelClusters(plot = ., id = \"ident\", repel = TRUE, box.padding = 1.5)\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Annotation_UMAP_resolution_group.png\", \n",
    "    width=1000, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Plot UMAP group (no labels)\n",
    "plot <- (DimPlot(object = seurat_anno, \n",
    "                reduction = \"umap\", \n",
    "                split.by = \"GROUP\") + \n",
    "          NoLegend())\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Annotation_UMAP_resolution_group_no_labels.png\", \n",
    "    width=1000, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Cell Markers\n",
    "# levels_active_ident <- factor(sort(levels(seurat_anno@active.ident)))\n",
    "# seurat_anno@active.ident <- levels_active_ident\n",
    "plot <- DotPlot(object = seurat_anno,\n",
    "                features = c(\"AQP4\",    # Astrocyte_Cells\n",
    "                             \"FLT1\",    # Endothelial_Cells\n",
    "                             \"SLC17A7\", # Excitatory_Neurons\n",
    "                             # \"CALB2\",   # Inhibitory_Neurons (CALB2+)\n",
    "                             # \"LHX6\",    # Inhibitory_Neurons (LHX6+)\n",
    "                             \"GAD1\",    # Inhibitory_Neurons (all)\n",
    "                             \"APBB1IP\", # Microglial_Cells\n",
    "                             \"MOBP\",    # Oligodendrocyte_Cells\n",
    "                             \"VCAN\"),   # Oligodendrocyte_Precursor_Cells\n",
    "                cols = c(\"blue\", \"red\"),  \n",
    "                dot.scale = 8) + \n",
    "          RotatedAxis()\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Annotation_Cell_Markers.png\", \n",
    "    width=800, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae8acd4f",
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 2,
    "name": "Seurat_PD_Gene_Figures"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_PD_Gene_Figures\n",
    "#----- PDgenes\n",
    "# genes are not included if they aren't in the top variable genes\n",
    "# (that means they aren't foudn in slot data, and errors occur.)\n",
    "PDgenes_ASAP <- c(\"SNCA\", \"LRRK2\", \"VPS35\", \"GBA\")\n",
    "PDneurogenes <- c(\"ATP13A2\", \"ATXN2\", \"CHCHD2\", \"DNAJC6\", \"EIF4G1\", \n",
    "                  \"GBA\", \"GIGYF2\", \"LRRK2\", \"MAPT\", \"PRKN\", \"PARK7\", \"PINK1\", \n",
    "                  \"PLA2G6\", \"PRKAR1B\", \"RAB39B\", \"SNCA\", \"SYNJ1\", \"VPS35\"\n",
    "                  #\"FBX07\", \"AGTR1\"\n",
    "                  )\n",
    "extra_genes <- c('DDRGK1', 'DLG2', 'ELOVL7', 'FGD4', 'FYN', 'GPNMB', 'HIP1R', 'HLA-DRB5', 'IP6K2', 'ITPKB', 'KANSL1', 'NSF', 'PAM', 'RIMS1', 'RPS12', 'STK39', 'TMEM163', 'ZNF608')\n",
    "PDgenes <- unique(c(PDgenes_ASAP, PDneurogenes, extra_genes))\n",
    "\n",
    "for (PDgene in PDgenes){\n",
    "  #----- Plot Expression: Overall\n",
    "  # UMAP\n",
    "  plot <- (FeaturePlot(object = seurat_anno, \n",
    "                      reduction = \"umap\",\n",
    "                      features = PDgene,\n",
    "                      order = TRUE) + \n",
    "            scale_colour_gradientn(colours = viridis(n = 11, \n",
    "                                                     direction = -1))) %>%\n",
    "           LabelClusters(plot = ., id = \"ident\", repel = TRUE, box.padding = 1)\n",
    "  print(plot)\n",
    "  png(file=paste0(\"seurat_results/figures/Seurat_Annotation_PDgene_UMAP_\", \n",
    "                  PDgene, \n",
    "                  \".png\"), \n",
    "      width=600, \n",
    "      height=600)\n",
    "  print(plot)\n",
    "  dev.off()\n",
    "  \n",
    "  # Violin Plot\n",
    "  plot <- VlnPlot(object = seurat_anno, \n",
    "                  features = PDgene,\n",
    "                  pt.size = 0) + \n",
    "            theme(legend.position = \"none\")\n",
    "  print(plot)\n",
    "  png(file=paste0(\"seurat_results/figures/Seurat_Annotation_PDgene_Violin_\", \n",
    "                  PDgene, \n",
    "                  \".png\"), \n",
    "      width=600, \n",
    "      height=600)\n",
    "  print(plot)\n",
    "  dev.off()\n",
    "  \n",
    "  #----- Plot Expression: Separate by PD vs. HC\n",
    "  # UMAP\n",
    "  plot_seperate_features(seurat_anno, PDgene, ident = \"GROUP\", pt.size = .1, numcol = 2)\n",
    "  ggsave(filename = paste0(\"seurat_results/figures/Seurat_Annotation_PDgene_UMAP_group_\", \n",
    "                           PDgene, \".png\"), \n",
    "         plot = plot_seperate_features(seurat_anno, PDgene, ident = \"GROUP\", pt.size = .05, numcol = 2),\n",
    "         width = 600*2,\n",
    "         height = 400*2,\n",
    "         dpi = 150,\n",
    "         units = \"px\"\n",
    "         )\n",
    "  \n",
    "  # Violin Plot\n",
    "  plot <- VlnPlot(object = seurat_anno, \n",
    "                  features = PDgene,\n",
    "                  pt.size = 0,\n",
    "                  split.by = \"GROUP\",\n",
    "                  split.plot = TRUE)\n",
    "  print(plot)\n",
    "  png(file=paste0(\"seurat_results/figures/Seurat_Annotation_PDgene_Violin_group_\", \n",
    "                  PDgene, \n",
    "                  \".png\"), \n",
    "      width=600, \n",
    "      height=600)\n",
    "  print(plot)\n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbf0f824",
   "metadata": {},
   "source": [
    "## Seurat Cell Proportions\n",
    "\n",
    "Celltype Proportion  \n",
    "Count of Celltype per sample / Count of all Celltypes per sample  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8c9f2f7",
   "metadata": {
    "eval": false,
    "name": "Seurat_Cell_Proportions"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Cell_Proportions\n",
    "# beta regression models are designed for data between 0 and 1\n",
    "# This give me percent differences\n",
    "# emmeans(betafit,pairwise~Group)\n",
    "# emmeans(betafit,pairwise~Group,type=\"response\")\n",
    "# To test that group is important, second way to be pvaule\n",
    "# betafit1 <- betareg(Oligodendrocytes ~ Group, data = cellproportions)\n",
    "# betafit2 <- betareg(Oligodendrocytes ~ 1, data = cellproportions)\n",
    "# lrtest(betafit1,betafit2)\n",
    "#----- Load data if needed\n",
    "if (!exists(\"seurat_anno\")) {\n",
    "  seurat_anno <- readRDS(file = \"seurat_results/seurat_anno.rds\")\n",
    "}\n",
    "\n",
    "#----- Cell Proportions\n",
    "cell_proportions <- seurat_anno@meta.data %>% \n",
    "  dplyr::count(cluster_annotation, GROUP, SAMPLE, AGE, SEX, PMI) %>% \n",
    "  dplyr::group_by(SAMPLE) %>% \n",
    "  dplyr::rename(Count = n) %>%\n",
    "  dplyr::mutate(Proportion = Count / sum(Count)) %>% \n",
    "  dplyr::select(-Count) %>%\n",
    "  tidyr::spread(cluster_annotation, Proportion) %>% \n",
    "  replace(is.na(.), 0)\n",
    "#----- Save Results\n",
    "write_delim(cell_proportions, \n",
    "            paste0(\"seurat_results/\", NameProject, \"_Cell_Proportions.txt\"), \n",
    "            delim = \"\\t\")\n",
    "\n",
    "#---------- Cell Proportions differences\n",
    "#----- Cell Types\n",
    "cell_types <- seurat_anno@meta.data %>% \n",
    "  dplyr::arrange(cluster_annotation) %>% \n",
    "  dplyr::select(cluster_annotation) %>% \n",
    "  unique() %>% \n",
    "  pull(cluster_annotation)\n",
    "#----- Cell Proportions Differences - Group\n",
    "cell_proportions_differences <- lapply(cell_types, function(cell_type) {\n",
    "  fit_data <- cell_proportions %>% \n",
    "    dplyr::select(GROUP, SAMPLE, AGE, cell_type)\n",
    "  fit_data1 <- fit_data %>% \n",
    "    dplyr::select(GROUP, SAMPLE, AGE, cell_type) %>%\n",
    "    dplyr::rename(celltype = cell_type) %>% \n",
    "    dplyr::mutate(lemon_squeezer = (celltype * (10 - 1) + 0.5) / 10)\n",
    "  fit_lm <- lm(formula = as.formula(paste(cell_type, \"~ GROUP + AGE\")), \n",
    "              data = fit_data)\n",
    "  fit_beta <- betareg(formula = \"lemon_squeezer ~ GROUP + AGE\", \n",
    "                     data = fit_data1) \n",
    "  fit_results <- data.frame(Celltype = cell_type,\n",
    "                   lm_pvalue = summary(fit_lm)$coefficients[2,4],\n",
    "                   lm_diff = summary(fit_lm)$coefficients[2,1],\n",
    "                   betareg_pvalue = summary(fit_beta)$coefficients$mean[2,4],\n",
    "                   betareg_change = exp(summary(fit_beta)$coefficients$mean[2,1]))\n",
    "  return(fit_results)\n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(cell_proportions_differences, \n",
    "            paste0(\"seurat_results/\", NameProject, \"_Cell_Proportions_Differences_Group.txt\"), \n",
    "            delim = \"\\t\")\n",
    "#----- Cell Proportions Differences - Group (with PMI & SEX as covariates)\n",
    "cell_proportions_differences_sex_pmi <- lapply(cell_types, function(cell_type) {\n",
    "  fit_data <- cell_proportions %>% \n",
    "    dplyr::select(GROUP, SAMPLE, AGE, SEX, PMI, cell_type)\n",
    "  fit_data1 <- fit_data %>% \n",
    "    dplyr::select(GROUP, SAMPLE, AGE, SEX, PMI, cell_type) %>%\n",
    "    dplyr::rename(celltype = cell_type) %>% \n",
    "    dplyr::mutate(lemon_squeezer = (celltype * (10 - 1) + 0.5) / 10)\n",
    "  fit_lm <- lm(formula = as.formula(paste(cell_type, \"~ GROUP + AGE + SEX + PMI\")), \n",
    "              data = fit_data)\n",
    "  fit_beta <- betareg(formula = \"lemon_squeezer ~ GROUP + AGE + SEX + PMI\", \n",
    "                     data = fit_data1) \n",
    "  fit_results <- data.frame(Celltype = cell_type,\n",
    "                   # (using more specific column names instead of indexes in case this changes)\n",
    "                   lm_pvalue = summary(fit_lm)$coefficients[\"GROUPPD\",\"Pr(>|t|)\"],\n",
    "                   lm_diff = summary(fit_lm)$coefficients[\"GROUPPD\",\"Estimate\"],\n",
    "                   betareg_pvalue = summary(fit_beta)$coefficients$mean[\"GROUPPD\",\"Pr(>|z|)\"],\n",
    "                   betareg_change = exp(summary(fit_beta)$coefficients$mean[\"GROUPPD\",\"Estimate\"]))\n",
    "  return(fit_results)\n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(cell_proportions_differences_sex_pmi, \n",
    "            paste0(\"seurat_results/\", NameProject,\n",
    "                   \"_Cell_Proportions_Differences_Group_PMI_Sex.txt\"), \n",
    "            delim = \"\\t\")\n",
    "\n",
    "#---------- Cell Cycle Phase Proportions\n",
    "cell_phase_proportions <- seurat_anno@meta.data %>% \n",
    "  dplyr::group_by(SAMPLE, AGE, SEX, PMI) %>% \n",
    "  dplyr::count(cluster_annotation, GROUP, SAMPLE, Phase) %>% \n",
    "  dplyr::rename(Count = n) %>%\n",
    "  tidyr::spread(Phase, Count) %>%\n",
    "  replace(is.na(.), 0) %>%\n",
    "  dplyr::mutate(G1_Proportion = G1/(G1+G2M+S),\n",
    "                G2M_Proportion = G2M/(G1+G2M+S),\n",
    "                S_Proportion = S/(G1+G2M+S)) %>%\n",
    "  dplyr::select(-G1,-G2M,-S) %>%  \n",
    "  melt(id.vars=c(\"cluster_annotation\", \"GROUP\", \"SAMPLE\", \"AGE\", \"SEX\", \"PMI\")) %>% \n",
    "  dplyr::rename(Phase = variable, \n",
    "                Proportion = value) %>% \n",
    "  tidyr::spread(cluster_annotation, Proportion)  %>% \n",
    "  replace(is.na(.), 0)\n",
    "#----- Save Results\n",
    "write_delim(cell_phase_proportions, \n",
    "            paste0(\"seurat_results/\", NameProject, \"_Cell_Proportions_Phase_Group.txt\"), \n",
    "            delim = \"\\t\")\n",
    "\n",
    "#---------- Cell Cycle Phase Proportions differences\n",
    "#----- Cell Types\n",
    "cell_types <- seurat_anno@meta.data %>% \n",
    "  dplyr::arrange(cluster_annotation) %>% \n",
    "  dplyr::select(cluster_annotation) %>% \n",
    "  unique() %>% \n",
    "  pull(cluster_annotation)\n",
    "#----- Phase\n",
    "phases <- c(\"G1_Proportion\", \"G2M_Proportion\", \"S_Proportion\")\n",
    "#----- Cell Cycle Phase Proportions Differences\n",
    "cell_phase_proportions_differences <- lapply(cell_types, function(cell_type) {\n",
    "  lapply(phases, function(phase){\n",
    "    fit_data <- cell_phase_proportions %>% \n",
    "      dplyr::select(GROUP, SAMPLE, AGE, Phase, cell_type) %>% \n",
    "      dplyr::filter(Phase == phase)\n",
    "    fit_data1 <- fit_data %>% \n",
    "      dplyr::rename(celltype = cell_type) %>% \n",
    "      dplyr::mutate(lemon_squeezer = (celltype * (10 - 1) + 0.5) / 10)\n",
    "    fit_lm <- lm(formula = as.formula(paste(cell_type, \"~ GROUP + AGE\")), \n",
    "                 data = fit_data)\n",
    "    fit_beta <- betareg(formula = \"lemon_squeezer ~ GROUP + AGE\", \n",
    "                        data = fit_data1) \n",
    "    fit_results <- data.frame(Celltype = cell_type,\n",
    "                     Phase = phase,\n",
    "                     lm_pvalue = summary(fit_lm)$coefficients[2,4],\n",
    "                     lm_diff = summary(fit_lm)$coefficients[2,1],\n",
    "                     betareg_pvalue = summary(fit_beta)$coefficients$mean[2,4],\n",
    "                     betareg_change = exp(summary(fit_beta)$coefficients$mean[2,1]))\n",
    "    return(fit_results)\n",
    "  }) \n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(cell_phase_proportions_differences, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Cell_Proportions_Phase_Differences_Group.txt\"), \n",
    "  delim = \"\\t\")\n",
    "#----- Cell Cycle Phase Proportions Differences (with PMI & SEX as covariates)\n",
    "cell_phase_proportions_differences_sex_pmi <- lapply(cell_types, function(cell_type) {\n",
    "  lapply(phases, function(phase){\n",
    "    fit_data <- cell_phase_proportions %>% \n",
    "      dplyr::select(GROUP, SAMPLE, AGE, SEX, PMI, Phase, cell_type) %>% \n",
    "      dplyr::filter(Phase == phase)\n",
    "    fit_data1 <- fit_data %>% \n",
    "      dplyr::rename(celltype = cell_type) %>% \n",
    "      dplyr::mutate(lemon_squeezer = (celltype * (10 - 1) + 0.5) / 10)\n",
    "    fit_lm <- lm(formula = as.formula(paste(cell_type, \"~ GROUP + AGE + SEX + PMI\")), \n",
    "                 data = fit_data)\n",
    "    fit_beta <- betareg(formula = \"lemon_squeezer ~ GROUP + AGE + SEX + PMI\", \n",
    "                        data = fit_data1) \n",
    "    fit_results <- data.frame(Celltype = cell_type,\n",
    "                     Phase = phase,\n",
    "                     # (using more specific column names instead of indexes in case this changes)\n",
    "                     lm_pvalue = summary(fit_lm)$coefficients[\"GROUPPD\",\"Pr(>|t|)\"],\n",
    "                     lm_diff = summary(fit_lm)$coefficients[\"GROUPPD\",\"Estimate\"],\n",
    "                     betareg_pvalue = summary(fit_beta)$coefficients$mean[\"GROUPPD\",\"Pr(>|z|)\"],\n",
    "                     betareg_change = exp(summary(fit_beta)$coefficients$mean[\"GROUPPD\",\"Estimate\"]))\n",
    "    return(fit_results)\n",
    "  }) \n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(cell_phase_proportions_differences_sex_pmi, \n",
    "  paste0(\"seurat_results/\", NameProject, \n",
    "         \"_Cell_Proportions_Phase_Differences_Group_PMI_Sex.txt\"), \n",
    "  delim = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cddde39",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Seurat_Cell_Proportions_Figures"
   },
   "outputs": [],
   "source": [
    "#---------- Cell Proportions\n",
    "cell_proportions <- \n",
    "  read_delim(paste0(\"seurat_results/\", NameProject, \"_Cell_Proportions.txt\"), \n",
    "  delim = \"\\t\", escape_double = FALSE, \n",
    "  trim_ws = TRUE)\n",
    "#----- Cell Proportions Data\n",
    "cell_proportions %>%\n",
    "  kable(caption = \"Proportion of Cells per Cluster\", \n",
    "        format.args = list(big.mark = \",\")) %>% \n",
    "  kable_styling(c(\"striped\", \"bordered\"))\n",
    "#----- Cell Proportions Figure by Sample\n",
    "# to be plotted later -- \n",
    "# plot <- ggSample +\n",
    "#   geom_boxplot(data = cell_proportions %>%\n",
    "#                  pivot_longer(cols = Oligodendrocyte_Cells:Endothelial_Cells,\n",
    "#                               names_to = \"Celltype\",\n",
    "#                               values_to = \"Proportion\") %>%\n",
    "#                  mutate(GROUP = as.factor(GROUP)\n",
    "#                         # outlier = case_when(.$Proportion == 0)\n",
    "#                         )\n",
    "#                ,\n",
    "#                aes(x = Celltype, y = Proportion, fill = GROUP)) +\n",
    "#   ggtitle(\"Proportion of Cells per Cluster\") +\n",
    "#   theme(legend.position = \"none\") #+\n",
    "#   # facet_wrap(~GROUP)\n",
    "#   #geom_text_repel(aes(label = ))\n",
    "#   # facet_wrap(~cluster_annotation, scale=\"free\", ncol =4)\n",
    "# plot\n",
    "\n",
    "#----- FIGURES & TABLES BY GROUP\n",
    "#----- Cell Proportions Figure\n",
    "plot <- ggGroup + \n",
    "  geom_boxplot(data = cell_proportions %>% \n",
    "                        melt(id.vars=c(\"GROUP\", \"SAMPLE\", \"AGE\", \"SEX\", \"PMI\")) %>% \n",
    "                        dplyr::rename(cluster_annotation = variable,\n",
    "                                      Proportion = value),\n",
    "               aes(x = GROUP, y = Proportion, fill = GROUP)) + \n",
    "  ggtitle(\"Proportion of Cells per Cluster\") + \n",
    "  theme(legend.position = \"none\") + \n",
    "  facet_wrap(~cluster_annotation, scale=\"free\", ncol =4)\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Cell_Proportions_group.png\", \n",
    "    width=800, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "ggsave(file=\"seurat_results/figures/Seurat_Cell_Proportions_group_high_res.png\")\n",
    "#----- Cell Proportions Differences\n",
    "cell_proportions_differences <- read_delim(\n",
    "  paste0(\"seurat_results/\", NameProject, \"_Cell_Proportions_Differences_Group.txt\"), \n",
    "    delim = \"\\t\", escape_double = FALSE, \n",
    "    trim_ws = TRUE)\n",
    "#----- Cell Proportions Differences Data\n",
    "cell_proportions_differences %>% \n",
    "  kable(caption = \"Proportion of Cells per Cluster Differeneces\", \n",
    "        format.args = list(big.mark = \",\")) %>% \n",
    "  kable_styling(c(\"striped\", \"bordered\"))\n",
    "\n",
    "#---------- Cell Cycle Phase Proportions\n",
    "cell_phase_proportions <- \n",
    "  read_delim(paste0(\"seurat_results/\", NameProject, \"_Cell_Proportions_Phase_Group.txt\"), \n",
    "  delim = \"\\t\", escape_double = FALSE, \n",
    "  trim_ws = TRUE)\n",
    "#----- Cell Cycle Phase Proportions Data\n",
    "cell_phase_proportions %>%\n",
    "  kable(caption = \"Proportion of Cells per Cluster\", \n",
    "        format.args = list(big.mark = \",\")) %>% \n",
    "  kable_styling(c(\"striped\", \"bordered\"))\n",
    "#----- Cell Cycle Phase Proportions Figure\n",
    "plot <- ggGroup + \n",
    "  geom_boxplot(data = cell_phase_proportions %>% \n",
    "                        melt(id.vars=c(\"GROUP\", \"SAMPLE\", \"AGE\", \"Phase\", \"SEX\", \"PMI\")) %>% \n",
    "                        dplyr::rename(cluster_annotation = variable,\n",
    "                                      Proportion = value),\n",
    "               aes(x = Phase, y = Proportion, fill = GROUP)) + \n",
    "  ggtitle(\"Proportion of Cells per Cluster\") + \n",
    "  # theme(legend.position = \"none\") + \n",
    "  facet_wrap(~cluster_annotation, scale=\"free\", ncol =4)\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Cell_Proportions_Phase_group.png\", \n",
    "    width=800, \n",
    "    height=600)\n",
    "plot\n",
    "dev.off()\n",
    "#----- Cell Proportions Differences\n",
    "cell_phase_proportions_differences <- read_delim(\n",
    "  paste0(\"seurat_results/\", NameProject, \"_Cell_Proportions_Phase_Differences_Group.txt\"), \n",
    "    delim = \"\\t\", escape_double = FALSE, \n",
    "    trim_ws = TRUE)\n",
    "#----- Cell Proportions Differences Data\n",
    "cell_phase_proportions_differences %>% \n",
    "  kable(caption = \"Proportion of Cells per Cluster Differeneces\", \n",
    "        format.args = list(big.mark = \",\")) %>% \n",
    "  kable_styling(c(\"striped\", \"bordered\"))\n",
    "\n",
    "\n",
    "#---------- Data\n",
    "#----- Remove Data\n",
    "rm(cell_proportions_differences, \n",
    "   cell_phase_proportions, cell_phase_proportions_differences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d957f1d2",
   "metadata": {
    "name": "Seurat_Cell_Proportions_Speckle"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Cell_Proportions_Speckle\n",
    "# install if package is not already installed.\n",
    "if (!(\"speckle\" %in% rownames(installed.packages()))) {\n",
    "  remotes::install_github(\"Oshlack/speckle\", \n",
    "                          build_vignettes = TRUE, \n",
    "                          dependencies = \"Suggest\")\n",
    "}\n",
    "library(BiocStyle)\n",
    "library(CellBench)\n",
    "library(limma)\n",
    "library(scater)\n",
    "library(speckle)\n",
    "#----- Our Data\n",
    "props <- getTransformedProps(clusters = seurat_anno$cluster_annotation, \n",
    "                             sample = seurat_anno$SAMPLE, \n",
    "                             transform = \"logit\")\n",
    "barplot(props$Proportions, legend=TRUE, ylab=\"Proportions\")\n",
    "plot <- props$Proportions %>% \n",
    "  melt(id.vars=c(\"clusters\")) %>% \n",
    "  dplyr::rename(Proportion = value) %>% \n",
    "  dplyr::mutate(clusters = as.character(clusters)) %>%\n",
    "  dplyr::mutate(sample = as.character(sample)) %>%\n",
    "  ggplot() + \n",
    "  geom_bar(aes(x = sample, y = Proportion, fill = clusters),\n",
    "           position=\"fill\", stat=\"identity\") + \n",
    "  ggtitle(\"Proportion of Cells per Cluster\") + \n",
    "  # scale_fill_brewer(palette = \"Spectral\") + \n",
    "  scale_fill_manual(values = hcl.colors(palette = \"Spectral\", n = 16)) +\n",
    "  theme_classic() + \n",
    "  theme(legend.position = \"right\") +\n",
    "  coord_flip()\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Cell_Proportions_group_Propeller.png\", \n",
    "    width=800, \n",
    "    height=400)\n",
    "plot\n",
    "dev.off()\n",
    "test_differences_propellor <- function(test_variable) {\n",
    "  design <- model.matrix(as.formula(paste(\"~ 0 + \", test_variable, \" + AGE\")), \n",
    "                         data = cell_proportions)\n",
    "  contrasts <- c(-1,1,0)\n",
    "  \n",
    "  # Determine & Conduct Test\n",
    "  if ((ncol(design) - 1) == 2) {\n",
    "    # two groups\n",
    "    # mycontr <- makeContrasts(GROUPPD-GROUPHC, levels=design)\n",
    "    test_type <- \"t-test\"\n",
    "    props_test <- propeller.ttest(prop.list = props, \n",
    "                    design = design, \n",
    "                    contrasts = contrasts, \n",
    "                    robust = TRUE, \n",
    "                    trend = FALSE, \n",
    "                    sort = FALSE)\n",
    "  } else if ((ncol(design) - 1) > 2) {\n",
    "    # More than two groups\n",
    "    # mycontr <- makeContrasts(GROUPPD-GROUPHC, levels=design)\n",
    "    test_type <- \"ANOVA\"\n",
    "    props_test <- propeller.anova(prop.list = props,\n",
    "                                  design = design,\n",
    "                                  coef = c(1,2,3),\n",
    "                                  robust = TRUE,\n",
    "                                  trend = FALSE,\n",
    "                                  sort = FALSE)\n",
    "  } else {\n",
    "    stop(\"You have less than 2 comparison groups -- propeller won't work.\")\n",
    "  }\n",
    "  # Return a named list with the stats table & the test type.\n",
    "  list(props_test = props_test,\n",
    "       test_type = test_type) %>%\n",
    "    return()\n",
    "  # return(props_test)\n",
    "}\n",
    "#----- Run Propeller\n",
    "props_test <- test_differences_propellor(\"GROUP\")\n",
    "#----- Cell Proportions Table\n",
    "kable(props_test$props_test, caption = paste(\"Proportion of Cells per Cluster (Propeller\", \n",
    "                                             props_test$test_type, \"test)\"), \n",
    "      format.args = list(big.mark = \",\")) %>% \n",
    "  kable_styling(c(\"striped\", \"bordered\"))\n",
    "#----- Save Results\n",
    "write_delim(props_test$props_test, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Cell_Proportions_Differences_Group_Propeller.txt\"), \n",
    "  delim = \"\\t\")\n",
    "#---------- Data\n",
    "#----- Remove Data\n",
    "rm(cell_proportions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ae59652",
   "metadata": {
    "eval": false,
    "lines_to_next_cell": 2,
    "name": "Seurat_Cell_Proportions_Speckle_Tutorial"
   },
   "outputs": [],
   "source": [
    "#----- Tutorial\n",
    "library(speckle)\n",
    "library(SingleCellExperiment)\n",
    "library(CellBench)\n",
    "library(limma)\n",
    "library(ggplot2)\n",
    "library(scater)\n",
    "library(patchwork)\n",
    "library(edgeR)\n",
    "library(AnnotationDbi)\n",
    "library(org.Hs.eg.db)\n",
    "# browseVignettes(\"speckle\")\n",
    "# Single Cell data\n",
    "sc_data <- load_sc_data()\n",
    "# Common Genes\n",
    "commongenes1 <- rownames(sc_data$sc_dropseq)[rownames(sc_data$sc_dropseq) %in% \n",
    "                                               rownames(sc_data$sc_celseq)]\n",
    "commongenes2 <-  commongenes1[commongenes1 %in% rownames(sc_data$sc_10x)]\n",
    "sce_10x <- sc_data$sc_10x[commongenes2,]\n",
    "sce_celseq <- sc_data$sc_celseq[commongenes2,] \n",
    "sce_dropseq <- sc_data$sc_dropseq[commongenes2,] \n",
    "# Bootstrap additional samples\n",
    "set.seed(10)\n",
    "boot.10x <- sample(seq_len(ncol(sce_10x)), replace=TRUE)\n",
    "boot.celseq <- sample(seq_len(ncol(sce_celseq)), replace=TRUE)\n",
    "boot.dropseq <- sample(seq_len(ncol(sce_dropseq)), replace=TRUE)\n",
    "sce_10x_rep2 <- sce_10x[,boot.10x]\n",
    "sce_celseq_rep2 <- sce_celseq[,boot.celseq]\n",
    "sce_dropseq_rep2 <- sce_dropseq[,boot.dropseq]\n",
    "# Combine all SingleCellExperiment objects\n",
    "allcounts <- cbind(counts(sce_10x),counts(sce_10x_rep2), \n",
    "                   counts(sce_celseq), counts(sce_celseq_rep2),\n",
    "                   counts(sce_dropseq), counts(sce_dropseq_rep2))\n",
    "sce_all <- SingleCellExperiment(assays = list(counts = allcounts))\n",
    "sce_all$sample <- rep(c(\"S1\",\"S2\",\"S3\",\"S4\",\"S5\",\"S6\"), \n",
    "                      c(ncol(sce_10x), ncol(sce_10x_rep2), \n",
    "                        ncol(sce_celseq), ncol(sce_celseq_rep2), \n",
    "                        ncol(sce_dropseq), ncol(sce_dropseq_rep2)))\n",
    "sce_all$group <- rep(c(\"10x\",\"celseq\",\"dropseq\"),\n",
    "                     c(2*ncol(sce_10x),2*ncol(sce_celseq),2*ncol(sce_dropseq)))\n",
    "sce_all$cluster <- c(sce_10x$cell_line, sce_10x_rep2$cell_line, \n",
    "                     sce_celseq$cell_line, sce_celseq_rep2$cell_line, \n",
    "                     sce_dropseq$cell_line, sce_dropseq_rep2$cell_line)\n",
    "rm(sc_data, commongenes1, commongenes2, sce_10x, sce_celseq, sce_dropseq, \n",
    "   boot.10x, boot.celseq, boot.dropseq, sce_10x_rep2, sce_celseq_rep2, sce_dropseq_rep2,\n",
    "   allcounts)\n",
    "sce_all <- scater::logNormCounts(sce_all)\n",
    "sce_all <- scater::runPCA(sce_all)\n",
    "sce_all <- scater::runUMAP(sce_all)\n",
    "# Test for differences in cell line proportions in the three technologies\n",
    "propeller(clusters = sce_all$cluster, \n",
    "          sample = sce_all$sample, \n",
    "          group = sce_all$group,\n",
    "          robust = TRUE,\n",
    "          transform = \"logit\")\n",
    "propeller(clusters = sce_all$cluster, \n",
    "          sample = sce_all$sample, \n",
    "          group = sce_all$group,\n",
    "          robust = TRUE,\n",
    "          transform = \"asin\")\n",
    "# Fitting linear models using the transformed proportions directly\n",
    "plotCellTypeProps(sce_all)\n",
    "props <- getTransformedProps(sce_all$cluster, sce_all$sample, transform=\"logit\")\n",
    "barplot(props$Proportions, col = c(\"orange\",\"purple\",\"dark green\"),legend=TRUE, \n",
    "        ylab=\"Proportions\")\n",
    "group <- rep(c(\"10x\",\"celseq\",\"dropseq\"),each=2)\n",
    "pair <- rep(c(1,2),3)\n",
    "design <- model.matrix(~ 0 + group + pair)\n",
    "propeller.anova(prop.list=props, \n",
    "                design=design, \n",
    "                coef = c(1,2,3), \n",
    "                robust=TRUE, \n",
    "                trend=FALSE, \n",
    "                sort=TRUE)\n",
    "mycontr <- makeContrasts(group10x-groupdropseq, levels=design)\n",
    "propeller.ttest(prop.list = props, \n",
    "                design = design, \n",
    "                contrasts = mycontr, \n",
    "                robust=TRUE, \n",
    "                trend=FALSE, \n",
    "                sort=TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dfceccd",
   "metadata": {},
   "source": [
    "## Seurat Conserved Markers\n",
    "  \n",
    "Identification of all markers for each cluster  \n",
    "This type of analysis is typically recommended for when evaluating a single  \n",
    "sample group/condition. With the FindAllMarkers() function we are comparing each  \n",
    "cluster against all other clusters to identify potential marker genes. The cells  \n",
    "in each cluster are treated as replicates, and essentially a differential  \n",
    "expression analysis is performed with some statistical test. Also, by default  \n",
    "this function will return to you genes that exhibit both positive and negative  \n",
    "expression changes. Typically, we add an argument only.pos to opt for keeping \n",
    "only the positive changes.  \n",
    "  \n",
    "Identification of conserved markers in all conditions  \n",
    "Since we have samples representing different conditions in our dataset, our best  \n",
    "option is to find conserved markers. This function internally separates out  \n",
    "cells by sample group/condition, and then performs differential gene expression  \n",
    "testing for a single specified cluster against all other clusters (or a second  \n",
    "cluster, if specified). Gene-level p-values are computed for each condition and  \n",
    "then combined across groups using meta-analysis methods from the MetaDE R  \n",
    "package. Before we start our marker identification we will explicitly set our  \n",
    "default assay, we want to use the original counts and not the integrated data.  \n",
    "DefaultAssay(seurat_anno) <- \"RNA\"  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c52dd737",
   "metadata": {
    "eval": false,
    "name": "Seurat_Conserved_Markers"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Conserved_Markers\n",
    "# RNA stores the original and normalized counts for finding markers\n",
    "# DefaultAssay(seurat_anno) <- \"RNA\"\n",
    "# Ensembl & Symbol Gene IDs \n",
    "#----- Cell Types\n",
    "cell_types <- seurat_anno@meta.data %>% \n",
    "  arrange(cluster_annotation) %>% \n",
    "  dplyr::select(cluster_annotation) %>% \n",
    "  unique() %>% \n",
    "  pull(cluster_annotation)\n",
    "#----- Conserved Markers across sample groups\n",
    "get_conserved <- function(seurat_obj, cell_type){\n",
    "  FindConservedMarkers(object = seurat_obj,\n",
    "                       ident.1 = cell_type, \n",
    "                       grouping.var = \"GROUP\", \n",
    "                       only.pos = TRUE, \n",
    "                       min.diff.pct = 0.1, \n",
    "                       min.pct = 0.1, \n",
    "                       logfc.threshold = 0.25) %>% \n",
    "    rownames_to_column(var = \"Gene_Symbol\") %>% \n",
    "    left_join(Gene_Emsembl_Symbol, by=\"Gene_Symbol\") %>% \n",
    "    dplyr::mutate(Celltype = cell_type)\n",
    "}\n",
    "Conserved_Markers <- mclapply(cell_types, function(i) {\n",
    "  print(i)\n",
    "  get_conserved(seurat_obj = seurat_anno,\n",
    "                cell_type = i) \n",
    "}) %>% bind_rows()\n",
    "#---- Save Results\n",
    "write_delim(Conserved_Markers, \n",
    "            paste0(\"seurat_results/\", NameProject, \"_Conserved_Markers.txt\"), \n",
    "            delim = \"\\t\")\n",
    "\n",
    "#---------- Conserved Markers Pathways\n",
    "#----- Cell Types\n",
    "cell_types <- seurat_anno@meta.data %>% \n",
    "  dplyr::arrange(cluster_annotation) %>% \n",
    "  dplyr::select(cluster_annotation) %>% \n",
    "  unique() %>% \n",
    "  pull(cluster_annotation)\n",
    "#----- Conserved Markers Pathways\n",
    "Conserved_Markers_Pathways <- lapply(cell_types, function(cell_type) {\n",
    "  genes <- Conserved_Markers %>% \n",
    "             dplyr::filter(Celltype == cell_type) %>% \n",
    "             dplyr::pull(Gene_Symbol)\n",
    "  genes_go <- enrichGO(gene = genes,\n",
    "                       OrgDb = \"org.Hs.eg.db\",\n",
    "                       ont =\"BP\",\n",
    "                       keyType = \"SYMBOL\",\n",
    "                       pAdjustMethod = \"BH\")\n",
    "  return(genes_go@result %>% mutate(Celltype = cell_type))\n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(Conserved_Markers_Pathways, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Conserved_Markers_Pathways_GOBP.txt\"), \n",
    "  delim = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74b4a8c2",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Seurat_Conserved_Markers_Figures"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Conserved_Markers_Figures\n",
    "Conserved_Markers <- \n",
    "  read_delim(paste0(\"seurat_results/\", NameProject, \"_Conserved_Markers.txt\"), \n",
    "  delim = \"\\t\", escape_double = FALSE, \n",
    "  trim_ws = TRUE)\n",
    "#----- Conserved Markers Heatmap\n",
    "# Idents(seurat_anno) <- seurat_anno$cluster_annotation\n",
    "plot <- DoHeatmap(object = seurat_anno, \n",
    "          features = Conserved_Markers %>% \n",
    "                       dplyr::group_by(Celltype) %>% \n",
    "                       dplyr::mutate(avg_log2FC = (HC_avg_log2FC + \n",
    "                                                   PD_avg_log2FC)/2) %>% \n",
    "                       top_n(10, avg_log2FC) %>% \n",
    "                       dplyr::pull(Gene_Symbol),\n",
    "          label = TRUE)\n",
    "plot\n",
    "png(file=\"seurat_results/figures/Seurat_Conserved_Markers_heatmap.png\", \n",
    "    width=1000, \n",
    "    height=1000)\n",
    "plot\n",
    "dev.off()\n",
    "\n",
    "\n",
    "#---------- Conserved Markers Pathways Figure\n",
    "Conserved_Markers_Pathways <- \n",
    "  read_delim(paste0(\"seurat_results/\", NameProject, \"_Conserved_Markers_Pathways_GOBP.txt\"), \n",
    "  delim = \"\\t\", escape_double = FALSE, \n",
    "  trim_ws = TRUE)\n",
    "#----- Cell Types\n",
    "cell_types <- seurat_anno@meta.data %>% \n",
    "  dplyr::arrange(cluster_annotation) %>% \n",
    "  dplyr::select(cluster_annotation) %>% \n",
    "  unique() %>% \n",
    "  pull(cluster_annotation)\n",
    "#----- Conserved Markers Pathways\n",
    "for (cell_type in cell_types){\n",
    "  genes <- Conserved_Markers %>% \n",
    "             dplyr::filter(Celltype == cell_type) %>% \n",
    "             dplyr::pull(Gene_Symbol)\n",
    "  genes_go <- enrichGO(gene = genes,\n",
    "                       OrgDb = \"org.Hs.eg.db\",\n",
    "                       ont =\"BP\",\n",
    "                       keyType = \"SYMBOL\",\n",
    "                       pAdjustMethod = \"BH\")\n",
    "  plot <- enrichplot::dotplot(genes_go, showCategory=20) + \n",
    "            ggtitle(paste0(cell_type, \" - GO:Biological Process\")) + \n",
    "            theme(plot.title = element_text(hjust = 0.5, \n",
    "                                            face = \"bold\"))\n",
    "  print(plot)\n",
    "  png(file=\n",
    "    paste0(\"seurat_results/figures/Seurat_Conserved_Markers_Pathways_GOBP_\", \n",
    "           cell_type, \".png\"), \n",
    "      width=1000, \n",
    "      height=600)\n",
    "  print(plot)\n",
    "  dev.off()\n",
    "  plot <- genes_go %>% pairwise_termsim() %>% \n",
    "            enrichplot::emapplot(showCategory=40) + \n",
    "              ggtitle(paste0(cell_type, \" - GO:Biological Process\")) + \n",
    "              theme(plot.title = element_text(hjust = 0.5, \n",
    "                                              face = \"bold\"))\n",
    "  print(plot)\n",
    "  png(file=\n",
    "    paste0(\"seurat_results/figures/Seurat_Conserved_Markers_Pathways_GOBPenrich_\", \n",
    "           cell_type, \".png\"), \n",
    "      width=1000, \n",
    "      height=1000)\n",
    "  print(plot)\n",
    "  dev.off()\n",
    "}  \n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f04a4565",
   "metadata": {},
   "source": [
    "## Seurat Differential Expression\n",
    "\n",
    "Find DEG between PD and HC per cell_type  \n",
    "Using 0.25 logFC  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "298e9ad2",
   "metadata": {
    "eval": false,
    "name": "Seurat_Differential_Expression"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Differential_Expression\n",
    "#----- RNA Assay stores the original and normalized counts\n",
    "# Original counts : seurat_anno@assays[[\"RNA\"]]@counts\n",
    "# Normalized counts : seurat_anno@assays[[\"RNA\"]]@data\n",
    "DefaultAssay(seurat_anno) <- \"RNA\"\n",
    "#----- Group Celltype ID\n",
    "# make this the Idents()\n",
    "seurat_anno$group_cluster <- paste(seurat_anno$GROUP, \n",
    "                                   seurat_anno$cluster_annotation, \n",
    "                                   sep = \"_\")\n",
    "Idents(seurat_anno) <- seurat_anno$group_cluster\n",
    "\n",
    "#---------- Differential Expressed Genes\n",
    "#----- Cell Types\n",
    "cell_types <- seurat_anno@meta.data %>% \n",
    "  dplyr::arrange(cluster_annotation) %>% \n",
    "  dplyr::select(cluster_annotation) %>% \n",
    "  unique() %>% \n",
    "  pull(cluster_annotation)\n",
    "#----- Cell_Type Differential Expressed Genes\n",
    "get_deg <- function(seurat_obj, cell_type, ident1, ident2){\n",
    "  FindMarkers(seurat_obj, \n",
    "              ident.1 = paste(ident1, cell_type, sep = \"_\"), \n",
    "              ident.2 = paste(ident2, cell_type, sep = \"_\"), \n",
    "              test.use = \"MAST\", \n",
    "              latent.vars = \"AGE\", \n",
    "              logfc.threshold = 0, \n",
    "              min.pct = 0,\n",
    "              min.diff.pct = -Inf, \n",
    "              min.cells.feature = 10, \n",
    "              min.cells.group = 10, \n",
    "              verbose = FALSE) %>% \n",
    "    rownames_to_column(var = \"Gene_Symbol\") %>% \n",
    "    left_join(Gene_Emsembl_Symbol, by=\"Gene_Symbol\") %>% \n",
    "    mutate(Comparison = paste(ident1, \"vs\", ident2, cell_type, sep = \"_\"))\n",
    "}\n",
    "Differential_Expression <- lapply(cell_types, function(cell_type) {\n",
    "  tryCatch(get_deg(seurat_obj = seurat_anno,\n",
    "                   cell_type = cell_type,\n",
    "                   ident1 = \"PD\",\n",
    "                   ident2 = \"HC\"),\n",
    "           error = function(e) \n",
    "             # empty dataframe with same column headers & types...\n",
    "             data.frame(Gene_Symbol = character(),\n",
    "                        p_val = double(),\n",
    "                        avg_log2FC = double(),\n",
    "                        pct.1 = double(),\n",
    "                        pct.2 = double(),\n",
    "                        p_val_adj = double(),\n",
    "                        Gene_Emsembl = character(),\n",
    "                        Comparison = character(),\n",
    "                        stringsAsFactors = FALSE))\n",
    "}) %>% bind_rows()\n",
    "#----- save the results\n",
    "write_delim(Differential_Expression, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_PD_vs_HC_AGE_MAST.txt\"), \n",
    "  delim = \"\\t\")\n",
    "saveRDS(Differential_Expression, file = \"seurat_results/differential_expression.rds\")\n",
    "#----- Number of DEGs\n",
    "Differential_Expression_count <- Differential_Expression %>% \n",
    "  mutate(Significant = case_when(\n",
    "           p_val_adj <= 0.05 & avg_log2FC >= 0.25 ~ 'Upregulated',\n",
    "           p_val_adj <= 0.05 & avg_log2FC <= -0.25 ~ 'Downregulated',\n",
    "           TRUE ~ 'Not_Significant')) %>% \n",
    "  dplyr::count(Comparison, Significant) %>% \n",
    "  dplyr::rename(Count = n) %>% \n",
    "  tidyr::spread(Significant, Count)\n",
    "#----- save the results\n",
    "write_delim(Differential_Expression_count, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_PD_vs_HC_AGE_MAST_count.txt\"), \n",
    "  delim = \"\\t\")\n",
    "\n",
    "\n",
    "#---------- Differential Expressed Genes Pathways\n",
    "#----- Comparisons\n",
    "comparisons <- Differential_Expression$Comparison %>% \n",
    "                 unique()\n",
    "#----- Pathways GSEA GOBP\n",
    "Differential_Expression_GOBP <- lapply(comparisons, function(comparison) {\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_go <- gseGO(geneList = genelist, \n",
    "                    ont =\"BP\", \n",
    "                    OrgDb = get(\"org.Hs.eg.db\"),\n",
    "                    keyType = \"SYMBOL\", \n",
    "                    minGSSize = 10, \n",
    "                    maxGSSize = 500, \n",
    "                    pvalueCutoff = 0.05, \n",
    "                    pAdjustMethod = \"BH\",\n",
    "                    verbose = TRUE)\n",
    "  return(genes_go@result %>% mutate(Comparison = comparison))\n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(Differential_Expression_GOBP, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_PD_vs_HC_AGE_MAST_Pathways_GSEA_GOBP.txt\"), \n",
    "  delim = \"\\t\")\n",
    "#----- Reactome\n",
    "reactome <- msigdbr(species = \"Homo sapiens\", \n",
    "                    category = \"C2\", \n",
    "                    subcategory = \"CP:REACTOME\") %>% \n",
    "              dplyr::select(gs_name, gene_symbol)\n",
    "#----- Pathways GSEA REACTOME\n",
    "Differential_Expression_REACTOME <- lapply(comparisons, function(comparison) {\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_gsea <- GSEA(geneList = genelist,\n",
    "                     minGSSize = 10,\n",
    "                     maxGSSize = 500,\n",
    "                     pvalueCutoff = 0.05,\n",
    "                     pAdjustMethod = \"BH\",\n",
    "                     TERM2GENE = reactome, \n",
    "                     by = \"fgsea\",\n",
    "                     verbose = TRUE)\n",
    "  return(genes_gsea@result %>% mutate(Comparison = comparison))\n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(Differential_Expression_REACTOME, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_PD_vs_HC_AGE_MAST_Pathways_GSEA_REACTOME.txt\"), \n",
    "  delim = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "882d0e47",
   "metadata": {
    "lines_to_next_cell": 2,
    "name": "Seurat_Differential_Expression_Figures"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Differential_Expression_Figures\n",
    "Differential_Expression <- \n",
    "  read_delim(paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_PD_vs_HC_AGE_MAST.txt\"), \n",
    "  delim = \"\\t\", escape_double = FALSE, \n",
    "  trim_ws = TRUE)\n",
    "\n",
    "#----- Differential_Expression_count\n",
    "Differential_Expression_count <- \n",
    "  read_delim(paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_PD_vs_HC_AGE_MAST_count.txt\"), \n",
    "  delim = \"\\t\", escape_double = FALSE, \n",
    "  trim_ws = TRUE)\n",
    "Differential_Expression_count %>% \n",
    "  kable(caption=\"Number of Differentially Expressed Genes\", \n",
    "        format.args = list(big.mark = \",\")) %>% \n",
    "  kable_styling(c(\"striped\", \"bordered\"))\n",
    "\n",
    "\n",
    "#---------- Differential Expressed Genes Volcano\n",
    "#----- Comparisons\n",
    "comparisons <- Differential_Expression$Comparison %>% \n",
    "                 unique()\n",
    "#----- Differential Expressed Genes Volcano\n",
    "for (comparison in comparisons){\n",
    "  plot <- Differential_Expression %>% \n",
    "            dplyr::filter(Comparison == comparison) %>%\n",
    "            dplyr::select(Comparison, \n",
    "                          Gene_Symbol, \n",
    "                          avg_log2FC, \n",
    "                          p_val,\n",
    "                          p_val_adj) %>% \n",
    "            mutate(Significant = case_when(\n",
    "                     p_val_adj <= 0.05 & avg_log2FC >= 0.25 ~ 'Upregulated',\n",
    "                     p_val_adj <= 0.05 & avg_log2FC <= -0.25 ~ 'Downregulated',\n",
    "                     TRUE ~ 'Not_Significant')) %>% \n",
    "            mutate(Label = case_when(\n",
    "                     Significant == \"Upregulated\" ~ Gene_Symbol,\n",
    "                     Significant == \"Downregulated\" ~ Gene_Symbol,\n",
    "                     TRUE ~ NA_character_)) %>% \n",
    "            ggplot(mapping = aes(x=avg_log2FC, \n",
    "                   y = -log10(p_val), \n",
    "                   color = Significant,\n",
    "                   label = Label)) + \n",
    "              geom_point(alpha = 0.5, stroke=0, size=2) + \n",
    "              geom_text_repel() +\n",
    "              ggtitle(comparison) + \n",
    "              xlab(\"log2 Fold Change\") +\n",
    "              ylab(\"-log10(P Value)\") + \n",
    "              theme_classic() + \n",
    "              theme(plot.title = element_text(hjust = 0.5, \n",
    "                                              face = \"bold\")) + \n",
    "              scale_colour_manual(values = c('Blue','Gray','Red')) + \n",
    "              geom_vline(xintercept = c(-0.25,0.25), \n",
    "                         linetype = \"dotted\")\n",
    "  print(plot)\n",
    "  # png(file = paste(\"seurat_results/figures/Seurat_Differential_Expression\", comparison, \"AGE_MAST_volcano.png\", sep = \"_\"),\n",
    "  #   width = 600*2, \n",
    "  #   height = 400*2,\n",
    "  #   pointsize = 6)\n",
    "  # print(plot)\n",
    "  # dev.off()\n",
    "  ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\", comparison, \"AGE_MAST_volcano.png\", sep = \"_\"),\n",
    "         plot = plot,\n",
    "         width = 600*2,\n",
    "         height = 400*2,\n",
    "         dpi = 150,\n",
    "         units = \"px\")\n",
    "}\n",
    "\n",
    "\n",
    "#---------- Differential Expressed Genes Pathways\n",
    "#----- Comparisons\n",
    "comparisons <- Differential_Expression$Comparison %>% \n",
    "                 unique()\n",
    "#----- Pathways GSEA GOBP\n",
    "for (comparison in comparisons){\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_go <- gseGO(geneList = genelist, \n",
    "                    ont =\"BP\", \n",
    "                    OrgDb = get(\"org.Hs.eg.db\"),\n",
    "                    keyType = \"SYMBOL\", \n",
    "                    minGSSize = 10, \n",
    "                    maxGSSize = 500, \n",
    "                    pvalueCutoff = 0.05, \n",
    "                    pAdjustMethod = \"BH\",\n",
    "                    verbose = TRUE)\n",
    "  if (nrow(genes_go@result) == 0) {\n",
    "     warning(paste0(\"No significant GOBP genes for: \", comparison, \"\\nSkipping to next comparison...\"))\n",
    "   } else {\n",
    "     plot <- enrichplot::dotplot(object = genes_go, \n",
    "                                  showCategory = 10, \n",
    "                                  split = \".sign\") + \n",
    "                facet_grid(.~.sign) + \n",
    "                ggtitle(paste0(comparison, \" - GO:Biological Process\")) + \n",
    "                theme(plot.title = element_text(hjust = 0.5, \n",
    "                                                face = \"bold\"))\n",
    "      print(plot)\n",
    "      png(file=\n",
    "        paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "               comparison, \"AGE_MAST_Pathways_gseaGOBP.png\", sep = \"_\"), \n",
    "          width=1000, \n",
    "          height=600)\n",
    "      print(plot)\n",
    "      dev.off()\n",
    "   } \n",
    "  \n",
    "  #----- Split GOBP Pathway Map by Activated / Suppressed\n",
    "  # (otherwise the activated & suppressed are shown together -- misleading)\n",
    "  genes_go_emapplot <- genes_go\n",
    "  genes_go_emapplot@result <- genes_go_emapplot@result %>%\n",
    "    mutate(pathway_direction = case_when(NES > 0 ~ \"activated\",\n",
    "                                         NES < 0 ~ \"suppressed\"))\n",
    "  genes_go_emapplot_activated <- dplyr::filter(genes_go_emapplot, \n",
    "                                               pathway_direction == \"activated\")\n",
    "  genes_go_emapplot_suppressed <- dplyr::filter(genes_go_emapplot, \n",
    "                                               pathway_direction == \"suppressed\")\n",
    "  \n",
    "  # plot activated\n",
    "  if (nrow(genes_go_emapplot_activated@result) == 0) {\n",
    "     warning(paste0(\"No significant activated GOBP genes for: \", \n",
    "                    comparison, \"\\nSkipping to next comparison...\"))\n",
    "    plot_activated <- ggplot() + theme_void()\n",
    "  } else {\n",
    "    genes_go_emapplot_activated <- pairwise_termsim(x = genes_go_emapplot_activated)\n",
    "    plot_activated <- genes_go_emapplot_activated %>% \n",
    "              enrichplot::emapplot(showCategory=40) +\n",
    "              theme(plot.title = element_text(hjust = 0.5,\n",
    "                                              face = \"bold\"))\n",
    "    print(plot_activated)\n",
    "  }\n",
    "  \n",
    "  # plot suppressed\n",
    "  if (nrow(genes_go_emapplot_suppressed@result) == 0) {\n",
    "     warning(paste0(\"No significant suppressed GOBP genes for: \", \n",
    "                    comparison, \"\\nSkipping to next comparison...\"))\n",
    "    plot_suppressed <- ggplot() + theme_void()\n",
    "  } else {\n",
    "    genes_go_emapplot_suppressed <- pairwise_termsim(x = genes_go_emapplot_suppressed)\n",
    "    plot_suppressed <- genes_go_emapplot_suppressed %>% \n",
    "              enrichplot::emapplot(showCategory=40) +\n",
    "                theme(plot.title = element_text(hjust = 0.5,\n",
    "                                                face = \"bold\"))\n",
    "    print(plot_suppressed)\n",
    "  } \n",
    "  \n",
    "  # combine & save plot\n",
    "  sub_plot_labels <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = \"ACTIVATED\",\n",
    "                                                                       size = 18,\n",
    "                                                                       hjust = 0.5,\n",
    "                                                                       fontface = \"bold\"),\n",
    "                                        ggdraw() + cowplot::draw_label(label = \"SUPPRESSED\",\n",
    "                                                                       size = 18,\n",
    "                                                                       hjust = 0.5,\n",
    "                                                                       fontface = \"bold\"),\n",
    "                                        ncol = 2)\n",
    "  \n",
    "  \n",
    "  plot <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = paste0(comparison, \n",
    "                                                     \" - GO:Biological Process\"),\n",
    "                                                     size = 24,\n",
    "                                                     hjust = 0.5,\n",
    "                                                     fontface = \"bold\"),\n",
    "                             sub_plot_labels,\n",
    "                             plot_grid(plot_activated, plot_suppressed),\n",
    "                             ncol = 1, rel_heights = c(0.1, .05, 1))\n",
    "\n",
    "  plot\n",
    "  \n",
    "  png(file=\n",
    "    paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "           comparison, \"AGE_MAST_Pathways_gseaGOBPenrich.png\", sep = \"_\"), \n",
    "      width=1400, \n",
    "      height=800)\n",
    "  print(plot)\n",
    "  dev.off()\n",
    "  #----- pathExplore: High-Level GSEA Clusters (GO:BP)\n",
    "  tryCatch({\n",
    "    # fix here: https://github.com/tidyverse/ggplot2/issues/2799\n",
    "    cf <- coord_fixed()\n",
    "    cf$default <- TRUE\n",
    "    plot <- enrichmentNetwork(genes_go@result, drawEllipses = TRUE, fontSize = 2.5) +\n",
    "      ggtitle(paste0(comparison, \" - GO:Biological Process\")) +\n",
    "      theme(plot.title = element_text(hjust = 0.5, face = \"bold\"),\n",
    "            plot.margin = margin(l = 6, t = 12, r = 3, unit = \"pt\")) +\n",
    "      cf +\n",
    "      coord_fixed(clip = \"off\")\n",
    "    print(plot)\n",
    "    ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\",\n",
    "                            comparison, \"AGE_MAST_Pathways_gseaGOBPenrich_pathExplore.png\", \n",
    "                            sep = \"_\"),\n",
    "           plot = plot,\n",
    "           width = 600*2,\n",
    "           height = 400*2,\n",
    "           dpi = 150,\n",
    "           units = \"px\")\n",
    "  },\n",
    "    error=function(e){cat(\"ERROR :\",conditionMessage(e), \"\\n\")})\n",
    "}\n",
    "#----- Reactome\n",
    "reactome <- msigdbr(species = \"Homo sapiens\", \n",
    "                    category = \"C2\", \n",
    "                    subcategory = \"CP:REACTOME\") %>% \n",
    "              dplyr::select(gs_name, gene_symbol)\n",
    "#----- Pathways GSEA REACTOME\n",
    "for (comparison in comparisons){\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_gsea <- GSEA(geneList = genelist,\n",
    "                     minGSSize = 10,\n",
    "                     maxGSSize = 500,\n",
    "                     pvalueCutoff = 0.05,\n",
    "                     pAdjustMethod = \"BH\",\n",
    "                     TERM2GENE = reactome, \n",
    "                     by = \"fgsea\",\n",
    "                     verbose = TRUE)\n",
    "  if (nrow(genes_gsea@result) == 0) {\n",
    "    warning(paste0(\"No significant REACTOME genes for: \", comparison, \"\\nSkipping to next comparison...\"))\n",
    "   } else {\n",
    "    plot <- enrichplot::dotplot(object = genes_gsea,#[genes_gsea@result$ID %in% c(\"REACTOME_SENESCENCE_ASSOCIATED_SECRETORY_PHENOTYPE_SASP\", \"REACTOME_CELLULAR_RESPONSE_TO_HEAT_STRESS\", \"REACTOME_PROGRAMMED_CELL_DEATH\", \"REACTOME_NUCLEOTIDE_EXCISION_REPAIR\", \"REACTOME_LAMININ_INTERACTIONS\", \"REACTOME_EXTRACELLULAR_MATRIX_ORGANIZATION\"), asis = TRUE], #REACTOME_SIGNALING_BY_INTERLEUKINS, REACTOME_RRNA_PROCESSING, REACTOME_SIGNALING_BY_NOTCH4\n",
    "                                showCategory = 10, \n",
    "                                split = \".sign\") + \n",
    "              facet_grid(.~.sign) + \n",
    "              ggtitle(paste0(comparison, \" - GSEA Reactome\")) + \n",
    "              theme(plot.title = element_text(hjust = 0.5, \n",
    "                                              face = \"bold\"))\n",
    "    print(plot)\n",
    "    png(file=\n",
    "      paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "             comparison, \"AGE_MAST_Pathways_gseaREACTOME.png\", sep = \"_\"), \n",
    "        width=1000, \n",
    "        height=600)\n",
    "    print(plot)\n",
    "    dev.off()\n",
    "    ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "             comparison, \"AGE_MAST_Pathways_gseaREACTOME_quality.png\", sep = \"_\"), \n",
    "       plot = plot,\n",
    "       width = 1000*2,\n",
    "       height = 600*2,\n",
    "       dpi = 200,\n",
    "       units = \"px\")\n",
    "    \n",
    "    #----- Split REACTOME Pathway Map by Activated / Suppressed\n",
    "    # (otherwise the activated & suppressed are shown together -- misleading)\n",
    "    genes_gsea_emapplot <- genes_gsea\n",
    "    genes_gsea_emapplot@result <- genes_gsea_emapplot@result %>%\n",
    "      mutate(pathway_direction = case_when(NES > 0 ~ \"activated\",\n",
    "                                           NES < 0 ~ \"suppressed\"))\n",
    "    genes_gsea_emapplot_activated <- dplyr::filter(genes_gsea_emapplot, \n",
    "                                                 pathway_direction == \"activated\")\n",
    "    genes_gsea_emapplot_suppressed <- dplyr::filter(genes_gsea_emapplot, \n",
    "                                                 pathway_direction == \"suppressed\")\n",
    "    \n",
    "    # plot activated\n",
    "    if (nrow(genes_gsea_emapplot_activated@result) == 0) {\n",
    "       warning(paste0(\"No significant activated REACTOME genes for: \", \n",
    "                      comparison, \"\\nSkipping to next comparison...\"))\n",
    "      plot_activated <- ggplot() + theme_void()\n",
    "    } else {\n",
    "      genes_gsea_emapplot_activated <- pairwise_termsim(x = genes_gsea_emapplot_activated)\n",
    "      plot_activated <- genes_gsea_emapplot_activated %>% \n",
    "                enrichplot::emapplot(showCategory=40) +\n",
    "                theme(plot.title = element_text(hjust = 0.5,\n",
    "                                                face = \"bold\"))\n",
    "      print(plot_activated)\n",
    "    }\n",
    "    \n",
    "    # plot suppressed\n",
    "    if (nrow(genes_gsea_emapplot_suppressed@result) == 0) {\n",
    "       warning(paste0(\"No significant suppressed REACTOME genes for: \", \n",
    "                      comparison, \"\\nSkipping to next comparison...\"))\n",
    "      plot_suppressed <- ggplot() + theme_void()\n",
    "    } else {\n",
    "      genes_gsea_emapplot_suppressed <- pairwise_termsim(x = genes_gsea_emapplot_suppressed)\n",
    "      plot_suppressed <- genes_gsea_emapplot_suppressed %>% \n",
    "                enrichplot::emapplot(showCategory=40) +\n",
    "                  theme(plot.title = element_text(hjust = 0.5,\n",
    "                                                  face = \"bold\"))\n",
    "      print(plot_suppressed)\n",
    "    } \n",
    "    \n",
    "    # combine & save plot\n",
    "    sub_plot_labels <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = \"ACTIVATED\",\n",
    "                                                                         size = 18,\n",
    "                                                                         hjust = 0.5,\n",
    "                                                                         fontface = \"bold\"),\n",
    "                                          ggdraw() + cowplot::draw_label(label = \"SUPPRESSED\",\n",
    "                                                                         size = 18,\n",
    "                                                                         hjust = 0.5,\n",
    "                                                                         fontface = \"bold\"),\n",
    "                                          ncol = 2)\n",
    "    \n",
    "    \n",
    "    plot <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = paste0(comparison, \n",
    "                                                       \" - GSEA Reactome\"),\n",
    "                                                       size = 24,\n",
    "                                                       hjust = 0.5,\n",
    "                                                       fontface = \"bold\"),\n",
    "                               sub_plot_labels,\n",
    "                               plot_grid(plot_activated, plot_suppressed),\n",
    "                               ncol = 1, rel_heights = c(0.1, .05, 1))\n",
    "    print(plot)\n",
    "    png(file=\n",
    "      paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "             comparison, \"AGE_MAST_Pathways_gseaREACTOMEenrich.png\", sep = \"_\"), \n",
    "        width=1400, \n",
    "        height=800)\n",
    "    print(plot)\n",
    "    dev.off()\n",
    "    #----- pathExplore: High-Level GSEA Clusters (GSEA Reactome)\n",
    "    tryCatch({\n",
    "      # fix here: https://github.com/tidyverse/ggplot2/issues/2799\n",
    "      cf <- coord_fixed()\n",
    "      cf$default <- TRUE\n",
    "      plot <- pathExplore::enrichmentNetwork(genes_gsea@result, drawEllipses = TRUE, fontSize = 2.5) +\n",
    "        ggtitle(paste0(comparison, \" - GSEA Reactome\")) +\n",
    "        theme(plot.title = element_text(hjust = 0.5, face = \"bold\"),\n",
    "              plot.margin = margin(l = 6, t = 12, r = 3, unit = \"pt\")) +\n",
    "        cf +\n",
    "        coord_fixed(clip = \"off\")\n",
    "      print(plot)\n",
    "      ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\",\n",
    "                              comparison, \"AGE_MAST_Pathways_gseaREACTOMEenrich_pathExplore.png\",\n",
    "                              sep = \"_\"),\n",
    "             plot = plot,\n",
    "             width = 600*2,\n",
    "             height = 400*2,\n",
    "             dpi = 150,\n",
    "             units = \"px\")\n",
    "    },\n",
    "      error=function(e){cat(\"ERROR :\",conditionMessage(e), \"\\n\")})\n",
    "   } \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5069a42e",
   "metadata": {},
   "source": [
    "## Seurat Differential Expression: Individual Sex\n",
    "\n",
    "Sex = 1: Male  \n",
    "Sex = 2: Female  \n",
    "\n",
    "Run the DEG workflow for all our Males, then all our Females. \n",
    "\n",
    "(Code displayed, but not run for this knitted html. \n",
    "This was used to check something previously.)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c4eacf5",
   "metadata": {
    "eval": false,
    "name": "Seurat_Differential_Expression_Function"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Differential_Expression\n",
    "do_all_differential_expression <- function(sex, seurat_anno) {\n",
    "  #----- RNA Assay stores the original and normalized counts\n",
    "  # Original counts : seurat_anno@assays[[\"RNA\"]]@counts\n",
    "  # Normalized counts : seurat_anno@assays[[\"RNA\"]]@data\n",
    "  DefaultAssay(seurat_anno) <- \"RNA\"\n",
    "  #----- Subset by Sex\n",
    "  # subset to only our Sex of interest.\n",
    "  if (sex %in% c(\"M\", \"Male\", \"male\", \"m\")) {\n",
    "    sex_coded <- 1\n",
    "    sex <- \"M\"\n",
    "  } else if (sex %in% c(\"F\", \"Female\", \"female\", \"f\")) {\n",
    "    sex_coded <- 2\n",
    "    sex <- \"F\"\n",
    "  } else {\n",
    "    stop(paste0(\"Invalid `sex` argument. You provided: `sex` = \", sex,\n",
    "                \"\\n`sex` must be one of the following: \",\n",
    "                '\\nfor Males: M, Male, male, or m',\n",
    "                '\\nfor Females: F, Female, female, f'))\n",
    "  }\n",
    "  seurat_anno <- subset(seurat_anno, subset = SEX %in% sex_coded)\n",
    "  #----- Group Celltype ID\n",
    "  # make this the Idents()\n",
    "  seurat_anno$group_cluster <- paste(seurat_anno$GROUP, \n",
    "                                     seurat_anno$cluster_annotation, \n",
    "                                     sep = \"_\")\n",
    "  Idents(seurat_anno) <- seurat_anno$group_cluster\n",
    "  \n",
    "  #---------- Differential Expressed Genes\n",
    "  #----- Cell Types\n",
    "  cell_types <- seurat_anno@meta.data %>% \n",
    "    dplyr::arrange(cluster_annotation) %>% \n",
    "    dplyr::select(cluster_annotation) %>% \n",
    "    unique() %>% \n",
    "    pull(cluster_annotation)\n",
    "  #----- Cell_Type Differential Expressed Genes\n",
    "  get_deg <- function(seurat_obj, cell_type, ident1, ident2){\n",
    "    FindMarkers(seurat_obj, \n",
    "                ident.1 = paste(ident1, cell_type, sep = \"_\"), \n",
    "                ident.2 = paste(ident2, cell_type, sep = \"_\"), \n",
    "                test.use = \"MAST\", \n",
    "                latent.vars = \"AGE\", \n",
    "                logfc.threshold = 0, \n",
    "                min.pct = 0,\n",
    "                min.diff.pct = -Inf, \n",
    "                min.cells.feature = 10, \n",
    "                min.cells.group = 10, \n",
    "                verbose = FALSE) %>% \n",
    "      rownames_to_column(var = \"Gene_Symbol\") %>% \n",
    "      left_join(Gene_Emsembl_Symbol, by=\"Gene_Symbol\") %>% \n",
    "      mutate(Comparison = paste(ident1, \"vs\", ident2, cell_type, sep = \"_\"))\n",
    "  }\n",
    "  Differential_Expression <- lapply(cell_types, function(cell_type) {\n",
    "  tryCatch(get_deg(seurat_obj = seurat_anno,\n",
    "                   cell_type = cell_type,\n",
    "                   ident1 = \"PD\",\n",
    "                   ident2 = \"HC\"),\n",
    "           error = function(e) \n",
    "             # empty dataframe with same column headers & types...\n",
    "             data.frame(Gene_Symbol = character(),\n",
    "                        p_val = double(),\n",
    "                        avg_log2FC = double(),\n",
    "                        pct.1 = double(),\n",
    "                        pct.2 = double(),\n",
    "                        p_val_adj = double(),\n",
    "                        Gene_Emsembl = character(),\n",
    "                        Comparison = character(),\n",
    "                        stringsAsFactors = FALSE))\n",
    "  }) %>% bind_rows()\n",
    "  #----- save the results\n",
    "  write_delim(Differential_Expression, \n",
    "    paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_PD_vs_HC_AGE_MAST_\", sex, \".txt\"), \n",
    "    delim = \"\\t\")\n",
    "  \n",
    "  #----- Number of DEGs\n",
    "  Differential_Expression_count <- Differential_Expression %>% \n",
    "    mutate(Significant = case_when(\n",
    "             p_val_adj <= 0.05 & avg_log2FC >= 0.25 ~ 'Upregulated',\n",
    "             p_val_adj <= 0.05 & avg_log2FC <= -0.25 ~ 'Downregulated',\n",
    "             TRUE ~ 'Not_Significant')) %>% \n",
    "    dplyr::count(Comparison, Significant) %>% \n",
    "    dplyr::rename(Count = n) %>% \n",
    "    tidyr::spread(Significant, Count)\n",
    "  #----- save the results\n",
    "  write_delim(Differential_Expression_count, \n",
    "    paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_PD_vs_HC_AGE_MAST_count_\", sex, \".txt\"), \n",
    "    delim = \"\\t\")\n",
    "  \n",
    "  \n",
    "  #---------- Differential Expressed Genes Pathways\n",
    "  #----- Comparisons\n",
    "  comparisons <- Differential_Expression$Comparison %>% \n",
    "                   unique()\n",
    "  #----- Pathways GSEA GOBP\n",
    "  Differential_Expression_GOBP <- lapply(comparisons, function(comparison) {\n",
    "    genes <- Differential_Expression %>% \n",
    "      dplyr::filter(Comparison == comparison) %>%\n",
    "      dplyr::select(Comparison, \n",
    "                    Gene_Symbol, \n",
    "                    avg_log2FC, \n",
    "                    p_val) %>%\n",
    "      dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                      TRUE ~ p_val)) %>%\n",
    "      dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "      dplyr::arrange(sign_log10Pvalue)\n",
    "    genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                         genes$Gene_Symbol) %>% \n",
    "      sort(decreasing = TRUE)\n",
    "    genes_go <- gseGO(geneList = genelist, \n",
    "                      ont =\"BP\", \n",
    "                      OrgDb = get(\"org.Hs.eg.db\"),\n",
    "                      keyType = \"SYMBOL\", \n",
    "                      minGSSize = 10, \n",
    "                      maxGSSize = 500, \n",
    "                      pvalueCutoff = 0.05, \n",
    "                      pAdjustMethod = \"BH\",\n",
    "                      verbose = TRUE)\n",
    "    return(genes_go@result %>% mutate(Comparison = comparison))\n",
    "  }) %>% bind_rows()\n",
    "  #----- Save Results\n",
    "  write_delim(Differential_Expression_GOBP, \n",
    "    paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_PD_vs_HC_AGE_MAST_Pathways_GSEA_GOBP_\", sex, \".txt\"), \n",
    "    delim = \"\\t\")\n",
    "  #----- Reactome\n",
    "  reactome <- msigdbr(species = \"Homo sapiens\", \n",
    "                      category = \"C2\", \n",
    "                      subcategory = \"CP:REACTOME\") %>% \n",
    "                dplyr::select(gs_name, gene_symbol)\n",
    "  #----- Pathways GSEA REACTOME\n",
    "  Differential_Expression_REACTOME <- lapply(comparisons, function(comparison) {\n",
    "    genes <- Differential_Expression %>% \n",
    "      dplyr::filter(Comparison == comparison) %>%\n",
    "      dplyr::select(Comparison, \n",
    "                    Gene_Symbol, \n",
    "                    avg_log2FC, \n",
    "                    p_val) %>%\n",
    "      dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                      TRUE ~ p_val)) %>%\n",
    "      dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "      dplyr::arrange(sign_log10Pvalue)\n",
    "    genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                         genes$Gene_Symbol) %>% \n",
    "      sort(decreasing = TRUE)\n",
    "    genes_gsea <- GSEA(geneList = genelist,\n",
    "                       minGSSize = 10,\n",
    "                       maxGSSize = 500,\n",
    "                       pvalueCutoff = 0.05,\n",
    "                       pAdjustMethod = \"BH\",\n",
    "                       TERM2GENE = reactome, \n",
    "                       by = \"fgsea\",\n",
    "                       verbose = TRUE)\n",
    "    return(genes_gsea@result %>% mutate(Comparison = comparison))\n",
    "  }) %>% bind_rows()\n",
    "  #----- Save Results\n",
    "  write_delim(Differential_Expression_REACTOME, \n",
    "    paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_PD_vs_HC_AGE_MAST_Pathways_GSEA_REACTOME_\", sex, \".txt\"), \n",
    "    delim = \"\\t\")\n",
    "}\n",
    "\n",
    "#----- Load data if needed\n",
    "if (!exists(\"seurat_anno\")) {\n",
    "  seurat_anno <- readRDS(file = \"seurat_results/seurat_anno.rds\")\n",
    "}\n",
    "#----- Do differential expression on each of the individual sexes\n",
    "do_all_differential_expression(sex = \"F\", \n",
    "                               seurat_anno = seurat_anno)\n",
    "do_all_differential_expression(sex = \"M\",\n",
    "                               seurat_anno = seurat_anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4d8835a",
   "metadata": {
    "name": "Seurat_Differential_Expression_Figures_Function"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Differential_Expression_Figures\n",
    "do_all_differential_expression_figures <- function(sex) {\n",
    "  #----- Process function arguments\n",
    "  if (sex %in% c(\"M\", \"Male\", \"male\", \"m\")) {\n",
    "    sex_coded <- 1\n",
    "    sex <- \"M\"\n",
    "  } else if (sex %in% c(\"F\", \"Female\", \"female\", \"f\")) {\n",
    "    sex_coded <- 2\n",
    "    sex <- \"F\"\n",
    "  } else {\n",
    "    stop(paste0(\"Invalid `sex` argument. You provided: `sex` = \", sex,\n",
    "                \"\\n`sex` must be one of the following: \",\n",
    "                '\\nfor Males: M, Male, male, or m',\n",
    "                '\\nfor Females: F, Female, female, f'))\n",
    "  }\n",
    "  \n",
    "  #----- Differential_Expression\n",
    "  Differential_Expression <- \n",
    "    read_delim(paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_PD_vs_HC_AGE_MAST_\", sex, \".txt\"), \n",
    "    delim = \"\\t\", escape_double = FALSE, \n",
    "    trim_ws = TRUE)\n",
    "  \n",
    "  #----- Differential_Expression_count\n",
    "  Differential_Expression_count <- \n",
    "    read_delim(paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_PD_vs_HC_AGE_MAST_count_\", sex, \".txt\"), \n",
    "    delim = \"\\t\", escape_double = FALSE, \n",
    "    trim_ws = TRUE)\n",
    "  Differential_Expression_count %>% \n",
    "    kable(caption=\"Number of Differentially Expressed Genes\", \n",
    "          format.args = list(big.mark = \",\")) %>% \n",
    "    kable_styling(c(\"striped\", \"bordered\"))\n",
    "  \n",
    "  \n",
    "  #---------- Differential Expressed Genes Volcano\n",
    "  #----- Comparisons\n",
    "  comparisons <- Differential_Expression$Comparison %>% \n",
    "                   unique()\n",
    "  #----- Differential Expressed Genes Volcano\n",
    "  for (comparison in comparisons){\n",
    "    plot <- Differential_Expression %>% \n",
    "              dplyr::filter(Comparison == comparison) %>%\n",
    "              dplyr::select(Comparison, \n",
    "                            Gene_Symbol, \n",
    "                            avg_log2FC, \n",
    "                            p_val,\n",
    "                            p_val_adj) %>% \n",
    "              mutate(Significant = case_when(\n",
    "                       p_val_adj <= 0.05 & avg_log2FC >= 0.25 ~ 'Upregulated',\n",
    "                       p_val_adj <= 0.05 & avg_log2FC <= -0.25 ~ 'Downregulated',\n",
    "                       TRUE ~ 'Not_Significant')) %>% \n",
    "              mutate(Label = case_when(\n",
    "                       Significant == \"Upregulated\" ~ Gene_Symbol,\n",
    "                       Significant == \"Downregulated\" ~ Gene_Symbol,\n",
    "                       TRUE ~ NA_character_)) %>% \n",
    "              ggplot(mapping = aes(x=avg_log2FC, \n",
    "                     y = -log10(p_val), \n",
    "                     color = Significant,\n",
    "                     label = Label)) + \n",
    "                geom_point(alpha = 0.5, stroke=0, size=2) + \n",
    "                geom_text_repel() +\n",
    "                ggtitle(paste0(comparison, \"\\n\", sex)) + \n",
    "                xlab(\"log2 Fold Change\") +\n",
    "                ylab(\"-log10(P Value)\") + \n",
    "                theme_classic() + \n",
    "                theme(plot.title = element_text(hjust = 0.5, \n",
    "                                                face = \"bold\")) + \n",
    "                scale_colour_manual(values = c('Downregulated' = 'Blue',\n",
    "                                               'Not_Significant' = 'Gray',\n",
    "                                               'Upregulated' = 'Red')) + \n",
    "                geom_vline(xintercept = c(-0.25,0.25), \n",
    "                           linetype = \"dotted\")\n",
    "    print(plot)\n",
    "    # png(file = paste(\"seurat_results/figures/Seurat_Differential_Expression\", comparison, \"AGE_MAST_volcano.png\", sep = \"_\"),\n",
    "    #   width = 600*2, \n",
    "    #   height = 400*2,\n",
    "    #   pointsize = 6)\n",
    "    # print(plot)\n",
    "    # dev.off()\n",
    "    ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\", comparison, \"AGE_MAST_volcano\", paste0(sex, \".png\"), sep = \"_\"),\n",
    "           plot = plot,\n",
    "           width = 600*2,\n",
    "           height = 400*2,\n",
    "           dpi = 150,\n",
    "           units = \"px\")\n",
    "  }\n",
    "  \n",
    "  \n",
    "  #---------- Differential Expressed Genes Pathways\n",
    "  #----- Comparisons\n",
    "  comparisons <- Differential_Expression$Comparison %>% \n",
    "                   unique()\n",
    "  #----- Pathways GSEA GOBP\n",
    "  for (comparison in comparisons){\n",
    "    genes <- Differential_Expression %>% \n",
    "      dplyr::filter(Comparison == comparison) %>%\n",
    "      dplyr::select(Comparison, \n",
    "                    Gene_Symbol, \n",
    "                    avg_log2FC, \n",
    "                    p_val) %>%\n",
    "      dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                      TRUE ~ p_val)) %>%\n",
    "      dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "      dplyr::arrange(sign_log10Pvalue)\n",
    "    genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                         genes$Gene_Symbol) %>% \n",
    "      sort(decreasing = TRUE)\n",
    "    genes_go <- gseGO(geneList = genelist, \n",
    "                      ont =\"BP\", \n",
    "                      OrgDb = get(\"org.Hs.eg.db\"),\n",
    "                      keyType = \"SYMBOL\", \n",
    "                      minGSSize = 10, \n",
    "                      maxGSSize = 500, \n",
    "                      pvalueCutoff = 0.05, \n",
    "                      pAdjustMethod = \"BH\",\n",
    "                      verbose = TRUE)\n",
    "    if (nrow(genes_go@result) == 0) {\n",
    "      # Error if this is empty & if we try to continue. Manual error handling...\n",
    "        # Error in `$<-.data.frame`(`*tmp*`, \".sign\", value = \"activated\") : \n",
    "        # replacement has 1 row, data has 0\n",
    "      warning(paste0(\"No Significant Pathways...\\n\",\n",
    "                     \"Skipping plot GSEA GOBP plot for: \", comparison, \" (\", sex, \")\"))\n",
    "    } else {\n",
    "      plot <- enrichplot::dotplot(object = genes_go, \n",
    "                                  showCategory = 10, \n",
    "                                  split = \".sign\") + \n",
    "                facet_grid(.~.sign) + \n",
    "                ggtitle(paste0(comparison, \" - GO:Biological Processs\\n\", sex)) + \n",
    "                theme(plot.title = element_text(hjust = 0.5, \n",
    "                                                face = \"bold\"))\n",
    "      print(plot)\n",
    "      png(file=\n",
    "        paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "               comparison, \"AGE_MAST_Pathways_gseaGOBP\", paste0(sex, \".png\"), sep = \"_\"), \n",
    "          width=1000, \n",
    "          height=600)\n",
    "      print(plot)\n",
    "      dev.off()\n",
    "      plot <- pairwise_termsim(x = genes_go) %>% \n",
    "                enrichplot::emapplot(showCategory=40) + \n",
    "                  ggtitle(paste0(comparison, \" - GO:Biological Process\\n\", sex)) + \n",
    "                  theme(plot.title = element_text(hjust = 0.5, \n",
    "                                                  face = \"bold\"))\n",
    "      print(plot)\n",
    "      png(file=\n",
    "        paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "               comparison, \"AGE_MAST_Pathways_gseaGOBPenrich\", paste0(sex, \".png\"), sep = \"_\"), \n",
    "          width=1400, \n",
    "          height=800)\n",
    "      print(plot)\n",
    "      dev.off()\n",
    "      }\n",
    "  }\n",
    "  #----- Reactome\n",
    "  reactome <- msigdbr(species = \"Homo sapiens\", \n",
    "                      category = \"C2\", \n",
    "                      subcategory = \"CP:REACTOME\") %>% \n",
    "                dplyr::select(gs_name, gene_symbol)\n",
    "  #----- Pathways GSEA REACTOME\n",
    "  for (comparison in comparisons){\n",
    "    genes <- Differential_Expression %>% \n",
    "      dplyr::filter(Comparison == comparison) %>%\n",
    "      dplyr::select(Comparison, \n",
    "                    Gene_Symbol, \n",
    "                    avg_log2FC, \n",
    "                    p_val) %>%\n",
    "      dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                      TRUE ~ p_val)) %>%\n",
    "      dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "      dplyr::arrange(sign_log10Pvalue)\n",
    "    genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                         genes$Gene_Symbol) %>% \n",
    "      sort(decreasing = TRUE)\n",
    "    genes_gsea <- GSEA(geneList = genelist,\n",
    "                       minGSSize = 10,\n",
    "                       maxGSSize = 500,\n",
    "                       pvalueCutoff = 0.05,\n",
    "                       pAdjustMethod = \"BH\",\n",
    "                       TERM2GENE = reactome, \n",
    "                       by = \"fgsea\",\n",
    "                       verbose = TRUE)\n",
    "    if (nrow(genes_gsea@result) == 0) {\n",
    "      # Error if this is empty & if we try to continue. Manual error handling...\n",
    "        # Error in `$<-.data.frame`(`*tmp*`, \".sign\", value = \"activated\") : \n",
    "        # replacement has 1 row, data has 0\n",
    "      warning(paste0(\"No Significant Pathways...\\n\",\n",
    "                     \"Skipping plot GSEA Reactome plot for: \", comparison, \" (\", sex, \")\"))\n",
    "    } else {\n",
    "      plot <- enrichplot::dotplot(object = genes_gsea, \n",
    "                                  showCategory = 10, \n",
    "                                  split = \".sign\") + \n",
    "                facet_grid(.~.sign) + \n",
    "                ggtitle(paste0(comparison, \" - GSEA Reactome\\n\", sex)) + \n",
    "                theme(plot.title = element_text(hjust = 0.5, \n",
    "                                                face = \"bold\"))\n",
    "      print(plot)\n",
    "      png(file=\n",
    "        paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "               comparison, \"AGE_MAST_Pathways_gseaREACTOME\", paste0(sex, \".png\"), sep = \"_\"), \n",
    "          width=1000, \n",
    "          height=600)\n",
    "      print(plot)\n",
    "      dev.off()\n",
    "      plot <- pairwise_termsim(x = genes_gsea) %>% \n",
    "                enrichplot::emapplot(showCategory=40) + \n",
    "                  ggtitle(paste0(comparison, \" - GSEA Reactome\\n\", sex)) + \n",
    "                  theme(plot.title = element_text(hjust = 0.5, \n",
    "                                                  face = \"bold\"))\n",
    "      print(plot)\n",
    "      png(file=\n",
    "        paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "               comparison, \"AGE_MAST_Pathways_gseaREACTOMEenrich\", paste0(sex, \".png\"), sep = \"_\"), \n",
    "          width=1400, \n",
    "          height=800)\n",
    "      print(plot)\n",
    "      dev.off()\n",
    "    } \n",
    "  }\n",
    "}\n",
    "do_all_differential_expression_figures(sex = \"F\")\n",
    "do_all_differential_expression_figures(sex = \"M\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb8d803a",
   "metadata": {},
   "source": [
    "## Export Main Data (Vugene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f6dfe85",
   "metadata": {
    "eval": false,
    "name": "Export_Data_Vugene"
   },
   "outputs": [],
   "source": [
    "#---------- Export_Data_Vugene\n",
    "#----- Load Anno Object\n",
    "if (!exists(\"seurat_anno\")) {\n",
    "  seurat_anno <- readRDS(file = \"seurat_results/seurat_anno.rds\")\n",
    "}\n",
    "\n",
    "#----- Get Data: Counts\n",
    "# ?FetchData\n",
    "# raw_counts_vugene <- FetchData(object = seurat_anno,\n",
    "#                                # need vars... I think I want everything\n",
    "#                                slot = \"counts\") # default is data, I want counts\n",
    "# ?GetAssayData()\n",
    "raw_counts_vugene <- GetAssayData(object = seurat_anno, \n",
    "                                  slot = \"counts\" # default is data, I want counts\n",
    "                                  ) # I only have one assay (RNA) so I won't specify\n",
    "\n",
    "#----- Get Data: Annotations\n",
    "annotations_vugene <- dplyr::select(seurat_anno@meta.data, cluster_annotation)\n",
    "\n",
    "#----- Save Data\n",
    "saveRDS(raw_counts_vugene, \n",
    "        file=paste0(\"seurat_results/\", NameProject, \"_raw_counts_vugene.rds\"))\n",
    "# write.table(as.matrix(raw_counts_vugene[1:3,1:3]),\n",
    "#             file = \"seurat_results/PD_MFG_snRNAseq_with_redos_raw_counts_vugene.txt\",\n",
    "#             quote = FALSE)\n",
    "\n",
    "write.table(annotations_vugene,\n",
    "            sep = \"\\t\",\n",
    "            quote = FALSE,\n",
    "            col.names = FALSE,\n",
    "            file = paste0(\"seurat_results/\", NameProject, \"_annotations_vugene.tsv\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3666ef2",
   "metadata": {},
   "source": [
    "## Seurat Differential Expression: p16+ / p16-\n",
    "\n",
    "Find DEG between PD and HC per cell_type\n",
    "Run the DEG workflow between p16+ and p16- cells.\n",
    "Using 0.25 logFC  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f491a330",
   "metadata": {
    "eval": false,
    "name": "Seurat_Differential_Expression_p16"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Differential_Expression_p16\n",
    "#----- Load data if needed\n",
    "if (!exists(\"seurat_anno\")) {\n",
    "  seurat_anno <- readRDS(file = \"seurat_results/seurat_anno.rds\")\n",
    "}\n",
    "\n",
    "#----- RNA Assay stores the original and normalized counts\n",
    "# Original counts : seurat_anno@assays[[\"RNA\"]]@counts\n",
    "# Normalized counts : seurat_anno@assays[[\"RNA\"]]@data\n",
    "DefaultAssay(seurat_anno) <- \"RNA\"\n",
    "#----- Group Celltype ID\n",
    "# make this the Idents()\n",
    "metadata_p16 <- seurat_anno@meta.data %>%\n",
    "  mutate(CDKN2A_Binary = case_when(CDKN2A_Count > 0 ~ \"CDKN2A_Pos\",\n",
    "                                   CDKN2A_Count == 0 ~ \"CDKN2A_Neg\"))\n",
    "seurat_anno@meta.data <- metadata_p16\n",
    "rm(metadata_p16)\n",
    "seurat_anno$group_cluster <- paste(seurat_anno$CDKN2A_Binary, \n",
    "                                   seurat_anno$cluster_annotation, \n",
    "                                   sep = \"_\")\n",
    "Idents(seurat_anno) <- seurat_anno$group_cluster\n",
    "\n",
    "#---------- Differential Expressed Genes\n",
    "#----- Cell Types\n",
    "cell_types <- seurat_anno@meta.data %>% \n",
    "  dplyr::arrange(cluster_annotation) %>% \n",
    "  dplyr::select(cluster_annotation) %>% \n",
    "  unique() %>% \n",
    "  pull(cluster_annotation)\n",
    "#----- Cell_Type Differential Expressed Genes\n",
    "get_deg <- function(seurat_obj, cell_type, ident1, ident2){\n",
    "  FindMarkers(seurat_obj, \n",
    "              ident.1 = paste(ident1, cell_type, sep = \"_\"), \n",
    "              ident.2 = paste(ident2, cell_type, sep = \"_\"), \n",
    "              test.use = \"MAST\", \n",
    "              latent.vars = \"AGE\", \n",
    "              logfc.threshold = 0, \n",
    "              min.pct = 0,\n",
    "              min.diff.pct = -Inf, \n",
    "              min.cells.feature = 10, \n",
    "              min.cells.group = 10, \n",
    "              verbose = FALSE) %>% \n",
    "    rownames_to_column(var = \"Gene_Symbol\") %>% \n",
    "    left_join(Gene_Emsembl_Symbol, by=\"Gene_Symbol\") %>% \n",
    "    mutate(Comparison = paste(ident1, \"vs\", ident2, cell_type, sep = \"_\"))\n",
    "}\n",
    "Differential_Expression <- lapply(cell_types, function(cell_type) {\n",
    "  tryCatch(get_deg(seurat_obj = seurat_anno,\n",
    "                   cell_type = cell_type,\n",
    "                   ident1 = \"CDKN2A_Pos\",\n",
    "                   ident2 = \"CDKN2A_Neg\"),\n",
    "           error = function(e) \n",
    "             # empty dataframe with same column headers & types...\n",
    "             data.frame(Gene_Symbol = character(),\n",
    "                        p_val = double(),\n",
    "                        avg_log2FC = double(),\n",
    "                        pct.1 = double(),\n",
    "                        pct.2 = double(),\n",
    "                        p_val_adj = double(),\n",
    "                        Gene_Emsembl = character(),\n",
    "                        Comparison = character(),\n",
    "                        stringsAsFactors = FALSE))\n",
    "}) %>% bind_rows()\n",
    "#----- save the results\n",
    "write_delim(Differential_Expression, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN2A_Pos_vs_CDKN2A_Neg_AGE_MAST.txt\"), \n",
    "  delim = \"\\t\")\n",
    "#----- Number of DEGs\n",
    "Differential_Expression_count <- Differential_Expression %>% \n",
    "  mutate(Significant = case_when(\n",
    "           p_val_adj <= 0.05 & avg_log2FC >= 0.25 ~ 'Upregulated',\n",
    "           p_val_adj <= 0.05 & avg_log2FC <= -0.25 ~ 'Downregulated',\n",
    "           TRUE ~ 'Not_Significant')) %>% \n",
    "  dplyr::count(Comparison, Significant) %>% \n",
    "  dplyr::rename(Count = n) %>% \n",
    "  tidyr::spread(Significant, Count)\n",
    "#----- save the results\n",
    "write_delim(Differential_Expression_count, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN2A_Pos_vs_CDKN2A_Neg_AGE_MAST_count.txt\"), \n",
    "  delim = \"\\t\")\n",
    "\n",
    "\n",
    "#---------- Differential Expressed Genes Pathways\n",
    "#----- Comparisons\n",
    "comparisons <- Differential_Expression$Comparison %>% \n",
    "                 unique()\n",
    "#----- Pathways GSEA GOBP\n",
    "Differential_Expression_GOBP <- lapply(comparisons, function(comparison) {\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_go <- gseGO(geneList = genelist, \n",
    "                    ont =\"BP\", \n",
    "                    OrgDb = get(\"org.Hs.eg.db\"),\n",
    "                    keyType = \"SYMBOL\", \n",
    "                    minGSSize = 10, \n",
    "                    maxGSSize = 500, \n",
    "                    pvalueCutoff = 0.05, \n",
    "                    pAdjustMethod = \"BH\",\n",
    "                    verbose = TRUE)\n",
    "  return(genes_go@result %>% mutate(Comparison = comparison))\n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(Differential_Expression_GOBP, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN2A_Pos_vs_CDKN2A_Neg_AGE_MAST_Pathways_GSEA_GOBP.txt\"), \n",
    "  delim = \"\\t\")\n",
    "#----- Reactome\n",
    "reactome <- msigdbr(species = \"Homo sapiens\", \n",
    "                    category = \"C2\", \n",
    "                    subcategory = \"CP:REACTOME\") %>% \n",
    "              dplyr::select(gs_name, gene_symbol)\n",
    "#----- Pathways GSEA REACTOME\n",
    "Differential_Expression_REACTOME <- lapply(comparisons, function(comparison) {\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_gsea <- GSEA(geneList = genelist,\n",
    "                     minGSSize = 10,\n",
    "                     maxGSSize = 500,\n",
    "                     pvalueCutoff = 0.05,\n",
    "                     pAdjustMethod = \"BH\",\n",
    "                     TERM2GENE = reactome, \n",
    "                     by = \"fgsea\",\n",
    "                     verbose = TRUE)\n",
    "  return(genes_gsea@result %>% mutate(Comparison = comparison))\n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(Differential_Expression_REACTOME, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN2A_Pos_vs_CDKN2A_Neg_AGE_MAST_Pathways_GSEA_REACTOME.txt\"), \n",
    "  delim = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1f00124",
   "metadata": {
    "name": "Seurat_Differential_Expression_Figures_p16"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Differential_Expression_Figures_p16\n",
    "Differential_Expression <- \n",
    "  read_delim(paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN2A_Pos_vs_CDKN2A_Neg_AGE_MAST.txt\"), \n",
    "  delim = \"\\t\", escape_double = FALSE, \n",
    "  trim_ws = TRUE)\n",
    "\n",
    "#----- Differential_Expression_count\n",
    "Differential_Expression_count <- \n",
    "  read_delim(paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN2A_Pos_vs_CDKN2A_Neg_AGE_MAST_count.txt\"), \n",
    "  delim = \"\\t\", escape_double = FALSE, \n",
    "  trim_ws = TRUE)\n",
    "Differential_Expression_count %>% \n",
    "  kable(caption=\"Number of Differentially Expressed Genes\", \n",
    "        format.args = list(big.mark = \",\")) %>% \n",
    "  kable_styling(c(\"striped\", \"bordered\"))\n",
    "\n",
    "\n",
    "#---------- Differential Expressed Genes Volcano\n",
    "#----- Comparisons\n",
    "comparisons <- Differential_Expression$Comparison %>% \n",
    "                 unique()\n",
    "#----- Differential Expressed Genes Volcano\n",
    "for (comparison in comparisons){\n",
    "  plot <- Differential_Expression %>% \n",
    "            dplyr::filter(Comparison == comparison) %>%\n",
    "            dplyr::select(Comparison, \n",
    "                          Gene_Symbol, \n",
    "                          avg_log2FC, \n",
    "                          p_val,\n",
    "                          p_val_adj) %>% \n",
    "            mutate(Significant = case_when(\n",
    "                     p_val_adj <= 0.05 & avg_log2FC >= 0.25 ~ 'Upregulated',\n",
    "                     p_val_adj <= 0.05 & avg_log2FC <= -0.25 ~ 'Downregulated',\n",
    "                     TRUE ~ 'Not_Significant')) %>% \n",
    "            mutate(Label = case_when(\n",
    "                     Significant == \"Upregulated\" ~ Gene_Symbol,\n",
    "                     Significant == \"Downregulated\" ~ Gene_Symbol,\n",
    "                     TRUE ~ NA_character_)) %>% \n",
    "            ggplot(mapping = aes(x=avg_log2FC, \n",
    "                   y = -log10(p_val), \n",
    "                   color = Significant,\n",
    "                   label = Label)) + \n",
    "              geom_point(alpha = 0.5, stroke=0, size=2) + \n",
    "              geom_text_repel() +\n",
    "              ggtitle(comparison) + \n",
    "              xlab(\"log2 Fold Change\") +\n",
    "              ylab(\"-log10(P Value)\") + \n",
    "              theme_classic() + \n",
    "              theme(plot.title = element_text(hjust = 0.5, \n",
    "                                              face = \"bold\")) + \n",
    "              scale_colour_manual(values = c('Downregulated' = 'Blue',\n",
    "                                             'Not_Significant' = 'Gray',\n",
    "                                             'Upregulated' = 'Red')) + \n",
    "              geom_vline(xintercept = c(-0.25,0.25), \n",
    "                         linetype = \"dotted\")\n",
    "  print(plot)\n",
    "  # png(file = paste(\"seurat_results/figures/Seurat_Differential_Expression\", comparison, \"AGE_MAST_volcano.png\", sep = \"_\"),\n",
    "  #   width = 600*2, \n",
    "  #   height = 400*2,\n",
    "  #   pointsize = 6)\n",
    "  # print(plot)\n",
    "  # dev.off()\n",
    "  ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\", comparison, \"AGE_MAST_volcano.png\", sep = \"_\"),\n",
    "         plot = plot,\n",
    "         width = 600*2,\n",
    "         height = 400*2,\n",
    "         dpi = 150,\n",
    "         units = \"px\")\n",
    "}\n",
    "\n",
    "\n",
    "#---------- Differential Expressed Genes Pathways\n",
    "#----- Comparisons\n",
    "comparisons <- Differential_Expression$Comparison %>% \n",
    "                 unique()\n",
    "#----- Pathways GSEA GOBP\n",
    "for (comparison in comparisons){\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_go <- gseGO(geneList = genelist, \n",
    "                    ont =\"BP\", \n",
    "                    OrgDb = get(\"org.Hs.eg.db\"),\n",
    "                    keyType = \"SYMBOL\", \n",
    "                    minGSSize = 10, \n",
    "                    maxGSSize = 500, \n",
    "                    pvalueCutoff = 0.05, \n",
    "                    pAdjustMethod = \"BH\",\n",
    "                    verbose = TRUE)\n",
    "  if (nrow(genes_go@result) == 0) {\n",
    "     warning(paste0(\"No significant GOBP genes for: \", comparison, \"\\nSkipping to next comparison...\"))\n",
    "   } else {\n",
    "     plot <- enrichplot::dotplot(object = genes_go, \n",
    "                                  showCategory = 10, \n",
    "                                  split = \".sign\") + \n",
    "                facet_grid(.~.sign) + \n",
    "                ggtitle(paste0(comparison, \" - GO:Biological Process\")) + \n",
    "                theme(plot.title = element_text(hjust = 0.5, \n",
    "                                                face = \"bold\"))\n",
    "      print(plot)\n",
    "      png(file=\n",
    "        paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "               comparison, \"AGE_MAST_Pathways_gseaGOBP.png\", sep = \"_\"), \n",
    "          width=1000, \n",
    "          height=600)\n",
    "      print(plot)\n",
    "      dev.off()\n",
    "   } \n",
    "  \n",
    "  #----- Split GOBP Pathway Map by Activated / Suppressed\n",
    "  # (otherwise the activated & suppressed are shown together -- misleading)\n",
    "  genes_go_emapplot <- genes_go\n",
    "  genes_go_emapplot@result <- genes_go_emapplot@result %>%\n",
    "    mutate(pathway_direction = case_when(NES > 0 ~ \"activated\",\n",
    "                                         NES < 0 ~ \"suppressed\"))\n",
    "  genes_go_emapplot_activated <- dplyr::filter(genes_go_emapplot, \n",
    "                                               pathway_direction == \"activated\")\n",
    "  genes_go_emapplot_suppressed <- dplyr::filter(genes_go_emapplot, \n",
    "                                               pathway_direction == \"suppressed\")\n",
    "  \n",
    "  # plot activated\n",
    "  if (nrow(genes_go_emapplot_activated@result) == 0) {\n",
    "     warning(paste0(\"No significant activated GOBP genes for: \", \n",
    "                    comparison, \"\\nSkipping to next comparison...\"))\n",
    "    plot_activated <- ggplot() + theme_void()\n",
    "  } else {\n",
    "    genes_go_emapplot_activated <- pairwise_termsim(x = genes_go_emapplot_activated)\n",
    "    plot_activated <- genes_go_emapplot_activated %>% \n",
    "              enrichplot::emapplot(showCategory=40) +\n",
    "              theme(plot.title = element_text(hjust = 0.5,\n",
    "                                              face = \"bold\"))\n",
    "    print(plot_activated)\n",
    "  }\n",
    "  \n",
    "  # plot suppressed\n",
    "  if (nrow(genes_go_emapplot_suppressed@result) == 0) {\n",
    "     warning(paste0(\"No significant suppressed GOBP genes for: \", \n",
    "                    comparison, \"\\nSkipping to next comparison...\"))\n",
    "    plot_suppressed <- ggplot() + theme_void()\n",
    "  } else {\n",
    "    genes_go_emapplot_suppressed <- pairwise_termsim(x = genes_go_emapplot_suppressed)\n",
    "    plot_suppressed <- genes_go_emapplot_suppressed %>% \n",
    "              enrichplot::emapplot(showCategory=40) +\n",
    "                theme(plot.title = element_text(hjust = 0.5,\n",
    "                                                face = \"bold\"))\n",
    "    print(plot_suppressed)\n",
    "  } \n",
    "  \n",
    "  # combine & save plot\n",
    "  sub_plot_labels <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = \"ACTIVATED\",\n",
    "                                                                       size = 18,\n",
    "                                                                       hjust = 0.5,\n",
    "                                                                       fontface = \"bold\"),\n",
    "                                        ggdraw() + cowplot::draw_label(label = \"SUPPRESSED\",\n",
    "                                                                       size = 18,\n",
    "                                                                       hjust = 0.5,\n",
    "                                                                       fontface = \"bold\"),\n",
    "                                        ncol = 2)\n",
    "  \n",
    "  \n",
    "  plot <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = paste0(comparison, \n",
    "                                                     \" - GO:Biological Process\"),\n",
    "                                                     size = 24,\n",
    "                                                     hjust = 0.5,\n",
    "                                                     fontface = \"bold\"),\n",
    "                             sub_plot_labels,\n",
    "                             plot_grid(plot_activated, plot_suppressed),\n",
    "                             ncol = 1, rel_heights = c(0.1, .05, 1))\n",
    "\n",
    "  plot\n",
    "  \n",
    "  png(file=\n",
    "    paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "           comparison, \"AGE_MAST_Pathways_gseaGOBPenrich.png\", sep = \"_\"), \n",
    "      width=1400, \n",
    "      height=800)\n",
    "  print(plot)\n",
    "  dev.off()\n",
    "  #----- pathExplore: High-Level GSEA Clusters (GO:BP)\n",
    "  tryCatch({\n",
    "    # fix here: https://github.com/tidyverse/ggplot2/issues/2799\n",
    "    cf <- coord_fixed()\n",
    "    cf$default <- TRUE\n",
    "    plot <- enrichmentNetwork(genes_go@result, drawEllipses = TRUE, fontSize = 2.5) +\n",
    "      ggtitle(paste0(comparison, \" - GO:Biological Process\")) +\n",
    "      theme(plot.title = element_text(hjust = 0.5, face = \"bold\"),\n",
    "            plot.margin = margin(l = 6, t = 12, r = 3, unit = \"pt\")) +\n",
    "      cf +\n",
    "      coord_fixed(clip = \"off\")\n",
    "    print(plot)\n",
    "    ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\",\n",
    "                            comparison, \"AGE_MAST_Pathways_gseaGOBPenrich_pathExplore.png\", \n",
    "                            sep = \"_\"),\n",
    "           plot = plot,\n",
    "           width = 600*2,\n",
    "           height = 400*2,\n",
    "           dpi = 150,\n",
    "           units = \"px\")\n",
    "  },\n",
    "    error=function(e){cat(\"ERROR :\",conditionMessage(e), \"\\n\")})\n",
    "}\n",
    "#----- Reactome\n",
    "reactome <- msigdbr(species = \"Homo sapiens\", \n",
    "                    category = \"C2\", \n",
    "                    subcategory = \"CP:REACTOME\") %>% \n",
    "              dplyr::select(gs_name, gene_symbol)\n",
    "#----- Pathways GSEA REACTOME\n",
    "for (comparison in comparisons){\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_gsea <- GSEA(geneList = genelist,\n",
    "                     minGSSize = 10,\n",
    "                     maxGSSize = 500,\n",
    "                     pvalueCutoff = 0.05,\n",
    "                     pAdjustMethod = \"BH\",\n",
    "                     TERM2GENE = reactome, \n",
    "                     by = \"fgsea\",\n",
    "                     verbose = TRUE)\n",
    "  if (nrow(genes_gsea@result) == 0) {\n",
    "    warning(paste0(\"No significant REACTOME genes for: \", comparison, \"\\nSkipping to next comparison...\"))\n",
    "   } else {\n",
    "    plot <- enrichplot::dotplot(object = genes_gsea,\n",
    "                                showCategory = 10, \n",
    "                                split = \".sign\") + \n",
    "              facet_grid(.~.sign) + \n",
    "              ggtitle(paste0(comparison, \" - GSEA Reactome\")) + \n",
    "              theme(plot.title = element_text(hjust = 0.5, \n",
    "                                              face = \"bold\"))\n",
    "    print(plot)\n",
    "    png(file=\n",
    "      paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "             comparison, \"AGE_MAST_Pathways_gseaREACTOME.png\", sep = \"_\"), \n",
    "        width=1000, \n",
    "        height=600)\n",
    "    print(plot)\n",
    "    dev.off()\n",
    "    ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "             comparison, \"AGE_MAST_Pathways_gseaREACTOME_quality.png\", sep = \"_\"), \n",
    "       plot = plot,\n",
    "       width = 1000*2,\n",
    "       height = 600*2,\n",
    "       dpi = 200,\n",
    "       units = \"px\")\n",
    "    \n",
    "    #----- Split REACTOME Pathway Map by Activated / Suppressed\n",
    "    # (otherwise the activated & suppressed are shown together -- misleading)\n",
    "    genes_gsea_emapplot <- genes_gsea\n",
    "    genes_gsea_emapplot@result <- genes_gsea_emapplot@result %>%\n",
    "      mutate(pathway_direction = case_when(NES > 0 ~ \"activated\",\n",
    "                                           NES < 0 ~ \"suppressed\"))\n",
    "    genes_gsea_emapplot_activated <- dplyr::filter(genes_gsea_emapplot, \n",
    "                                                 pathway_direction == \"activated\")\n",
    "    genes_gsea_emapplot_suppressed <- dplyr::filter(genes_gsea_emapplot, \n",
    "                                                 pathway_direction == \"suppressed\")\n",
    "    \n",
    "    # plot activated\n",
    "    if (nrow(genes_gsea_emapplot_activated@result) == 0) {\n",
    "       warning(paste0(\"No significant activated REACTOME genes for: \", \n",
    "                      comparison, \"\\nSkipping to next comparison...\"))\n",
    "      plot_activated <- ggplot() + theme_void()\n",
    "    } else {\n",
    "      genes_gsea_emapplot_activated <- pairwise_termsim(x = genes_gsea_emapplot_activated)\n",
    "      plot_activated <- genes_gsea_emapplot_activated %>% \n",
    "                enrichplot::emapplot(showCategory=40) +\n",
    "                theme(plot.title = element_text(hjust = 0.5,\n",
    "                                                face = \"bold\"))\n",
    "      print(plot_activated)\n",
    "    }\n",
    "    \n",
    "    # plot suppressed\n",
    "    if (nrow(genes_gsea_emapplot_suppressed@result) == 0) {\n",
    "       warning(paste0(\"No significant suppressed REACTOME genes for: \", \n",
    "                      comparison, \"\\nSkipping to next comparison...\"))\n",
    "      plot_suppressed <- ggplot() + theme_void()\n",
    "    } else {\n",
    "      genes_gsea_emapplot_suppressed <- pairwise_termsim(x = genes_gsea_emapplot_suppressed)\n",
    "      plot_suppressed <- genes_gsea_emapplot_suppressed %>% \n",
    "                enrichplot::emapplot(showCategory=40) +\n",
    "                  theme(plot.title = element_text(hjust = 0.5,\n",
    "                                                  face = \"bold\"))\n",
    "      print(plot_suppressed)\n",
    "    } \n",
    "    \n",
    "    # combine & save plot\n",
    "    sub_plot_labels <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = \"ACTIVATED\",\n",
    "                                                                         size = 18,\n",
    "                                                                         hjust = 0.5,\n",
    "                                                                         fontface = \"bold\"),\n",
    "                                          ggdraw() + cowplot::draw_label(label = \"SUPPRESSED\",\n",
    "                                                                         size = 18,\n",
    "                                                                         hjust = 0.5,\n",
    "                                                                         fontface = \"bold\"),\n",
    "                                          ncol = 2)\n",
    "    \n",
    "    \n",
    "    plot <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = paste0(comparison, \n",
    "                                                       \" - GSEA Reactome\"),\n",
    "                                                       size = 24,\n",
    "                                                       hjust = 0.5,\n",
    "                                                       fontface = \"bold\"),\n",
    "                               sub_plot_labels,\n",
    "                               plot_grid(plot_activated, plot_suppressed),\n",
    "                               ncol = 1, rel_heights = c(0.1, .05, 1))\n",
    "    print(plot)\n",
    "    png(file=\n",
    "      paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "             comparison, \"AGE_MAST_Pathways_gseaREACTOMEenrich.png\", sep = \"_\"), \n",
    "        width=1400, \n",
    "        height=800)\n",
    "    print(plot)\n",
    "    dev.off()\n",
    "    #----- pathExplore: High-Level GSEA Clusters (GSEA Reactome)\n",
    "    tryCatch({\n",
    "      # fix here: https://github.com/tidyverse/ggplot2/issues/2799\n",
    "      cf <- coord_fixed()\n",
    "      cf$default <- TRUE\n",
    "      plot <- pathExplore::enrichmentNetwork(genes_gsea@result, drawEllipses = TRUE, fontSize = 2.5) +\n",
    "        ggtitle(paste0(comparison, \" - GSEA Reactome\")) +\n",
    "        theme(plot.title = element_text(hjust = 0.5, face = \"bold\"),\n",
    "              plot.margin = margin(l = 6, t = 12, r = 3, unit = \"pt\")) +\n",
    "        cf +\n",
    "        coord_fixed(clip = \"off\")\n",
    "      print(plot)\n",
    "      ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\",\n",
    "                              comparison, \"AGE_MAST_Pathways_gseaREACTOMEenrich_pathExplore.png\",\n",
    "                              sep = \"_\"),\n",
    "             plot = plot,\n",
    "             width = 600*2,\n",
    "             height = 400*2,\n",
    "             dpi = 150,\n",
    "             units = \"px\")\n",
    "    },\n",
    "      error=function(e){cat(\"ERROR :\",conditionMessage(e), \"\\n\")})\n",
    "   } \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31d66c2f",
   "metadata": {},
   "source": [
    "## Seurat Differential Expression: p21+ / p21-\n",
    "\n",
    "Find DEG between p21+ and p21- per cell_type\n",
    "Using 0.25 logFC  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cd60517",
   "metadata": {
    "eval": false,
    "name": "Seurat_Differential_Expression_p21"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Differential_Expression_p21\n",
    "#----- Load data if needed\n",
    "if (!exists(\"seurat_anno\")) {\n",
    "  seurat_anno <- readRDS(file = \"seurat_results/seurat_anno.rds\")\n",
    "}\n",
    "\n",
    "#----- RNA Assay stores the original and normalized counts\n",
    "# Original counts : seurat_anno@assays[[\"RNA\"]]@counts\n",
    "# Normalized counts : seurat_anno@assays[[\"RNA\"]]@data\n",
    "DefaultAssay(seurat_anno) <- \"RNA\"\n",
    "#----- Group Celltype ID\n",
    "# make this the Idents()\n",
    "metadata_p21 <- seurat_anno@meta.data %>%\n",
    "  mutate(CDKN1A_Binary = case_when(CDKN1A_Count > 0 ~ \"CDKN1A_Pos\",\n",
    "                                   CDKN1A_Count == 0 ~ \"CDKN1A_Neg\"))\n",
    "seurat_anno@meta.data <- metadata_p21\n",
    "rm(metadata_p21)\n",
    "seurat_anno$group_cluster <- paste(seurat_anno$CDKN1A_Binary, \n",
    "                                   seurat_anno$cluster_annotation, \n",
    "                                   sep = \"_\")\n",
    "Idents(seurat_anno) <- seurat_anno$group_cluster\n",
    "\n",
    "#---------- Differential Expressed Genes\n",
    "#----- Cell Types\n",
    "cell_types <- seurat_anno@meta.data %>% \n",
    "  dplyr::arrange(cluster_annotation) %>% \n",
    "  dplyr::select(cluster_annotation) %>% \n",
    "  unique() %>% \n",
    "  pull(cluster_annotation)\n",
    "#----- Cell_Type Differential Expressed Genes\n",
    "get_deg <- function(seurat_obj, cell_type, ident1, ident2){\n",
    "  FindMarkers(seurat_obj, \n",
    "              ident.1 = paste(ident1, cell_type, sep = \"_\"), \n",
    "              ident.2 = paste(ident2, cell_type, sep = \"_\"), \n",
    "              test.use = \"MAST\", \n",
    "              latent.vars = \"AGE\", \n",
    "              logfc.threshold = 0, \n",
    "              min.pct = 0,\n",
    "              min.diff.pct = -Inf, \n",
    "              min.cells.feature = 10, \n",
    "              min.cells.group = 10, \n",
    "              verbose = FALSE) %>% \n",
    "    rownames_to_column(var = \"Gene_Symbol\") %>% \n",
    "    left_join(Gene_Emsembl_Symbol, by=\"Gene_Symbol\") %>% \n",
    "    mutate(Comparison = paste(ident1, \"vs\", ident2, cell_type, sep = \"_\"))\n",
    "}\n",
    "Differential_Expression <- lapply(cell_types, function(cell_type) {\n",
    "  tryCatch(get_deg(seurat_obj = seurat_anno,\n",
    "                   cell_type = cell_type,\n",
    "                   ident1 = \"CDKN1A_Pos\",\n",
    "                   ident2 = \"CDKN1A_Neg\"),\n",
    "           error = function(e) \n",
    "             # empty dataframe with same column headers & types...\n",
    "             data.frame(Gene_Symbol = character(),\n",
    "                        p_val = double(),\n",
    "                        avg_log2FC = double(),\n",
    "                        pct.1 = double(),\n",
    "                        pct.2 = double(),\n",
    "                        p_val_adj = double(),\n",
    "                        Gene_Emsembl = character(),\n",
    "                        Comparison = character(),\n",
    "                        stringsAsFactors = FALSE))\n",
    "}) %>% bind_rows()\n",
    "#----- save the results\n",
    "write_delim(Differential_Expression, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN1A_Pos_vs_CDKN1A_Neg_AGE_MAST.txt\"), \n",
    "  delim = \"\\t\")\n",
    "#----- Number of DEGs\n",
    "Differential_Expression_count <- Differential_Expression %>% \n",
    "  mutate(Significant = case_when(\n",
    "           p_val_adj <= 0.05 & avg_log2FC >= 0.25 ~ 'Upregulated',\n",
    "           p_val_adj <= 0.05 & avg_log2FC <= -0.25 ~ 'Downregulated',\n",
    "           TRUE ~ 'Not_Significant')) %>% \n",
    "  dplyr::count(Comparison, Significant) %>% \n",
    "  dplyr::rename(Count = n) %>% \n",
    "  tidyr::spread(Significant, Count)\n",
    "#----- save the results\n",
    "write_delim(Differential_Expression_count, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN1A_Pos_vs_CDKN1A_Neg_AGE_MAST_count.txt\"), \n",
    "  delim = \"\\t\")\n",
    "\n",
    "\n",
    "#---------- Differential Expressed Genes Pathways\n",
    "#----- Comparisons\n",
    "comparisons <- Differential_Expression$Comparison %>% \n",
    "                 unique()\n",
    "#----- Pathways GSEA GOBP\n",
    "Differential_Expression_GOBP <- lapply(comparisons, function(comparison) {\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_go <- gseGO(geneList = genelist, \n",
    "                    ont =\"BP\", \n",
    "                    OrgDb = get(\"org.Hs.eg.db\"),\n",
    "                    keyType = \"SYMBOL\", \n",
    "                    minGSSize = 10, \n",
    "                    maxGSSize = 500, \n",
    "                    pvalueCutoff = 0.05, \n",
    "                    pAdjustMethod = \"BH\",\n",
    "                    verbose = TRUE)\n",
    "  return(genes_go@result %>% mutate(Comparison = comparison))\n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(Differential_Expression_GOBP, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN1A_Pos_vs_CDKN1A_Neg_AGE_MAST_Pathways_GSEA_GOBP.txt\"), \n",
    "  delim = \"\\t\")\n",
    "#----- Reactome\n",
    "reactome <- msigdbr(species = \"Homo sapiens\", \n",
    "                    category = \"C2\", \n",
    "                    subcategory = \"CP:REACTOME\") %>% \n",
    "              dplyr::select(gs_name, gene_symbol)\n",
    "#----- Pathways GSEA REACTOME\n",
    "Differential_Expression_REACTOME <- lapply(comparisons, function(comparison) {\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_gsea <- GSEA(geneList = genelist,\n",
    "                     minGSSize = 10,\n",
    "                     maxGSSize = 500,\n",
    "                     pvalueCutoff = 0.05,\n",
    "                     pAdjustMethod = \"BH\",\n",
    "                     TERM2GENE = reactome, \n",
    "                     by = \"fgsea\",\n",
    "                     verbose = TRUE)\n",
    "  return(genes_gsea@result %>% mutate(Comparison = comparison))\n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(Differential_Expression_REACTOME, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN1A_Pos_vs_CDKN1A_Neg_AGE_MAST_Pathways_GSEA_REACTOME.txt\"), \n",
    "  delim = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d53b163",
   "metadata": {
    "name": "Seurat_Differential_Expression_Figures_p21"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Differential_Expression_Figures_p21\n",
    "Differential_Expression <- \n",
    "  read_delim(paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN1A_Pos_vs_CDKN1A_Neg_AGE_MAST.txt\"), \n",
    "  delim = \"\\t\", escape_double = FALSE, \n",
    "  trim_ws = TRUE)\n",
    "\n",
    "#----- Differential_Expression_count\n",
    "Differential_Expression_count <- \n",
    "  read_delim(paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN1A_Pos_vs_CDKN1A_Neg_AGE_MAST_count.txt\"), \n",
    "  delim = \"\\t\", escape_double = FALSE, \n",
    "  trim_ws = TRUE)\n",
    "Differential_Expression_count %>% \n",
    "  kable(caption=\"Number of Differentially Expressed Genes\", \n",
    "        format.args = list(big.mark = \",\")) %>% \n",
    "  kable_styling(c(\"striped\", \"bordered\"))\n",
    "\n",
    "\n",
    "#---------- Differential Expressed Genes Volcano\n",
    "#----- Comparisons\n",
    "comparisons <- Differential_Expression$Comparison %>% \n",
    "                 unique()\n",
    "#----- Differential Expressed Genes Volcano\n",
    "for (comparison in comparisons){\n",
    "  plot <- Differential_Expression %>% \n",
    "            dplyr::filter(Comparison == comparison) %>%\n",
    "            dplyr::select(Comparison, \n",
    "                          Gene_Symbol, \n",
    "                          avg_log2FC, \n",
    "                          p_val,\n",
    "                          p_val_adj) %>% \n",
    "            mutate(Significant = case_when(\n",
    "                     p_val_adj <= 0.05 & avg_log2FC >= 0.25 ~ 'Upregulated',\n",
    "                     p_val_adj <= 0.05 & avg_log2FC <= -0.25 ~ 'Downregulated',\n",
    "                     TRUE ~ 'Not_Significant')) %>% \n",
    "            mutate(Label = case_when(\n",
    "                     Significant == \"Upregulated\" ~ Gene_Symbol,\n",
    "                     Significant == \"Downregulated\" ~ Gene_Symbol,\n",
    "                     TRUE ~ NA_character_)) %>% \n",
    "            ggplot(mapping = aes(x=avg_log2FC, \n",
    "                   y = -log10(p_val), \n",
    "                   color = Significant,\n",
    "                   label = Label)) + \n",
    "              geom_point(alpha = 0.5, stroke=0, size=2) + \n",
    "              geom_text_repel() +\n",
    "              ggtitle(comparison) + \n",
    "              xlab(\"log2 Fold Change\") +\n",
    "              ylab(\"-log10(P Value)\") + \n",
    "              theme_classic() + \n",
    "              theme(plot.title = element_text(hjust = 0.5, \n",
    "                                              face = \"bold\")) + \n",
    "              scale_colour_manual(values = c('Downregulated' = 'Blue',\n",
    "                                             'Not_Significant' = 'Gray',\n",
    "                                             'Upregulated' = 'Red')) + \n",
    "              geom_vline(xintercept = c(-0.25,0.25), \n",
    "                         linetype = \"dotted\")\n",
    "  print(plot)\n",
    "  # png(file = paste(\"seurat_results/figures/Seurat_Differential_Expression\", comparison, \"AGE_MAST_volcano.png\", sep = \"_\"),\n",
    "  #   width = 600*2, \n",
    "  #   height = 400*2,\n",
    "  #   pointsize = 6)\n",
    "  # print(plot)\n",
    "  # dev.off()\n",
    "  ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\", comparison, \"AGE_MAST_volcano.png\", sep = \"_\"),\n",
    "         plot = plot,\n",
    "         width = 600*2,\n",
    "         height = 400*2,\n",
    "         dpi = 150,\n",
    "         units = \"px\")\n",
    "}\n",
    "\n",
    "\n",
    "#---------- Differential Expressed Genes Pathways\n",
    "#----- Comparisons\n",
    "comparisons <- Differential_Expression$Comparison %>% \n",
    "                 unique()\n",
    "#----- Pathways GSEA GOBP\n",
    "for (comparison in comparisons){\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_go <- gseGO(geneList = genelist, \n",
    "                    ont =\"BP\", \n",
    "                    OrgDb = get(\"org.Hs.eg.db\"),\n",
    "                    keyType = \"SYMBOL\", \n",
    "                    minGSSize = 10, \n",
    "                    maxGSSize = 500, \n",
    "                    pvalueCutoff = 0.05, \n",
    "                    pAdjustMethod = \"BH\",\n",
    "                    verbose = TRUE)\n",
    "  if (nrow(genes_go@result) == 0) {\n",
    "     warning(paste0(\"No significant GOBP genes for: \", comparison, \"\\nSkipping to next comparison...\"))\n",
    "   } else {\n",
    "     plot <- enrichplot::dotplot(object = genes_go, \n",
    "                                  showCategory = 10, \n",
    "                                  split = \".sign\") + \n",
    "                facet_grid(.~.sign) + \n",
    "                ggtitle(paste0(comparison, \" - GO:Biological Process\")) + \n",
    "                theme(plot.title = element_text(hjust = 0.5, \n",
    "                                                face = \"bold\"))\n",
    "      print(plot)\n",
    "      png(file=\n",
    "        paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "               comparison, \"AGE_MAST_Pathways_gseaGOBP.png\", sep = \"_\"), \n",
    "          width=1000, \n",
    "          height=600)\n",
    "      print(plot)\n",
    "      dev.off()\n",
    "   } \n",
    "  \n",
    "  #----- Split GOBP Pathway Map by Activated / Suppressed\n",
    "  # (otherwise the activated & suppressed are shown together -- misleading)\n",
    "  genes_go_emapplot <- genes_go\n",
    "  genes_go_emapplot@result <- genes_go_emapplot@result %>%\n",
    "    mutate(pathway_direction = case_when(NES > 0 ~ \"activated\",\n",
    "                                         NES < 0 ~ \"suppressed\"))\n",
    "  genes_go_emapplot_activated <- dplyr::filter(genes_go_emapplot, \n",
    "                                               pathway_direction == \"activated\")\n",
    "  genes_go_emapplot_suppressed <- dplyr::filter(genes_go_emapplot, \n",
    "                                               pathway_direction == \"suppressed\")\n",
    "  \n",
    "  # plot activated\n",
    "  if (nrow(genes_go_emapplot_activated@result) == 0) {\n",
    "     warning(paste0(\"No significant activated GOBP genes for: \", \n",
    "                    comparison, \"\\nSkipping to next comparison...\"))\n",
    "    plot_activated <- ggplot() + theme_void()\n",
    "  } else {\n",
    "    genes_go_emapplot_activated <- pairwise_termsim(x = genes_go_emapplot_activated)\n",
    "    plot_activated <- genes_go_emapplot_activated %>% \n",
    "              enrichplot::emapplot(showCategory=40) +\n",
    "              theme(plot.title = element_text(hjust = 0.5,\n",
    "                                              face = \"bold\"))\n",
    "    print(plot_activated)\n",
    "  }\n",
    "  \n",
    "  # plot suppressed\n",
    "  if (nrow(genes_go_emapplot_suppressed@result) == 0) {\n",
    "     warning(paste0(\"No significant suppressed GOBP genes for: \", \n",
    "                    comparison, \"\\nSkipping to next comparison...\"))\n",
    "    plot_suppressed <- ggplot() + theme_void()\n",
    "  } else {\n",
    "    genes_go_emapplot_suppressed <- pairwise_termsim(x = genes_go_emapplot_suppressed)\n",
    "    plot_suppressed <- genes_go_emapplot_suppressed %>% \n",
    "              enrichplot::emapplot(showCategory=40) +\n",
    "                theme(plot.title = element_text(hjust = 0.5,\n",
    "                                                face = \"bold\"))\n",
    "    print(plot_suppressed)\n",
    "  } \n",
    "  \n",
    "  # combine & save plot\n",
    "  sub_plot_labels <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = \"ACTIVATED\",\n",
    "                                                                       size = 18,\n",
    "                                                                       hjust = 0.5,\n",
    "                                                                       fontface = \"bold\"),\n",
    "                                        ggdraw() + cowplot::draw_label(label = \"SUPPRESSED\",\n",
    "                                                                       size = 18,\n",
    "                                                                       hjust = 0.5,\n",
    "                                                                       fontface = \"bold\"),\n",
    "                                        ncol = 2)\n",
    "  \n",
    "  \n",
    "  plot <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = paste0(comparison, \n",
    "                                                     \" - GO:Biological Process\"),\n",
    "                                                     size = 24,\n",
    "                                                     hjust = 0.5,\n",
    "                                                     fontface = \"bold\"),\n",
    "                             sub_plot_labels,\n",
    "                             plot_grid(plot_activated, plot_suppressed),\n",
    "                             ncol = 1, rel_heights = c(0.1, .05, 1))\n",
    "\n",
    "  plot\n",
    "  \n",
    "  png(file=\n",
    "    paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "           comparison, \"AGE_MAST_Pathways_gseaGOBPenrich.png\", sep = \"_\"), \n",
    "      width=1400, \n",
    "      height=800)\n",
    "  print(plot)\n",
    "  dev.off()\n",
    "  #----- pathExplore: High-Level GSEA Clusters (GO:BP)\n",
    "  tryCatch({\n",
    "    # fix here: https://github.com/tidyverse/ggplot2/issues/2799\n",
    "    cf <- coord_fixed()\n",
    "    cf$default <- TRUE\n",
    "    plot <- enrichmentNetwork(genes_go@result, drawEllipses = TRUE, fontSize = 2.5) +\n",
    "      ggtitle(paste0(comparison, \" - GO:Biological Process\")) +\n",
    "      theme(plot.title = element_text(hjust = 0.5, face = \"bold\"),\n",
    "            plot.margin = margin(l = 6, t = 12, r = 3, unit = \"pt\")) +\n",
    "      cf +\n",
    "      coord_fixed(clip = \"off\")\n",
    "    print(plot)\n",
    "    ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\",\n",
    "                            comparison, \"AGE_MAST_Pathways_gseaGOBPenrich_pathExplore.png\", \n",
    "                            sep = \"_\"),\n",
    "           plot = plot,\n",
    "           width = 600*2,\n",
    "           height = 400*2,\n",
    "           dpi = 150,\n",
    "           units = \"px\")\n",
    "  },\n",
    "    error=function(e){cat(\"ERROR :\",conditionMessage(e), \"\\n\")})\n",
    "}\n",
    "#----- Reactome\n",
    "reactome <- msigdbr(species = \"Homo sapiens\", \n",
    "                    category = \"C2\", \n",
    "                    subcategory = \"CP:REACTOME\") %>% \n",
    "              dplyr::select(gs_name, gene_symbol)\n",
    "#----- Pathways GSEA REACTOME\n",
    "for (comparison in comparisons){\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_gsea <- GSEA(geneList = genelist,\n",
    "                     minGSSize = 10,\n",
    "                     maxGSSize = 500,\n",
    "                     pvalueCutoff = 0.05,\n",
    "                     pAdjustMethod = \"BH\",\n",
    "                     TERM2GENE = reactome, \n",
    "                     by = \"fgsea\",\n",
    "                     verbose = TRUE)\n",
    "  if (nrow(genes_gsea@result) == 0) {\n",
    "    warning(paste0(\"No significant REACTOME genes for: \", comparison, \"\\nSkipping to next comparison...\"))\n",
    "   } else {\n",
    "    plot <- enrichplot::dotplot(object = genes_gsea,\n",
    "                                showCategory = 10, \n",
    "                                split = \".sign\") + \n",
    "              facet_grid(.~.sign) + \n",
    "              ggtitle(paste0(comparison, \" - GSEA Reactome\")) + \n",
    "              theme(plot.title = element_text(hjust = 0.5, \n",
    "                                              face = \"bold\"))\n",
    "    print(plot)\n",
    "    png(file=\n",
    "      paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "             comparison, \"AGE_MAST_Pathways_gseaREACTOME.png\", sep = \"_\"), \n",
    "        width=1000, \n",
    "        height=600)\n",
    "    print(plot)\n",
    "    dev.off()\n",
    "    ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "             comparison, \"AGE_MAST_Pathways_gseaREACTOME_quality.png\", sep = \"_\"), \n",
    "       plot = plot,\n",
    "       width = 1000*2,\n",
    "       height = 600*2,\n",
    "       dpi = 200,\n",
    "       units = \"px\")\n",
    "    \n",
    "    #----- Split REACTOME Pathway Map by Activated / Suppressed\n",
    "    # (otherwise the activated & suppressed are shown together -- misleading)\n",
    "    genes_gsea_emapplot <- genes_gsea\n",
    "    genes_gsea_emapplot@result <- genes_gsea_emapplot@result %>%\n",
    "      mutate(pathway_direction = case_when(NES > 0 ~ \"activated\",\n",
    "                                           NES < 0 ~ \"suppressed\"))\n",
    "    genes_gsea_emapplot_activated <- dplyr::filter(genes_gsea_emapplot, \n",
    "                                                 pathway_direction == \"activated\")\n",
    "    genes_gsea_emapplot_suppressed <- dplyr::filter(genes_gsea_emapplot, \n",
    "                                                 pathway_direction == \"suppressed\")\n",
    "    \n",
    "    # plot activated\n",
    "    if (nrow(genes_gsea_emapplot_activated@result) == 0) {\n",
    "       warning(paste0(\"No significant activated REACTOME genes for: \", \n",
    "                      comparison, \"\\nSkipping to next comparison...\"))\n",
    "      plot_activated <- ggplot() + theme_void()\n",
    "    } else {\n",
    "      genes_gsea_emapplot_activated <- pairwise_termsim(x = genes_gsea_emapplot_activated)\n",
    "      plot_activated <- genes_gsea_emapplot_activated %>% \n",
    "                enrichplot::emapplot(showCategory=40) +\n",
    "                theme(plot.title = element_text(hjust = 0.5,\n",
    "                                                face = \"bold\"))\n",
    "      print(plot_activated)\n",
    "    }\n",
    "    \n",
    "    # plot suppressed\n",
    "    if (nrow(genes_gsea_emapplot_suppressed@result) == 0) {\n",
    "       warning(paste0(\"No significant suppressed REACTOME genes for: \", \n",
    "                      comparison, \"\\nSkipping to next comparison...\"))\n",
    "      plot_suppressed <- ggplot() + theme_void()\n",
    "    } else {\n",
    "      genes_gsea_emapplot_suppressed <- pairwise_termsim(x = genes_gsea_emapplot_suppressed)\n",
    "      plot_suppressed <- genes_gsea_emapplot_suppressed %>% \n",
    "                enrichplot::emapplot(showCategory=40) +\n",
    "                  theme(plot.title = element_text(hjust = 0.5,\n",
    "                                                  face = \"bold\"))\n",
    "      print(plot_suppressed)\n",
    "    } \n",
    "    \n",
    "    # combine & save plot\n",
    "    sub_plot_labels <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = \"ACTIVATED\",\n",
    "                                                                         size = 18,\n",
    "                                                                         hjust = 0.5,\n",
    "                                                                         fontface = \"bold\"),\n",
    "                                          ggdraw() + cowplot::draw_label(label = \"SUPPRESSED\",\n",
    "                                                                         size = 18,\n",
    "                                                                         hjust = 0.5,\n",
    "                                                                         fontface = \"bold\"),\n",
    "                                          ncol = 2)\n",
    "    \n",
    "    \n",
    "    plot <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = paste0(comparison, \n",
    "                                                       \" - GSEA Reactome\"),\n",
    "                                                       size = 24,\n",
    "                                                       hjust = 0.5,\n",
    "                                                       fontface = \"bold\"),\n",
    "                               sub_plot_labels,\n",
    "                               plot_grid(plot_activated, plot_suppressed),\n",
    "                               ncol = 1, rel_heights = c(0.1, .05, 1))\n",
    "    print(plot)\n",
    "    png(file=\n",
    "      paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "             comparison, \"AGE_MAST_Pathways_gseaREACTOMEenrich.png\", sep = \"_\"), \n",
    "        width=1400, \n",
    "        height=800)\n",
    "    print(plot)\n",
    "    dev.off()\n",
    "    #----- pathExplore: High-Level GSEA Clusters (GSEA Reactome)\n",
    "    tryCatch({\n",
    "      # fix here: https://github.com/tidyverse/ggplot2/issues/2799\n",
    "      cf <- coord_fixed()\n",
    "      cf$default <- TRUE\n",
    "      plot <- pathExplore::enrichmentNetwork(genes_gsea@result, drawEllipses = TRUE, fontSize = 2.5) +\n",
    "        ggtitle(paste0(comparison, \" - GSEA Reactome\")) +\n",
    "        theme(plot.title = element_text(hjust = 0.5, face = \"bold\"),\n",
    "              plot.margin = margin(l = 6, t = 12, r = 3, unit = \"pt\")) +\n",
    "        cf +\n",
    "        coord_fixed(clip = \"off\")\n",
    "      print(plot)\n",
    "      ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\",\n",
    "                              comparison, \"AGE_MAST_Pathways_gseaREACTOMEenrich_pathExplore.png\",\n",
    "                              sep = \"_\"),\n",
    "             plot = plot,\n",
    "             width = 600*2,\n",
    "             height = 400*2,\n",
    "             dpi = 150,\n",
    "             units = \"px\")\n",
    "    },\n",
    "      error=function(e){cat(\"ERROR :\",conditionMessage(e), \"\\n\")})\n",
    "   } \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad369dbb",
   "metadata": {},
   "source": [
    "## Seurat Differential Expression: p16+ & p21+ Co-Expression\n",
    "\n",
    "Find DEG between (p16+ & p21+) and (p16- or p21-) per cell_type\n",
    "Using 0.25 logFC  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21cad9e3",
   "metadata": {
    "eval": false,
    "name": "Seurat_Differential_Expression_p16_p21"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Differential_Expression_p16_p21\n",
    "#----- Load data if needed\n",
    "if (!exists(\"seurat_anno\")) {\n",
    "  seurat_anno <- readRDS(file = \"seurat_results/seurat_anno.rds\")\n",
    "}\n",
    "\n",
    "#----- RNA Assay stores the original and normalized counts\n",
    "# Original counts : seurat_anno@assays[[\"RNA\"]]@counts\n",
    "# Normalized counts : seurat_anno@assays[[\"RNA\"]]@data\n",
    "DefaultAssay(seurat_anno) <- \"RNA\"\n",
    "#----- Group Celltype ID\n",
    "# make this the Idents()\n",
    "metadata_p16_p21 <- seurat_anno@meta.data %>%\n",
    "  mutate(CDKN1A_CDKN2A_Binary = case_when((CDKN1A_Count > 0) & (CDKN2A_Count > 0) ~ \"CDKN1A_CDKN2A_Pos\",\n",
    "                                          (CDKN1A_Count == 0) | (CDKN2A_Count == 0) ~ \"CDKN1A_or_CDKN2A_Neg\"))\n",
    "seurat_anno@meta.data <- metadata_p16_p21\n",
    "rm(metadata_p16_p21)\n",
    "seurat_anno$group_cluster <- paste(seurat_anno$CDKN1A_CDKN2A_Binary, \n",
    "                                   seurat_anno$cluster_annotation, \n",
    "                                   sep = \"_\")\n",
    "Idents(seurat_anno) <- seurat_anno$group_cluster\n",
    "\n",
    "#---------- Differential Expressed Genes\n",
    "#----- Cell Types\n",
    "cell_types <- seurat_anno@meta.data %>% \n",
    "  dplyr::arrange(cluster_annotation) %>% \n",
    "  dplyr::select(cluster_annotation) %>% \n",
    "  unique() %>% \n",
    "  pull(cluster_annotation)\n",
    "#----- Cell_Type Differential Expressed Genes\n",
    "get_deg <- function(seurat_obj, cell_type, ident1, ident2){\n",
    "  FindMarkers(seurat_obj, \n",
    "              ident.1 = paste(ident1, cell_type, sep = \"_\"), \n",
    "              ident.2 = paste(ident2, cell_type, sep = \"_\"), \n",
    "              test.use = \"MAST\", \n",
    "              latent.vars = \"AGE\", \n",
    "              logfc.threshold = 0, \n",
    "              min.pct = 0,\n",
    "              min.diff.pct = -Inf, \n",
    "              min.cells.feature = 10, \n",
    "              min.cells.group = 10, \n",
    "              verbose = FALSE) %>% \n",
    "    rownames_to_column(var = \"Gene_Symbol\") %>% \n",
    "    left_join(Gene_Emsembl_Symbol, by=\"Gene_Symbol\") %>% \n",
    "    mutate(Comparison = paste(ident1, \"vs\", ident2, cell_type, sep = \"_\"))\n",
    "}\n",
    "Differential_Expression <- lapply(cell_types, function(cell_type) {\n",
    "  tryCatch(get_deg(seurat_obj = seurat_anno,\n",
    "                   cell_type = cell_type,\n",
    "                   ident1 = \"CDKN1A_CDKN2A_Pos\",\n",
    "                   ident2 = \"CDKN1A_or_CDKN2A_Neg\"),\n",
    "           error = function(e) \n",
    "             # empty dataframe with same column headers & types...\n",
    "             data.frame(Gene_Symbol = character(),\n",
    "                        p_val = double(),\n",
    "                        avg_log2FC = double(),\n",
    "                        pct.1 = double(),\n",
    "                        pct.2 = double(),\n",
    "                        p_val_adj = double(),\n",
    "                        Gene_Emsembl = character(),\n",
    "                        Comparison = character(),\n",
    "                        stringsAsFactors = FALSE))\n",
    "}) %>% bind_rows()\n",
    "#----- save the results\n",
    "write_delim(Differential_Expression, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN1A_CDKN2A_Pos_vs_CDKN1A_or_CDKN2A_Neg_AGE_MAST.txt\"), \n",
    "  delim = \"\\t\")\n",
    "#----- Number of DEGs\n",
    "Differential_Expression_count <- Differential_Expression %>% \n",
    "  mutate(Significant = case_when(\n",
    "           p_val_adj <= 0.05 & avg_log2FC >= 0.25 ~ 'Upregulated',\n",
    "           p_val_adj <= 0.05 & avg_log2FC <= -0.25 ~ 'Downregulated',\n",
    "           TRUE ~ 'Not_Significant')) %>% \n",
    "  dplyr::count(Comparison, Significant) %>% \n",
    "  dplyr::rename(Count = n) %>% \n",
    "  tidyr::spread(Significant, Count)\n",
    "#----- save the results\n",
    "write_delim(Differential_Expression_count, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN1A_CDKN2A_Pos_vs_CDKN1A_or_CDKN2A_Neg_AGE_MAST_count.txt\"), \n",
    "  delim = \"\\t\")\n",
    "\n",
    "\n",
    "#---------- Differential Expressed Genes Pathways\n",
    "#----- Comparisons\n",
    "comparisons <- Differential_Expression$Comparison %>% \n",
    "                 unique()\n",
    "#----- Pathways GSEA GOBP\n",
    "Differential_Expression_GOBP <- lapply(comparisons, function(comparison) {\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_go <- gseGO(geneList = genelist, \n",
    "                    ont =\"BP\", \n",
    "                    OrgDb = get(\"org.Hs.eg.db\"),\n",
    "                    keyType = \"SYMBOL\", \n",
    "                    minGSSize = 10, \n",
    "                    maxGSSize = 500, \n",
    "                    pvalueCutoff = 0.05, \n",
    "                    pAdjustMethod = \"BH\",\n",
    "                    verbose = TRUE)\n",
    "  return(genes_go@result %>% mutate(Comparison = comparison))\n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(Differential_Expression_GOBP, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN1A_CDKN2A_Pos_vs_CDKN1A_or_CDKN2A_Neg_AGE_MAST_Pathways_GSEA_GOBP.txt\"), \n",
    "  delim = \"\\t\")\n",
    "#----- Reactome\n",
    "reactome <- msigdbr(species = \"Homo sapiens\", \n",
    "                    category = \"C2\", \n",
    "                    subcategory = \"CP:REACTOME\") %>% \n",
    "              dplyr::select(gs_name, gene_symbol)\n",
    "#----- Pathways GSEA REACTOME\n",
    "Differential_Expression_REACTOME <- lapply(comparisons, function(comparison) {\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_gsea <- GSEA(geneList = genelist,\n",
    "                     minGSSize = 10,\n",
    "                     maxGSSize = 500,\n",
    "                     pvalueCutoff = 0.05,\n",
    "                     pAdjustMethod = \"BH\",\n",
    "                     TERM2GENE = reactome, \n",
    "                     by = \"fgsea\",\n",
    "                     verbose = TRUE)\n",
    "  return(genes_gsea@result %>% mutate(Comparison = comparison))\n",
    "}) %>% bind_rows()\n",
    "#----- Save Results\n",
    "write_delim(Differential_Expression_REACTOME, \n",
    "  paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN1A_CDKN2A_Pos_vs_CDKN1A_or_CDKN2A_Neg_AGE_MAST_Pathways_GSEA_REACTOME.txt\"), \n",
    "  delim = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2c50012",
   "metadata": {
    "name": "Seurat_Differential_Expression_Figures_p16_p21"
   },
   "outputs": [],
   "source": [
    "#---------- Seurat_Differential_Expression_Figures_p16_p21\n",
    "Differential_Expression <- \n",
    "  read_delim(paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN1A_CDKN2A_Pos_vs_CDKN1A_or_CDKN2A_Neg_AGE_MAST.txt\"), \n",
    "  delim = \"\\t\", escape_double = FALSE, \n",
    "  trim_ws = TRUE)\n",
    "\n",
    "#----- Differential_Expression_count\n",
    "Differential_Expression_count <- \n",
    "  read_delim(paste0(\"seurat_results/\", NameProject, \"_Differential_Expression_CDKN1A_CDKN2A_Pos_vs_CDKN1A_or_CDKN2A_Neg_AGE_MAST_count.txt\"), \n",
    "  delim = \"\\t\", escape_double = FALSE, \n",
    "  trim_ws = TRUE)\n",
    "Differential_Expression_count %>% \n",
    "  kable(caption=\"Number of Differentially Expressed Genes\", \n",
    "        format.args = list(big.mark = \",\")) %>% \n",
    "  kable_styling(c(\"striped\", \"bordered\"))\n",
    "\n",
    "\n",
    "#---------- Differential Expressed Genes Volcano\n",
    "#----- Comparisons\n",
    "comparisons <- Differential_Expression$Comparison %>% \n",
    "                 unique()\n",
    "#----- Differential Expressed Genes Volcano\n",
    "for (comparison in comparisons){\n",
    "  plot <- Differential_Expression %>% \n",
    "            dplyr::filter(Comparison == comparison) %>%\n",
    "            dplyr::select(Comparison, \n",
    "                          Gene_Symbol, \n",
    "                          avg_log2FC, \n",
    "                          p_val,\n",
    "                          p_val_adj) %>% \n",
    "            mutate(Significant = case_when(\n",
    "                     p_val_adj <= 0.05 & avg_log2FC >= 0.25 ~ 'Upregulated',\n",
    "                     p_val_adj <= 0.05 & avg_log2FC <= -0.25 ~ 'Downregulated',\n",
    "                     TRUE ~ 'Not_Significant')) %>% \n",
    "            mutate(Label = case_when(\n",
    "                     Significant == \"Upregulated\" ~ Gene_Symbol,\n",
    "                     Significant == \"Downregulated\" ~ Gene_Symbol,\n",
    "                     TRUE ~ NA_character_)) %>% \n",
    "            ggplot(mapping = aes(x=avg_log2FC, \n",
    "                   y = -log10(p_val), \n",
    "                   color = Significant,\n",
    "                   label = Label)) + \n",
    "              geom_point(alpha = 0.5, stroke=0, size=2) + \n",
    "              geom_text_repel() +\n",
    "              ggtitle(comparison) + \n",
    "              xlab(\"log2 Fold Change\") +\n",
    "              ylab(\"-log10(P Value)\") + \n",
    "              theme_classic() + \n",
    "              theme(plot.title = element_text(hjust = 0.5, \n",
    "                                              face = \"bold\")) + \n",
    "              scale_colour_manual(values = c('Downregulated' = 'Blue',\n",
    "                                             'Not_Significant' = 'Gray',\n",
    "                                             'Upregulated' = 'Red')) + \n",
    "              geom_vline(xintercept = c(-0.25,0.25), \n",
    "                         linetype = \"dotted\")\n",
    "  print(plot)\n",
    "  # png(file = paste(\"seurat_results/figures/Seurat_Differential_Expression\", comparison, \"AGE_MAST_volcano.png\", sep = \"_\"),\n",
    "  #   width = 600*2, \n",
    "  #   height = 400*2,\n",
    "  #   pointsize = 6)\n",
    "  # print(plot)\n",
    "  # dev.off()\n",
    "  ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\", comparison, \"AGE_MAST_volcano.png\", sep = \"_\"),\n",
    "         plot = plot,\n",
    "         width = 600*2,\n",
    "         height = 400*2,\n",
    "         dpi = 150,\n",
    "         units = \"px\")\n",
    "}\n",
    "\n",
    "\n",
    "#---------- Differential Expressed Genes Pathways\n",
    "#----- Comparisons\n",
    "comparisons <- Differential_Expression$Comparison %>% \n",
    "                 unique()\n",
    "#----- Pathways GSEA GOBP\n",
    "for (comparison in comparisons){\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_go <- gseGO(geneList = genelist, \n",
    "                    ont =\"BP\", \n",
    "                    OrgDb = get(\"org.Hs.eg.db\"),\n",
    "                    keyType = \"SYMBOL\", \n",
    "                    minGSSize = 10, \n",
    "                    maxGSSize = 500, \n",
    "                    pvalueCutoff = 0.05, \n",
    "                    pAdjustMethod = \"BH\",\n",
    "                    verbose = TRUE)\n",
    "  if (nrow(genes_go@result) == 0) {\n",
    "     warning(paste0(\"No significant GOBP genes for: \", comparison, \"\\nSkipping to next comparison...\"))\n",
    "   } else {\n",
    "     plot <- enrichplot::dotplot(object = genes_go, \n",
    "                                  showCategory = 10, \n",
    "                                  split = \".sign\") + \n",
    "                facet_grid(.~.sign) + \n",
    "                ggtitle(paste0(comparison, \" - GO:Biological Process\")) + \n",
    "                theme(plot.title = element_text(hjust = 0.5, \n",
    "                                                face = \"bold\"))\n",
    "      print(plot)\n",
    "      png(file=\n",
    "        paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "               comparison, \"AGE_MAST_Pathways_gseaGOBP.png\", sep = \"_\"), \n",
    "          width=1000, \n",
    "          height=600)\n",
    "      print(plot)\n",
    "      dev.off()\n",
    "   } \n",
    "  \n",
    "  #----- Split GOBP Pathway Map by Activated / Suppressed\n",
    "  # (otherwise the activated & suppressed are shown together -- misleading)\n",
    "  genes_go_emapplot <- genes_go\n",
    "  genes_go_emapplot@result <- genes_go_emapplot@result %>%\n",
    "    mutate(pathway_direction = case_when(NES > 0 ~ \"activated\",\n",
    "                                         NES < 0 ~ \"suppressed\"))\n",
    "  genes_go_emapplot_activated <- dplyr::filter(genes_go_emapplot, \n",
    "                                               pathway_direction == \"activated\")\n",
    "  genes_go_emapplot_suppressed <- dplyr::filter(genes_go_emapplot, \n",
    "                                               pathway_direction == \"suppressed\")\n",
    "  \n",
    "  # plot activated\n",
    "  if (nrow(genes_go_emapplot_activated@result) == 0) {\n",
    "     warning(paste0(\"No significant activated GOBP genes for: \", \n",
    "                    comparison, \"\\nSkipping to next comparison...\"))\n",
    "    plot_activated <- ggplot() + theme_void()\n",
    "  } else {\n",
    "    genes_go_emapplot_activated <- pairwise_termsim(x = genes_go_emapplot_activated)\n",
    "    plot_activated <- genes_go_emapplot_activated %>% \n",
    "              enrichplot::emapplot(showCategory=40) +\n",
    "              theme(plot.title = element_text(hjust = 0.5,\n",
    "                                              face = \"bold\"))\n",
    "    print(plot_activated)\n",
    "  }\n",
    "  \n",
    "  # plot suppressed\n",
    "  if (nrow(genes_go_emapplot_suppressed@result) == 0) {\n",
    "     warning(paste0(\"No significant suppressed GOBP genes for: \", \n",
    "                    comparison, \"\\nSkipping to next comparison...\"))\n",
    "    plot_suppressed <- ggplot() + theme_void()\n",
    "  } else {\n",
    "    genes_go_emapplot_suppressed <- pairwise_termsim(x = genes_go_emapplot_suppressed)\n",
    "    plot_suppressed <- genes_go_emapplot_suppressed %>% \n",
    "              enrichplot::emapplot(showCategory=40) +\n",
    "                theme(plot.title = element_text(hjust = 0.5,\n",
    "                                                face = \"bold\"))\n",
    "    print(plot_suppressed)\n",
    "  } \n",
    "  \n",
    "  # combine & save plot\n",
    "  sub_plot_labels <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = \"ACTIVATED\",\n",
    "                                                                       size = 18,\n",
    "                                                                       hjust = 0.5,\n",
    "                                                                       fontface = \"bold\"),\n",
    "                                        ggdraw() + cowplot::draw_label(label = \"SUPPRESSED\",\n",
    "                                                                       size = 18,\n",
    "                                                                       hjust = 0.5,\n",
    "                                                                       fontface = \"bold\"),\n",
    "                                        ncol = 2)\n",
    "  \n",
    "  \n",
    "  plot <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = paste0(comparison, \n",
    "                                                     \" - GO:Biological Process\"),\n",
    "                                                     size = 24,\n",
    "                                                     hjust = 0.5,\n",
    "                                                     fontface = \"bold\"),\n",
    "                             sub_plot_labels,\n",
    "                             plot_grid(plot_activated, plot_suppressed),\n",
    "                             ncol = 1, rel_heights = c(0.1, .05, 1))\n",
    "\n",
    "  plot\n",
    "  \n",
    "  png(file=\n",
    "    paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "           comparison, \"AGE_MAST_Pathways_gseaGOBPenrich.png\", sep = \"_\"), \n",
    "      width=1400, \n",
    "      height=800)\n",
    "  print(plot)\n",
    "  dev.off()\n",
    "  #----- pathExplore: High-Level GSEA Clusters (GO:BP)\n",
    "  tryCatch({\n",
    "    # fix here: https://github.com/tidyverse/ggplot2/issues/2799\n",
    "    cf <- coord_fixed()\n",
    "    cf$default <- TRUE\n",
    "    plot <- enrichmentNetwork(genes_go@result, drawEllipses = TRUE, fontSize = 2.5) +\n",
    "      ggtitle(paste0(comparison, \" - GO:Biological Process\")) +\n",
    "      theme(plot.title = element_text(hjust = 0.5, face = \"bold\"),\n",
    "            plot.margin = margin(l = 6, t = 12, r = 3, unit = \"pt\")) +\n",
    "      cf +\n",
    "      coord_fixed(clip = \"off\")\n",
    "    print(plot)\n",
    "    ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\",\n",
    "                            comparison, \"AGE_MAST_Pathways_gseaGOBPenrich_pathExplore.png\", \n",
    "                            sep = \"_\"),\n",
    "           plot = plot,\n",
    "           width = 600*2,\n",
    "           height = 400*2,\n",
    "           dpi = 150,\n",
    "           units = \"px\")\n",
    "  },\n",
    "    error=function(e){cat(\"ERROR :\",conditionMessage(e), \"\\n\")})\n",
    "}\n",
    "#----- Reactome\n",
    "reactome <- msigdbr(species = \"Homo sapiens\", \n",
    "                    category = \"C2\", \n",
    "                    subcategory = \"CP:REACTOME\") %>% \n",
    "              dplyr::select(gs_name, gene_symbol)\n",
    "#----- Pathways GSEA REACTOME\n",
    "for (comparison in comparisons){\n",
    "  genes <- Differential_Expression %>% \n",
    "    dplyr::filter(Comparison == comparison) %>%\n",
    "    dplyr::select(Comparison, \n",
    "                  Gene_Symbol, \n",
    "                  avg_log2FC, \n",
    "                  p_val) %>%\n",
    "    dplyr::mutate(p_val = case_when(p_val == 0 ~ 5e-324,\n",
    "                                    TRUE ~ p_val)) %>%\n",
    "    dplyr::mutate(sign_log10Pvalue = sign(avg_log2FC) * -log10(p_val)) %>% \n",
    "    dplyr::arrange(sign_log10Pvalue)\n",
    "  genelist <- setNames(genes$sign_log10Pvalue, \n",
    "                       genes$Gene_Symbol) %>% \n",
    "    sort(decreasing = TRUE)\n",
    "  genes_gsea <- GSEA(geneList = genelist,\n",
    "                     minGSSize = 10,\n",
    "                     maxGSSize = 500,\n",
    "                     pvalueCutoff = 0.05,\n",
    "                     pAdjustMethod = \"BH\",\n",
    "                     TERM2GENE = reactome, \n",
    "                     by = \"fgsea\",\n",
    "                     verbose = TRUE)\n",
    "  if (nrow(genes_gsea@result) == 0) {\n",
    "    warning(paste0(\"No significant REACTOME genes for: \", comparison, \"\\nSkipping to next comparison...\"))\n",
    "   } else {\n",
    "    plot <- enrichplot::dotplot(object = genes_gsea,\n",
    "                                showCategory = 10, \n",
    "                                split = \".sign\") + \n",
    "              facet_grid(.~.sign) + \n",
    "              ggtitle(paste0(comparison, \" - GSEA Reactome\")) + \n",
    "              theme(plot.title = element_text(hjust = 0.5, \n",
    "                                              face = \"bold\"))\n",
    "    print(plot)\n",
    "    png(file=\n",
    "      paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "             comparison, \"AGE_MAST_Pathways_gseaREACTOME.png\", sep = \"_\"), \n",
    "        width=1000, \n",
    "        height=600)\n",
    "    print(plot)\n",
    "    dev.off()\n",
    "    ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "             comparison, \"AGE_MAST_Pathways_gseaREACTOME_quality.png\", sep = \"_\"), \n",
    "       plot = plot,\n",
    "       width = 1000*2,\n",
    "       height = 600*2,\n",
    "       dpi = 200,\n",
    "       units = \"px\")\n",
    "    \n",
    "    #----- Split REACTOME Pathway Map by Activated / Suppressed\n",
    "    # (otherwise the activated & suppressed are shown together -- misleading)\n",
    "    genes_gsea_emapplot <- genes_gsea\n",
    "    genes_gsea_emapplot@result <- genes_gsea_emapplot@result %>%\n",
    "      mutate(pathway_direction = case_when(NES > 0 ~ \"activated\",\n",
    "                                           NES < 0 ~ \"suppressed\"))\n",
    "    genes_gsea_emapplot_activated <- dplyr::filter(genes_gsea_emapplot, \n",
    "                                                 pathway_direction == \"activated\")\n",
    "    genes_gsea_emapplot_suppressed <- dplyr::filter(genes_gsea_emapplot, \n",
    "                                                 pathway_direction == \"suppressed\")\n",
    "    \n",
    "    # plot activated\n",
    "    if (nrow(genes_gsea_emapplot_activated@result) == 0) {\n",
    "       warning(paste0(\"No significant activated REACTOME genes for: \", \n",
    "                      comparison, \"\\nSkipping to next comparison...\"))\n",
    "      plot_activated <- ggplot() + theme_void()\n",
    "    } else {\n",
    "      genes_gsea_emapplot_activated <- pairwise_termsim(x = genes_gsea_emapplot_activated)\n",
    "      plot_activated <- genes_gsea_emapplot_activated %>% \n",
    "                enrichplot::emapplot(showCategory=40) +\n",
    "                theme(plot.title = element_text(hjust = 0.5,\n",
    "                                                face = \"bold\"))\n",
    "      print(plot_activated)\n",
    "    }\n",
    "    \n",
    "    # plot suppressed\n",
    "    if (nrow(genes_gsea_emapplot_suppressed@result) == 0) {\n",
    "       warning(paste0(\"No significant suppressed REACTOME genes for: \", \n",
    "                      comparison, \"\\nSkipping to next comparison...\"))\n",
    "      plot_suppressed <- ggplot() + theme_void()\n",
    "    } else {\n",
    "      genes_gsea_emapplot_suppressed <- pairwise_termsim(x = genes_gsea_emapplot_suppressed)\n",
    "      plot_suppressed <- genes_gsea_emapplot_suppressed %>% \n",
    "                enrichplot::emapplot(showCategory=40) +\n",
    "                  theme(plot.title = element_text(hjust = 0.5,\n",
    "                                                  face = \"bold\"))\n",
    "      print(plot_suppressed)\n",
    "    } \n",
    "    \n",
    "    # combine & save plot\n",
    "    sub_plot_labels <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = \"ACTIVATED\",\n",
    "                                                                         size = 18,\n",
    "                                                                         hjust = 0.5,\n",
    "                                                                         fontface = \"bold\"),\n",
    "                                          ggdraw() + cowplot::draw_label(label = \"SUPPRESSED\",\n",
    "                                                                         size = 18,\n",
    "                                                                         hjust = 0.5,\n",
    "                                                                         fontface = \"bold\"),\n",
    "                                          ncol = 2)\n",
    "    \n",
    "    \n",
    "    plot <- cowplot::plot_grid(ggdraw() + cowplot::draw_label(label = paste0(comparison, \n",
    "                                                       \" - GSEA Reactome\"),\n",
    "                                                       size = 24,\n",
    "                                                       hjust = 0.5,\n",
    "                                                       fontface = \"bold\"),\n",
    "                               sub_plot_labels,\n",
    "                               plot_grid(plot_activated, plot_suppressed),\n",
    "                               ncol = 1, rel_heights = c(0.1, .05, 1))\n",
    "    print(plot)\n",
    "    png(file=\n",
    "      paste(\"seurat_results/figures/Seurat_Differential_Expression\", \n",
    "             comparison, \"AGE_MAST_Pathways_gseaREACTOMEenrich.png\", sep = \"_\"), \n",
    "        width=1400, \n",
    "        height=800)\n",
    "    print(plot)\n",
    "    dev.off()\n",
    "    #----- pathExplore: High-Level GSEA Clusters (GSEA Reactome)\n",
    "    tryCatch({\n",
    "      # fix here: https://github.com/tidyverse/ggplot2/issues/2799\n",
    "      cf <- coord_fixed()\n",
    "      cf$default <- TRUE\n",
    "      plot <- pathExplore::enrichmentNetwork(genes_gsea@result, drawEllipses = TRUE, fontSize = 2.5) +\n",
    "        ggtitle(paste0(comparison, \" - GSEA Reactome\")) +\n",
    "        theme(plot.title = element_text(hjust = 0.5, face = \"bold\"),\n",
    "              plot.margin = margin(l = 6, t = 12, r = 3, unit = \"pt\")) +\n",
    "        cf +\n",
    "        coord_fixed(clip = \"off\")\n",
    "      print(plot)\n",
    "      ggsave(filename = paste(\"seurat_results/figures/Seurat_Differential_Expression\",\n",
    "                              comparison, \"AGE_MAST_Pathways_gseaREACTOMEenrich_pathExplore.png\",\n",
    "                              sep = \"_\"),\n",
    "             plot = plot,\n",
    "             width = 600*2,\n",
    "             height = 400*2,\n",
    "             dpi = 150,\n",
    "             units = \"px\")\n",
    "    },\n",
    "      error=function(e){cat(\"ERROR :\",conditionMessage(e), \"\\n\")})\n",
    "   } \n",
    "}\n"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "name,tags,fig.cap,out.width,eval,-all",
   "main_language": "R",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "scverse10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
